---
title: "cscplots Tutorial 1"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cscplots Tutorial 1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### __Tutorial 1: Identifying and Characterising Cancer Stem Cells In SCLC CDX Single-Cell RNA-Seq__

cscplots is a collection of functions that can be combined to create a pipeline, to achieve the objective of finding
and characterising putative cancer stem cells.

Results from applying this method to scRNA-Seq from several cancer types, will be published in the literature
later this year. The tutorials shown here replicate some of these results, while trying to familiarise the reader with
this pipeline, so that you may generate similar results with your own scRNA-Seq data.

During the construction and use of the pipeline, we will use t-SNE plots to identify high transcription cells
with predicted high stemmess compared to the other cells from their cell population. This will require prediction
of stemness in each cancer cell, using several similar machine learning methods - different implementations of
random forests. We'll use this information to identify a subset of cells as putative stem cells, and then
analyse them for preferential expression, relative to the other cells in the cell population they are selected
from. This will aim to identifty genes that are driving these cells, which are in turn driving the tumour they are part of.
In addition, understanding which genes are elevated and suppressed in cancer stem cells (CSCs), can reveal how the tumour
is resisting drug treatments, and achieving immune evasion, by giving CSCs exceptionally hardy molecular characteristics.

Tutorial 1 shows how to apply this methodology to SCLC CDX scRNA-Seq expression matrices published by Byers et al. (Nature Cancer 2020).
This example uses only raw read counts, and we'll focus on just two of the samples from this dataset, for brevity. A CDX
is a tumour produced from patient circulating tumour cells (CTCs).

Tutorial 2 will use a similar pipeline, this time applied to melanoma data published by Tirosh et al. (Science 2016).
This is a more complex dataset in that the 4558 cells in this population are of different types,
i.e. malginant cancer cells, immune cells, fibroblasts, etc., and is sourced from 19 patients. We'll focus on CSCs from
eight of these patients' tumours.

OK, let's get started!

### Step 1: Load SCLC CDX scRNA-Seq Expression Matrices ##################################################################################

Firstly, prepare the CDX expression data included data with this package.

The two samples they're from, are taken from tumours created from CTCs, from the same SCLC patient.

GSM4104152_SC4.LB17009 is from a CDX created before the patient was treated with Talazoparib; GSM4104161_SC4_Talazoparib.LB17011
is from another CDX generated after this patient received clinical treatment with Talazoparib. (Talazoparib is a PARP inhibitor
that interferes with DNA repair in the cancer cells.)

```{r message=FALSE}
library("Matrix")
library("gtools")
library("Rtsne")
library("caret")
library("ranger")
library("e1071")
library("randomForest")
library("ggplot2")
library("ggrepel"); options(ggrepel.max.overlaps = Inf)   
library("cscplots")
```

```{r}
project_name   <- "SCLC_CDXs_project" 
wd             <- './' # set the working directory
saveDir        <- paste0(wd, 'save/')
resultsBaseDir <- paste0('results/')
resultsDir     <- paste0('results/', project_name, '/')
dataDir        <- paste0(wd, '../data/')
dir.create(saveDir, showWarnings = FALSE)
dir.create(resultsBaseDir, showWarnings = FALSE)
dir.create(resultsDir, showWarnings = FALSE)
dir.create(dataDir, showWarnings = FALSE)

CDX_sample_IDs <- c(
    'GSM4104161_SC4_Talazoparib.LB17011',
    'GSM4104152_SC4.LB17009'
)
number_of_samples <- length(CDX_sample_IDs)

resultsDirs  <- list()
tSNEsDirs    <- list()
pCSCsDirs    <- list()
prefExprDirs <- list()

for(i in 1:length(CDX_sample_IDs)){
    resultsDirs[i]  <- paste0(resultsDir, CDX_sample_IDs[i], '/')
    tSNEsDirs[i]    <- paste0(resultsDir, CDX_sample_IDs[i], '/tSNEs/')
    pCSCsDirs[i]    <- paste0(resultsDir, CDX_sample_IDs[i], '/tSNEs/pCSC_group_selection/')
    prefExprDirs[i] <- paste0(resultsDir, CDX_sample_IDs[i], '/preferential_expression/')
    
    dir.create(resultsDirs[[i]], showWarnings = FALSE)
    dir.create(tSNEsDirs[[i]], showWarnings = FALSE)
    dir.create(pCSCsDirs[[i]], showWarnings = FALSE)
    dir.create(prefExprDirs[[i]], showWarnings = FALSE)
}
```

The CDX data we'll work with is available in this package, but if you wanted to download it from the NIH, you could uncomment
the following code:

```{r}
#CDX_urls <- c(
#'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4104nnn/GSM4104161/suppl/GSM4104161_SC4_Talazoparib.LB17011.matrix.tar.gz',
#'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4104nnn/GSM4104152/suppl/GSM4104152_SC4.LB17009.matrix.tar.gz')
#        
#for(i in 1:length(CDX_sample_IDs)){
#    sample_ID <- CDX_sample_IDs[[i]]
#    #cat(paste0(sample_ID, "\n"))
#    gz_file_name <- paste0(sample_ID, '.matrix.tar.gz') 
#    gz_path   <- paste0(wd, sample_ID, '/', gz_file_name)
#    dir.create(file.path(wd, sample_ID), showWarnings = FALSE)
#    download.file(CDX_urls[[i]], gz_path) # TEMP TURN OFF
#    untar(gz_path, list=TRUE)
#    untar(gz_path, exdir=paste0(wd, sample_ID))
#    #unlink(gz_file_name)
#}
#
#number_of_samples <- length(CDX_sample_IDs)
#CDX_tRawCounts_list <- list()
#
#for(i in 1:number_of_samples){
#    this_sample_ID <- CDX_sample_IDs[[i]]
#    matrix_dir <- paste0(wd, this_sample_ID, '/')
#    barcode_path  <- paste0(matrix_dir, "barcodes.tsv")
#    gene_path     <- paste0(matrix_dir, "genes.tsv")
#    matrix_path   <- paste0(matrix_dir, "matrix.mtx")
#    this_matrix <- readMM(file = matrix_path)
#    gene_names = read.delim(gene_path, 
#                            header = FALSE,
#                            stringsAsFactors = FALSE)
#    barcode_names = read.delim(barcode_path, 
#                               header = FALSE,
#                               stringsAsFactors = FALSE)
#    colnames(this_matrix) = barcode_names[[1]]
#    rownames(this_matrix) = gene_names[[2]]
#    this_matrix_no_duplicates <- this_matrix[which(!duplicated(rownames(this_matrix))),]
#    t_matrix <- t(this_matrix_no_duplicates)
#    tRawCounts <- as.matrix(t_matrix)
#    tRawCounts_reorder_pos <- order(colnames(tRawCounts))
#    tRawCounts_reordered <- tRawCounts[,tRawCounts_reorder_pos]
#    this_tRawCounts <- tRawCounts_reordered
#    CDX_tRawCounts_list[[i]] <- this_tRawCounts
#    names(CDX_tRawCounts_list)[[i]] <- this_sample_ID
#}
#save(CDX_tRawCounts_list, file = paste0(saveDir, "CDX_tRawCounts_list.RData"))
##load(paste0(saveDir, "CDX_tRawCounts_list.RData"))
```

We'll use the data already in the package though (which is the same):

```{r}
CDX_tRawCounts_list <- list()
# data automatically available in this R package:
CDX_tRawCounts_list[[1]] <- CDX_tRawCounts_GSM4104161_SC4_Talazoparib.LB17011 
CDX_tRawCounts_list[[2]] <- CDX_tRawCounts_GSM4104152_SC4.LB17009
names(CDX_tRawCounts_list) <- CDX_sample_IDs
```

----------

### Step 2: t-SNE Generation ###########################################################################################

Here, we produce t-SNEs using the SCLC CDX scRNA-Seq matrices, just stored in CDX_tRawCounts_list.

The recommended perplexity value used for the t-SNE is 70, but you can use another value, or a list of values, if you prefer.

This may take some time to run (hours).

```{r}
CDX_tSNE_list <- list()

#perplexityValues <<- c(5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130) # Full range
#perplexityValues <<- c(30,70)
perplexityValues <<- c(70) # Recommended
perplexityValuesPrint <- add_zeros_to_perplexity_values(perplexityValues)
tSNE_run_names        <- paste0("tSNE_", perplexityValuesPrint)

CDX_tRawCounts_list <- list()
CDX_tRawCounts_list[[1]] <- CDX_tRawCounts_GSM4104161_SC4_Talazoparib.LB17011
CDX_tRawCounts_list[[2]] <- CDX_tRawCounts_GSM4104152_SC4.LB17009
names(CDX_tRawCounts_list) <- CDX_sample_IDs

set.seed(1) # for reproducibility

for(i in 1:number_of_samples){
    this_sample_ID  <- names(CDX_tRawCounts_list)[[i]]
    this_tRawCounts <- CDX_tRawCounts_list[[i]]

    this_raw_tSNE_results <- run_all_tSNEs(perplexityValues, this_tRawCounts, paste0("rawCounts_",
                                           this_sample_ID), outDir <- tSNEsDirs[[i]], montage <- 0)
    save(this_raw_tSNE_results, file = paste0(saveDir, "CDX_tSNE_", this_sample_ID, ".RData"))

    CDX_tSNE_list[[i]] <- this_raw_tSNE_results
    names(CDX_tSNE_list)[[i]] <- this_sample_ID
}
save(CDX_tSNE_list, file = paste0(saveDir, "CDX_tSNE_list.RData"))
#load(paste0(saveDir, "CDX_tSNE_list.RData"))
```

----------

### Step 3: Stemness Predictions #######################################################################################

Each cell, in the SCLC CDX population of cells, will be scored using the predicted probability of it being a stem cell.

These stemness scores will be later used, to assist the identification putative cancer stem cells.

These scores are produced using machine learning methods - in this example they are random forests (RFs). These are trained on stem cell RNA-Seq
expression data from induced pluripotent stem cells from the HipSci consortium. After the RFs are trained on the HipSci data, they are used to
predict the likelihood that each of the cancer cells are stem cells.

The HipSci data is an expression matrix describing 196 samples; some of the samples and fibroblasts or mononuclear cells, but most are from iPSC
bulk RNA-Seq samples.

#### Run the Machine Learning Methods to Predict Stemness in Each of the Cancer Cells

```{r}
print(hipsci_training_data[1:2,1:5]) # data automatically available in this R package
```

Now we'll use the run_ML_stemness_predictors function in the cscplots package, to use three random forest implementations (ranger, rf, randomForest) to predict
stemness in each of the cancer cells in the two samples we will analyse (GSM4104161_SC4_Talazoparib.LB17011 & GSM4104152_SC4.LB17009).
The number of genes compared, between the HipSci and cancer expression datasets, will be limited to 16,000. If your computational resources are limited, or you
would like the predictions to run more quickly, you can try using less, but lower than 10,000 is not recommeded.

```{r}
number_of_genes_to_use <- 16000
#number_of_genes_to_use <- 10000 # can give reasonable results
#number_of_genes_to_use <- 16700 # fails on my R installation

CDX_stemness_prediction_DFs_lists <- list()
CDX_stemness_prediction_DFs_lists[[1]] <- run_ML_stemness_predictors(model_type <- 'ranger',
                                                                     CDX_tRawCounts_list,
                                                                     hipsci_training_data,
                                                                     stemness_type <- 'hipsci',
                                                                     number_of_genes_to_use,
                                                                     sample_data_type <- 'raw',
                                                                     resultsDir)
names(CDX_stemness_prediction_DFs_lists)[1] <- "ranger_hipsci_raw"
CDX_stemness_prediction_DFs_lists[[2]] <- run_ML_stemness_predictors(model_type <- 'rf',
                                                                     CDX_tRawCounts_list,
                                                                     hipsci_training_data,
                                                                     stemness_type <- 'hipsci',
                                                                     number_of_genes_to_use,
                                                                     sample_data_type <- 'raw',
                                                                     resultsDir)
names(CDX_stemness_prediction_DFs_lists)[2] <- "rf_hipsci_raw"
CDX_stemness_prediction_DFs_lists[[3]] <- run_ML_stemness_predictors(model_type <- 'randomForest',
                                                                     CDX_tRawCounts_list,
                                                                     hipsci_training_data,
                                                                     stemness_type <- 'hipsci',
                                                                     number_of_genes_to_use,
                                                                     sample_data_type <- 'raw',
                                                                     resultsDir)
names(CDX_stemness_prediction_DFs_lists)[3] <- "randomForest_hipsci_raw"

save(CDX_stemness_prediction_DFs_lists, file = paste0(saveDir, "CDX_stemness_prediction_DFs_lists.RData"))
#load(paste0(saveDir, "CDX_stemness_prediction_DFs_lists.RData"))
```

----------

### Step 4: Genes Correlated with Elevated Stemness and/or High Transcription ####################################################################

Now that we have predicted stemness for each cancer cell, we can use that to see which genes/transcripts are most positively or negatively correlated
with stemness in the two cancer cell populations, i.e. genes that are expressed more, or less, as stemness goes up

In a mixed cell population, or a population composed of cells from multiple samples and/or patients, this could be calculated in subsets of
the cell population. The subsets could be for separate tumours, or subsets by cell type, or a combination of both, e.g. groups of malignant cancer cells
separate patients.

However, here we will calculate for each of the two samples being studied correlations for all cells in each sample. The most postively correlated
will later be used as stemness markers, to assist identification of cancer stem cells.

Genes correlated with transcription increases (total transcription and total transcribed genes) in the cancer cell populations will also be calculated
for both samples, in step 4b.

#### __4a)__: Calculate Stemness Correlations

Display the top genes most correlated with stemness, for each of the three predictions of stemness that we performed in step 3:

```{r}
CDX_stemness_correlations_list <- list()
for(i in 1:length(CDX_tRawCounts_list)){
    tCounts_to_use <- CDX_tRawCounts_list[[i]]

    scores_to_use_list <- list()
    for(j in 1:length(CDX_stemness_prediction_DFs_lists)){
        scores_to_use_list[[j]] <- CDX_stemness_prediction_DFs_lists[[j]][[i]]
    }
    names(scores_to_use_list) <- names(CDX_stemness_prediction_DFs_lists)

    for(k in 1:length(scores_to_use_list)){
        score_to_use_index <- which(names(scores_to_use_list[[k]]) == 'IPS') # for HipSci
                                                                             # IPS predictions
                                                                             # (vs fibro & mononuclear)
        scores_to_use <- scores_to_use_list[[k]][,score_to_use_index]
        
        CDX_stemness_correlations_list[[length(CDX_stemness_correlations_list)+1]] <-
            get_expression_scores_correlations(tCounts_to_use, scores_to_use)
        names(CDX_stemness_correlations_list)[length(CDX_stemness_correlations_list)] <-
            paste0(names(CDX_tRawCounts_list)[i], "; ", names(scores_to_use_list)[k])

        cat(paste0(names(CDX_tRawCounts_list)[i], "; ", names(scores_to_use_list)[k], ":\n"))
        cat(paste0("\nPositive correlations (top 10):\n"))
        print(CDX_stemness_correlations_list[[length(CDX_stemness_correlations_list)]][[1]][1:10])
        cat(paste0("\nNegative correlations (top 10):\n"))
        print(CDX_stemness_correlations_list[[length(CDX_stemness_correlations_list)]][[2]][1:10])
        cat("\n")    
    }
}
```

#### __4b)__: Increased Transcription Correlations

Similarly, we'll now find the genes positively or negatively correlated with increase in transcription in the cell populations from
each sample. We will correlate against total transcription increase as measured by the count of all reads in all genes, for each
cell, and separately against total transcribed genes in each cell (number of genes with >= 1 read).

```{r}
CDX_total_transcription_levels_list <- list()
CDX_total_transcribed_genes_list    <- list()
for(i in 1:length(CDX_tRawCounts_list)){
    df <- CDX_tRawCounts_list[[i]]
    total_transcription_levels <- rowSums(df)
    total_transcribed_genes_vec <- vector()
    for(j in 1:nrow(df)){
        thisCount <- sum(df[j,] >= 1)
        if(is.na(thisCount)){ cat("Is na\n") }; if(thisCount <= 0){ cat("0 or less\n") }
        total_transcribed_genes_vec[[j]] <- thisCount
    }
    names(total_transcribed_genes_vec) <- rownames(df)
    print(cor.test(as.numeric(total_transcribed_genes_vec), as.numeric(total_transcription_levels),
                   method="pearson")) # cor ~0.95
    CDX_total_transcription_levels_list[[i]] <- total_transcription_levels
    CDX_total_transcribed_genes_list[[i]] <- total_transcribed_genes_vec
    names(CDX_total_transcription_levels_list)[i] <- names(CDX_tRawCounts_list)[i]
    names(CDX_total_transcribed_genes_list)[i] <- names(CDX_tRawCounts_list)[i]
    df <- NULL    
}

CDX_transcription_correlations_list <- list()
for(i in 1:length(CDX_tRawCounts_list)){

    tCounts_to_use <- CDX_tRawCounts_list[[i]]
    scores_to_use_list <- list(CDX_total_transcription_levels_list[[i]],
                               CDX_total_transcribed_genes_list[[i]])
    names(scores_to_use_list) <- c("Total transcription", "Total transcribed genes")

    for(j in 1:length(scores_to_use_list)){
        scores_to_use <- scores_to_use_list[[j]]        

        CDX_transcription_correlations_list[[length(CDX_transcription_correlations_list)+1]] <-
            get_expression_scores_correlations(tCounts_to_use, scores_to_use)
        names(CDX_transcription_correlations_list)[length(CDX_transcription_correlations_list)] <-
            paste0(names(CDX_tRawCounts_list)[i], "; ", names(scores_to_use_list)[j])
        
        cat(paste0(names(CDX_tRawCounts_list)[i], "; ", names(scores_to_use_list)[j], ":\n"))
        cat(paste0("\nPositive correlations (top 10):\n"))
        print(CDX_transcription_correlations_list[[length(CDX_transcription_correlations_list)]][[1]][1:10])
        cat(paste0("\nNegative correlations (top 10):\n"))
        print(CDX_transcription_correlations_list[[length(CDX_transcription_correlations_list)]][[2]][1:10])
        cat("\n")
    }
}
```

Info about full HUGO gene names is available within this R package, and will be used to help understand the correlations:

```{r}
head(all_HUGO_abbreviations_and_fullnames)
```

However, HUGO continually updates this info. To get the most recent data for HUGO full names, you can uncomment the following code:

```{r}
use_biomart <- FALSE # Set to TRUE to use Biomart to get the full names, when FALSE downloads from ftp.ebi.ac.uk
if(use_biomart == TRUE){  library("biomaRt")  }

#if(use_biomart == FALSE){
#    HUGO_ftp_address <- 'ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt'
#    download.file(HUGO_ftp_address, "hgnc_complete_set.txt")
#    all_HUGO_info <- read.csv(paste0("hgnc_complete_set.txt"), header=TRUE, sep="\t",
#                              stringsAsFactors = FALSE)
#    all_HUGO_abbreviations_and_fullnames <- all_HUGO_info[,c(2,3)]
#    save(all_HUGO_abbreviations_and_fullnames, file = paste0(saveDir,
#        "all_HUGO_abbreviations_and_fullnames.RData"))
#} else {
#    mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
#}
```

You may have noticed that there are a lot of ribosomal genes in the most positivley correlated genes, both for stemness and transcription.
The gene IDs may also not have a lot of meaning for you, if you are not familiar with them. For these reasons, we'll look up a full
gene name/description for each of them, and remove those from the mostly highly correlated that have "ribosomal" in the full
gene name.

First we'll do that for the genes correlated with stemness increase, in the cancer cell populations. We'll just do it for some of
the stemness correlations (Ranger generated), but you could modify the stemness_correlations_indices_to_use value in the code to look
at the others:

```{r}
CDX_top_10_stemness_positive_correlations_no_ribo <- list()
number_of_correlations <- 100
stemness_correlations_indices_to_use <- c(1,4) # Manually set for
                                               # 1="GSM4104161_SC4_Talazoparib.LB17011; ranger_hipsci_raw"
                                               # 4="GSM4104152_SC4.LB17009; ranger_hipsci_raw"
for(i in stemness_correlations_indices_to_use){
#for(i in 1:length(CDX_stemness_correlations_list)){ # Use this for all
    gene_names_to_get <- names(CDX_stemness_correlations_list[[i]][[1]][1:number_of_correlations])

    if(use_biomart == FALSE){
        match_pos <- match(gene_names_to_get, all_HUGO_abbreviations_and_fullnames[,1])
        gene_names_and_functions <- data.frame(
            hgnc_symbol = gene_names_to_get,
            description = all_HUGO_abbreviations_and_fullnames[match_pos,2],
            stringsAsFactors = FALSE
        )
        gene_names_and_functions[which(is.na(gene_names_and_functions[,2])),2] <- '-'
    } else {
        gene_names_and_functions <- getBM(attributes = c("hgnc_symbol", "description"),
                                          filters = 'hgnc_symbol', gene_names_to_get, mart = mart)
    }

    #print(gene_names_and_functions)
    correlation_results <- CDX_stemness_correlations_list[[i]][[1]][1:number_of_correlations]
    correlations_with_functions_df <- data.frame(
      genes <- names(correlation_results),
      correlations <- correlation_results,
      functions <- gene_names_and_functions[match(names(correlation_results),
                   gene_names_and_functions[,1]),2],
      stringsAsFactors = FALSE
    )
    colnames(correlations_with_functions_df) <- c('Genes', 'Correlation', 'Function')
    rownames(correlations_with_functions_df) <- NULL
    correlations_with_functions_df[which(is.na(correlations_with_functions_df[,3])),3] <- '-'
    correlations_with_functions_df[,3] <- gsub(' \\[Source.*', "",
                                               correlations_with_functions_df[,3])
    to_remove_pos <- grep("ribosomal",  correlations_with_functions_df[,3])
    if(length(to_remove_pos > 0)){
        correlations_with_functions_df_no_ribo <- correlations_with_functions_df[-c(to_remove_pos),]
    } else {
        correlations_with_functions_df_no_ribo <- correlations_with_functions_df
    }
    #cat(paste0('Correlations (top ', number_of_correlations, '):\n'))
    #print(correlations_with_functions_df)
    #cat(paste0('Correlations (ribosomal removed):\n'))
    #print(correlations_with_functions_df_no_ribo)

    top_10_stemness_position <- (length(CDX_top_10_stemness_positive_correlations_no_ribo) + 1)
    CDX_top_10_stemness_positive_correlations_no_ribo[[(top_10_stemness_position)]] <-
        correlations_with_functions_df_no_ribo[1:10,1]
    names(CDX_top_10_stemness_positive_correlations_no_ribo)[top_10_stemness_position] <-
        names(CDX_stemness_correlations_list)[i]
}

cat(paste0("Top 10 positive correlations with ribosomal removed:\n"))
CDX_top_10_stemness_positive_correlations_no_ribo
```

And we'll do the same, for the high transcription associated genes:

```{r}
CDX_top_10_transcription_positive_correlations_no_ribo <- list()
number_of_correlations <- 100
transcription_correlations_indices_to_use <- c(1,3) # Manually set for
                                                    # 1="GSM4104161_SC4_Talazoparib.LB17011; Total transcription"
                                                    # 3="GSM4104152_SC4.LB17009; Total transcription"
for(i in transcription_correlations_indices_to_use){
#for(i in 1:length(CDX_transcription_correlations_list)){ # Use for all
    gene_names_to_get <- names(CDX_transcription_correlations_list[[i]][[1]][1:number_of_correlations])
    if(use_biomart == 0){
        match_pos <- match(gene_names_to_get, all_HUGO_abbreviations_and_fullnames[,1])
        gene_names_and_functions <- data.frame(
            hgnc_symbol = gene_names_to_get,
            description = all_HUGO_abbreviations_and_fullnames[match_pos,2],
            stringsAsFactors = FALSE
        )
        gene_names_and_functions[which(is.na(gene_names_and_functions[,2])),2] <- '-'
    } else {
        gene_names_and_functions <- getBM(attributes = c("hgnc_symbol", "description"),
                                          filters = 'hgnc_symbol', gene_names_to_get, mart = mart)
    }
    #print(gene_names_and_functions)
    correlation_results <- CDX_transcription_correlations_list[[i]][[1]][1:number_of_correlations]
    correlations_with_functions_df <- data.frame(
      genes <- names(correlation_results),
      correlations <- correlation_results,
      functions <- gene_names_and_functions[match(names(correlation_results),
                   gene_names_and_functions[,1]),2],
      stringsAsFactors = FALSE
    )
    colnames(correlations_with_functions_df) <- c('Genes', 'Correlation', 'Function')
    rownames(correlations_with_functions_df) <- NULL
    correlations_with_functions_df[which(is.na(correlations_with_functions_df[,3])),3] <- '-'
    correlations_with_functions_df[,3] <- gsub(' \\[Source.*', "",
                                               correlations_with_functions_df[,3])
    correlations_with_functions_df_no_ribo <-
        correlations_with_functions_df[-c(grep("ribosomal",  correlations_with_functions_df[,3])),]

    #cat(paste0('Correlations (top ', number_of_correlations, '):\n'))
    #print(correlations_with_functions_df)
    #cat(paste0('Correlations (ribosomal removed):\n'))
    #print(correlations_with_functions_df_no_ribo)

    top_10_transcription_position <- (length(CDX_top_10_transcription_positive_correlations_no_ribo) + 1)
    CDX_top_10_transcription_positive_correlations_no_ribo[[(top_10_transcription_position)]] <-
        correlations_with_functions_df_no_ribo[1:10,1]
    names(CDX_top_10_transcription_positive_correlations_no_ribo)[top_10_transcription_position] <-
        names(CDX_transcription_correlations_list)[i]
}

cat(paste0("Top 10 positive correlations with ribosomal removed:\n"))
print(CDX_top_10_transcription_positive_correlations_no_ribo)
```

----------

### Step 5: t-SNE Analysis #############################################################################################

We'll now look at the t-SNEs that were generated in step 2, and explore their topologies. The stemness scores that were
produced in step 3, will be overlaid onto each of the cells represented in the t-SNEs, to look for any stemness gradients that are
apparent. Similarly, we'll alse look at expression patterns of the genes most correlated with stemness and transcription
that we identified in step 4. This is with the aim of identifying a subset of cells that appear to have high transcription,
high stemness, and that are interesting in the context of the shape of the t-SNE.

The two samples that we are analysing are quite homogenous, both in terms of cell type i.e. all are expected to be malignant cancer cells
from a CDX, and that the two cell populations are each from a separate tumour.

If you are following this tutorial on the cscplots web pages at Github, the generated plots that are shown may appear
small, and are harder to study in detail. It's recommended to use the high resolution tif files that are printed by the code in this tutorial.
We'll be looking at the relationships between small numbers of cells in the t-SNEs, so it's important to see them at high resolution.
This is also true for other plots generated in this tutorial - it's much better to look at the tif files.

#### __5a)__: Annotate the t-SNEs

Add the scores to a list of t-SNEs, for later plotting and analysis:

```{r}
# Use the perplexity values you set previously, or a subset of them, if appropriate:
#perplexityValues <<- c(5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130)
#perplexityValues <<- 70

CDX_tSNE_list_with_annotations <- list()
stemness_score_indices_to_use <- c(2,2,2) # manually set chosen scores to add: 2 for HipSci IPS predictions (fibro = 1; mononuclear = 3)
for(i in 1:length(CDX_tSNE_list)){

    tSNE_results_with_annotations_tmp <- CDX_tSNE_list[[i]]

    tSNE_results_with_annotations_tmp <-
        add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp,
                                   perplexityValues,
                                   as.numeric(CDX_total_transcription_levels_list[[i]]),
                                   "Total_transcription")
    tSNE_results_with_annotations_tmp <-
        add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp,
                                   perplexityValues,
                                   as.numeric(CDX_total_transcribed_genes_list[[i]]),
                                   "Total_transcribed_genes")

    for(j in 1:length(CDX_stemness_prediction_DFs_lists)){
        tSNE_results_with_annotations_tmp <-
            add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp,
                                       perplexityValues,
                                       scores <- CDX_stemness_prediction_DFs_lists[[j]][[i]][,stemness_score_indices_to_use[j]],
                                       scoresName <- paste0(names(CDX_stemness_prediction_DFs_lists[j]), "_",
                                       names(CDX_stemness_prediction_DFs_lists[[j]][[i]])[stemness_score_indices_to_use[j]]))
    }

    CDX_tSNE_list_with_annotations[[i]] <- tSNE_results_with_annotations_tmp
    names(CDX_tSNE_list_with_annotations)[i] <- names(CDX_tSNE_list)[i]
    tSNE_results_with_annotations_tmp <- NULL
}
save(CDX_tSNE_list_with_annotations, file = paste0(saveDir, "CDX_tSNE_list_with_annotations.RData"))
#load(paste0(saveDir, "CDX_tSNE_list_with_annotations.RData"))
```

#### __5b)__: Plot t-SNE Stemness and Transcription Gradients
Use the newly annotated tSNE list (CDX_tSNE_list_with_annotations), to plot the t-SNEs overlaid with stemness scores and transcription levels:

```{r}
# Print all the annotated scores previously added i.e. Total_transcription, Total_transcribed_genes,
# ranger_hipsci_raw_IPS, etc.
for(i in 1:length(CDX_tRawCounts_list)){
    #j <- 1 # perplexity values index
    for(j in 1:length(perplexityValues)){
        #outDir <- paste0(tSNEsDir, names(CDX_tSNE_list_with_annotations)[i], "/")
        #dir.create(file.path(outDir), showWarnings = FALSE)
        outDir <- tSNEsDirs[[i]]
        
        for(k in 3:length(names(CDX_tSNE_list_with_annotations[[i]][[j]]))){ # 1:2 are the Y1 & Y2
                                                                             # values so skip them
            plot_scores(CDX_tSNE_list_with_annotations[[i]][[j]], "Y1", "Y2",
                        runName <- tSNE_run_names[j],
                        scoresName <- names(CDX_tSNE_list_with_annotations[[i]][[j]])[k],
                        datasetName <- names(CDX_tSNE_list_with_annotations)[i], outDir,
                        labelPositions <- NULL, colourPositions <- NULL, highestOnTop <- TRUE)
        }
    }
}
```

#### __5c)__: Plot t-SNEs with Expression of High Stemness and Transcription Associated Genes

To continue building up a picture of which areas of the t-SNEs may contain putative cancer stem cells, genes correlated with high stemness
and/or high transcription, will now be displayed.

We will select the genes to print their expression within the t-SNE; here we'll use genes in the top 10 of stemness __and__ transcription
positive correlations, in __either__ sample:

```{r}
gene_names_to_print_1 <- c(CDX_top_10_stemness_positive_correlations_no_ribo[[1]],
                           CDX_top_10_transcription_positive_correlations_no_ribo[[1]])
gene_names_to_print_2 <- c(CDX_top_10_stemness_positive_correlations_no_ribo[[2]],
                           CDX_top_10_transcription_positive_correlations_no_ribo[[2]])
gene_names_to_print <- unique(c(gene_names_to_print_1[which(gene_names_to_print_1 %in%
                                                            gene_names_to_print_2)],
                                gene_names_to_print_2[which(gene_names_to_print_2 %in%
                                                            gene_names_to_print_1)]))

perpl_index <- 1 # if multiple perplexity values were used for t-SNEs, this number defines the
                 # index in the list of t-SNEs to use
run_name <- tSNE_run_names[perpl_index]
```

Then we print the t-SNEs showing expression of the chosen genes, for manual inspection:

```{r}
for(cohort_index in 1:2){
    print_tSNE_coloured_by_gene(gene_names_to_print,
                                CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                                CDX_tRawCounts_list[[cohort_index]],
                                run_name, outDir <- tSNEsDirs[[cohort_index]], point_size <- 2,
                                all_HUGO_abbreviations_and_fullnames, dataDir)
}
```

We can also introduce biological and/or clinical information at this stage, to further assist identification of the cancer stem cells. SCLC is thought to
have genes associated with tumour subtypes: these are ASCL1, NEUROD1, YAP1 and POU2F3. Looking at the expression of these genes in the t-SNEs indicates
that ASCL1 is influential in the samples that we are studying:

```{r}
gene_names_to_print <- 'ASCL1'
for(cohort_index in 1:2){
    print_tSNE_coloured_by_gene(gene_names_to_print,
                                CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                                CDX_tRawCounts_list[[cohort_index]],
                                run_name, tSNEsDirs[[cohort_index]], point_size <- 2,
                                all_HUGO_abbreviations_and_fullnames, dataDir)
}
```

N.B. In addition, comparison of PARP1 expression between the Talazoparib PARP inhibitor treated and untreated samples i.e. GSM4104161_SC4_Talazoparib.LB17011 vs
GSM4104152_SC4.LB17009, shows that PARP1 expression has a higher maximum after treatment (in GSM4104161_SC4_Talazoparib.LB17011). This corresponds to the
area of the t-SNE that has cells with high transcription and predicted stemness. This may be of clinical interest, because post-therapy, the drug's target is
elevated in a subset of cells that may be driving the tumour.

```{r}
gene_names_to_print <- 'PARP1'
for(cohort_index in 1:2){
    print_tSNE_coloured_by_gene(gene_names_to_print,
                                CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                                CDX_tRawCounts_list[[cohort_index]],
                                run_name, tSNEsDirs[[cohort_index]], point_size <- 2,
                                all_HUGO_abbreviations_and_fullnames, dataDir)
}
```

### Step 6: pCSC Group Selection #######################################################################################

Step 5 compared the topologies of the t-SNEs with elevated predicted stemness and very high transcription. This indicated that there
are regions of the t-SNEs where the cells appear to have increased stemness and aberrantly high transcription. These cells are good candidates
for further study as putative cancer stem cells. We will select a group of cells from these areas, which can then be used to
create some preferential expression plots.

To identify a group of cells to examine is may be arbitrary to some extent, as we are imposing a cutoff on the gradients of transcription and stemness,
However, we'll look for clusters of cells, that are located in the region of interest.

Generally I try to select somewhere in the region of between 2 and 30 cells, but it depends on the topology of the t-SNE and characteristics of the cell population.

```{r}
cohort_index <- 1
perpl_index <- 1

outDir <- pCSCsDirs[[cohort_index]]
run_name <- tSNE_run_names[perpl_index]

plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, scoresName <- "Total_transcription",
                      datasetName <- names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions <- NULL, colourPositions <- NULL, highestOnTop <- TRUE)
```

In the plot that was just generated, on the Y2 axis, in my plot (yours may be slightly different) cells located above Y2 = 39 appear interesting. In the plots
that we previously created for this sample, this area of the t-SNE has high total transcribed genes, high total transcription, high predicted
stemness using HipSci (ranger, rf and randomForest predictions). The genes highly correlated with stemness and transcriptiion are also highly
expressed in this region, e.g. YBX1, CHCHD2, H3F3A, PPIA, HINT1 & RAN. You could select cells slightly above or below this cutoff, and see how the
preferential results generated in steps 7 & 8 change, but they likely will be very similar.

To assist pCSC group selection, we will use t-SNEs with total transcription displayed. We'll identify cells of interest by colouring them black or labelling them,
and then noting their positions (row numbers) in the melanoma expression matrix. We'll use plots with grids in the background, to enable this. Generated tif files
will are written to the relevant sample's results directory under "tSNEs/pCSC_group_selection/", where you can view them.

```{r}
found_pos1 <- which(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Y2 > 39)
#found_pos1 <- c(165, 889, 893, 1127, 1186, 1277, 1303, 3448, 4175,
#                4203, 4609, 4874, 5295, 5297, 5425, 5429) # In my plot these are the cells noted in found_pos1

# reorder by total transcription
found_pos1_ordered <-
    found_pos1[rev(order(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos1,3]))]
```

To check the selection, plot with highlighted or labelled selected cells - output is in a file ending in "with_grid-selected_cells.tif" or "with_grid-annotated.tif":

```{r}
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir, NULL,
                      colourPositions <- found_pos1_ordered,
                      highestOnTop <- TRUE) # selected cells in black

plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions <- found_pos1_ordered, NULL,
                      highestOnTop <- TRUE) # selected cell names labelled
```

These are the cell IDs for this pCSC group, and we'll store their expression matrix positions for later use:

```{r}
CDX_pCSC_group_1_SC4_Talazoparib.LB17011 <- found_pos1_ordered

CDX_pCSC_group_1_SC4_Talazoparib.LB17011_df <-
    CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos1_ordered,]
CDX_pCSC_group_1_SC4_Talazoparib.LB17011_cell_names <-
    rownames(CDX_pCSC_group_1_SC4_Talazoparib.LB17011_df)
cat(CDX_pCSC_group_1_SC4_Talazoparib.LB17011_cell_names, sep = '\n')
```

We will now select some pCSCs from the 2nd sample, in the same way.

```{r}
cohort_index <- 2
perpl_index <- 1

outDir <- pCSCsDirs[[cohort_index]]
run_name <- tSNE_run_names[perpl_index]

plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir, NULL, NULL,
                      highestOnTop <- TRUE)
```

In my version of the t-SNE for this sample, I selected cells below -14 on the Y2 axis due to their exceptionally high transcription levels, and because these cells
were predicted to have high stemness, including high expression of the stemness markers that we found using the stemness correlations, as shown in previous plots.

```{r}
found_pos2 <- which(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Y2 < -40)
#found_pos2 <- c(186, 266, 560, 1213, 2108, 2490, 3474, 3604, 4071, 4141, 5038, 5064, 5364) # these are the cells I selected

# reorder by total transcription
found_pos2_ordered <-
    found_pos2[rev(order(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos2,3]))]
```

Highlighted cells in black:

```{r}
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir, NULL,
                      colourPositions <- found_pos2_ordered, highestOnTop <- TRUE)
```

And with cell labels:

```{r}
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions <- found_pos2_ordered, NULL, highestOnTop <- TRUE)
```

Store the pCSC group for this sample for later use, and show the cell names:

```{r}
CDX_pCSC_group_2_SC4.LB17009 <- found_pos2_ordered

CDX_pCSC_group_2_SC4.LB17009_df <-
    CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos2_ordered,]
CDX_pCSC_group_2_SC4.LB17009_cell_names <- rownames(CDX_pCSC_group_2_SC4.LB17009_df)
cat(CDX_pCSC_group_2_SC4.LB17009_cell_names, sep = '\n')
```

We could have done something similar, automatically, by selecting the top n genes using total transcription as follows. However, I prefer not to do this, as it's
not taking into consideration the topology of the t-SNE or the stemness scores. It's better to look at all of this information in combination.

```{r}
n <- 10 # take the top 10 genes automatically

cohort_index <- 1
perpl_index <- 1

outDir <- pCSCsDirs[[cohort_index]]
run_name <- tSNE_run_names[perpl_index]

found_pos3 <- rev(order(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Total_transcription))[1:n]
CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos3,3]
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir, NULL,
                      colourPositions <- found_pos3, highestOnTop <- TRUE) # selected cells in black
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions <- found_pos3, NULL, highestOnTop <- TRUE) # labelled cells
```

And now, automatic selection for the second sample:

```{r}
n <- 10

cohort_index <- 2
perpl_index <- 1

outDir <- pCSCsDirs[[cohort_index]]
run_name <- tSNE_run_names[perpl_index]

found_pos4 <- rev(order(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Total_transcription))[1:n]
CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]][found_pos4,3]
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir, NULL,
                      colourPositions <- found_pos4, highestOnTop <- TRUE) # black cells
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions <- found_pos4, NULL, highestOnTop <- TRUE) # cell labels
```

... but looking at the t-SNE, I would refine it to the top 8, as this is more consistent with the tip of the t-SNE shape.

### Step 7: Preferential Expression Plots ForSelected pCSC Groups ############################################

Now that we have some groups of pCSCs selected for both samples, we'll look at the preferential expression in these cells compared to all
of the cells in the cell populations they were selected from.

Some info needs to be prepared for preferential expression plot of HUGO gene groups in the pCSCs (e.g. families of related genes, or components
of a molecular complex). 

```{r}
head(all_HUGO_groups_mordered) # automatically available in this package -
                               # you can get the most recent info from the HUGO website

cohort_index <- 1
CDX_transcript_names <- colnames(CDX_tRawCounts_list[[cohort_index]])

# Create a list of HUGO gene groups, with each group containing the position of each group member in the
# expression matrix columns 
CDX_all_HUGO_groups_positions <-
    map_gene_positions_to_HUGO_gene_groups(CDX_transcript_names, all_HUGO_groups_mordered)
save(CDX_all_HUGO_groups_positions, file = paste0(saveDir, "CDX_all_HUGO_group_positions.RData"))
#load(paste0(saveDir, "CDX_all_HUGO_group_positions.RData"))

# Subset to use HUGO gene groups that have two or more members in the gene expression matrix
CDX_all_HUGO_groups_positions_two_or_more <-
    CDX_all_HUGO_groups_positions[which(lengths(CDX_all_HUGO_groups_positions) >= 2)]
gene_group_positions_to_use <- CDX_all_HUGO_groups_positions_two_or_more
```

To run the preferential expression plots, set some parameters:

```{r}
CDX_pCSC_groups_to_run <- list()

CDX_pCSC_groups_info_list_tmp <- list()
CDX_pCSC_groups_info_list_tmp[[1]] <- 'pCSC_group_1_SC4_Talazoparib.LB17011'   # pCSC group name
CDX_pCSC_groups_info_list_tmp[[2]] <- CDX_pCSC_group_1_SC4_Talazoparib.LB17011 # pCSC group positions
CDX_pCSC_groups_info_list_tmp[[3]] <- 1                      # matrix position in CDX_tRawCounts_list
names(CDX_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions',
                                          'Expression matrix list index')
CDX_pCSC_groups_to_run[[1]] <- CDX_pCSC_groups_info_list_tmp

CDX_pCSC_groups_info_list_tmp <- list()
CDX_pCSC_groups_info_list_tmp[[1]] <- 'pCSC_group_2_SC4.LB17009'          # pCSC group name
CDX_pCSC_groups_info_list_tmp[[2]] <- CDX_pCSC_group_2_SC4.LB17009        # pCSC group positions
CDX_pCSC_groups_info_list_tmp[[3]] <- 2                 # matrix position in CDX_tRawCounts_list
names(CDX_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions',
                                          'Expression matrix list index')
CDX_pCSC_groups_to_run[[2]] <- CDX_pCSC_groups_info_list_tmp

tissue_type       <<- "CDX"
frag_count_method <- 'raw'  # expression measurement type, but just a name, can be anything

no_ribo                                      <- TRUE  # set to TRUE to remove ribosomal transcripts
                                                      # from some of the expression bar charts
separate_row_for_high_expression_transcripts <- FALSE # set to TRUE to print highest expression bars
                                                      # on a seperate row
number_of_graph_rows_stacked_transcripts     <- 1
number_of_graph_rows_stacked_families        <- 1
```

Now that preparations are complete, we can __print the preferential expression plots__.

(If the images generated look small when viewing them in the web pages, you could try opening each of them in a new tab
e.g. by using a right mouse click.)

```{r fig.width=30}
CDX_all_results_ranks_list <- list() # preferential expression ranks will be stored in here

for(r in 1:length(CDX_pCSC_groups_to_run)){
    
    expression_matrix    <- NULL
    pCSC_group_pos       <- NULL
    
    this_matrix_index    <- CDX_pCSC_groups_to_run[[r]][[3]]
    expression_matrix    <- CDX_tRawCounts_list[[this_matrix_index]]
    this_CDX_name        <- names(CDX_tRawCounts_list[this_matrix_index])
    this_pCSC_group_name <- CDX_pCSC_groups_to_run[[r]][[1]]
    cat(paste0(this_pCSC_group_name, "; ", this_CDX_name, "\n"))

    CDX_tumour_only_positions <- 1:nrow(expression_matrix) # can be restricted to a subset,
                                                           # not necessary for CDXs

    # Subset transcription info for pCSC group
    pCSC_group_pos <- CDX_pCSC_groups_to_run[[r]][[2]]

    # Generate graphs
    CDX_all_results_ranks_list[[r]] <- print_preferential_expression_bar_graphs(
        pCSC_group_name                    <- this_pCSC_group_name,
        pCSC_group_positions               <- pCSC_group_pos,
        HUGO_groups_sets_of_columnNumbers  <- gene_group_positions_to_use,
        tExprMatr                          <- expression_matrix,
        tumour_only_positions              <- CDX_tumour_only_positions,
        this_patient_tumour_only_positions <- CDX_tumour_only_positions,
        outDir                             <- prefExprDirs[[(CDX_pCSC_groups_to_run[[r]][[3]])]],
        tissue_type,
        family_type_name                   <- 'HUGO',
        patient_source_name                <- this_CDX_name,
        frag_count_method,
        no_ribo,
        separate_row_for_high_expression_transcripts,
        number_of_graph_rows_stacked_transcripts,
        number_of_graph_rows_stacked_families,
        all_HUGO_abbreviations_and_fullnames,
        dataDir
    )
    names(CDX_all_results_ranks_list)[[r]] <- this_pCSC_group_name #names(CDX_pCSC_groups)[r]
}
save(CDX_all_results_ranks_list, file = paste0(saveDir, "CDX_all_results_ranks_list_raw.RData"))
#load(paste0(saveDir, "CDX_all_results_ranks_list_raw.RData"))
```

We need to print the t-SNEs with the cell numbers of the pCSC group used in the stacked transcripts graphs. (The tif file is written to the
results/tSNEs directory under the relevant samples directory, and ends in "n"_cells-labelled.tif, where n is the number of cells.) This can be
used to compare these preferential expression graphs with the t-SNEs, to understand the pCSCs in great detail.

```{r}
perpl_index <- 1

for(i in 1:length(CDX_pCSC_groups_to_run)){
    exprMatr_list_index  <- CDX_pCSC_groups_to_run[[i]][[3]]
    this_pCSC_group_name <- CDX_pCSC_groups_to_run[[i]][[1]]
    this_pCSC_group_pos  <- CDX_pCSC_groups_to_run[[i]][[2]]
    this_perplexity_name <- names(CDX_tSNE_list_with_annotations[[(exprMatr_list_index)]])[perpl_index]
    this_sample_name     <- names(CDX_tSNE_list_with_annotations)[[(exprMatr_list_index)]]
    this_tSNE_results_with_annotations <-
        CDX_tSNE_list_with_annotations[[(exprMatr_list_index)]][[perpl_index]]
    rownames(this_tSNE_results_with_annotations)[(this_pCSC_group_pos)] <-
        paste0("#", as.character(1:length(this_pCSC_group_pos)))
    plot_scores(this_tSNE_results_with_annotations, "Y1", "Y2", this_pCSC_group_name,
                "Total_transcription", names(CDX_tSNE_list_with_annotations)[exprMatr_list_index],
                tSNEsDirs[[exprMatr_list_index]], this_pCSC_group_pos)
    
    outFile <- paste0(tSNEsDirs[[i]], this_pCSC_group_name, '-Total_transcription_Y1vsY2.tif') 
    renamed_file <- paste0(tSNEsDirs[[i]], this_pCSC_group_name, "-Total_transcription_Y1vsY2-",
                           length(this_pCSC_group_pos), "_cells-labelled.tif")
    file.rename(outFile, renamed_file)
}
```

These are the __top ranked preferential expressed genes and gene groups__ from the preferential expression plots, for each sample.

```{r}
for(i in 1:length(CDX_all_results_ranks_list)){
    cat(paste0("\n## Rankings for ", names(CDX_all_results_ranks_list[i]), " ##\n\n"))
    cat(paste0("Top 100 transcripts for ", names(CDX_all_results_ranks_list[i]), ":\n"))
    print(names(CDX_all_results_ranks_list[[i]][[1]]))
    cat(paste0("\nTop 100 families for ", names(CDX_all_results_ranks_list[i]), ":\n"))
    print(names(CDX_all_results_ranks_list[[i]][[2]]))
    cat(paste0("\nTop 100 transcripts in ranked families for ", names(CDX_all_results_ranks_list[i]), ":\n"))
    print(names(CDX_all_results_ranks_list[[i]][[3]][1:100]))
}
```

### Step 8: t-SNEs for Highest Ranked Genes ############################################

To understand further the importance of the the highest ranked preferentially expressed genes, we'll  view their expression patterns within the
t-SNEs.

For example, let's look at AP-1 transcription factor components FOS, JUN & JUNB, as they ranked highly.
```{r}
gene_names_to_print <- c('FOS', 'JUN', 'JUNB')

perpl_index <- 1
run_name <- tSNE_run_names[perpl_index]

for(cohort_index in 1:2){
    print_tSNE_coloured_by_gene(gene_names_to_print,
                                CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                                CDX_tRawCounts_list[[cohort_index]],
                                run_name, outDir <- tSNEsDirs[[cohort_index]], point_size <- 2,
                                all_HUGO_abbreviations_and_fullnames, dataDir)
}
```

This is the end of cscplots tutorial 1.

__Now try [tutorial 2](tutorial_2.html)__, which looks at CSCs from melanoma and evaluates their preferential expression. This is a more
complex example from the perspective of having one cell population composed of multiple cell types from multiple samples.