---
title: "cscplots Tutorial 2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cscplots Tutorial 2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### __Tutorial 2: Identifying and Characterising Cancer Stem Cells In Melanoma Single-Cell RNA-Seq__

Tutorial 2 uses a melanoma scRNA-Seq dataset published by Tirosh et al. (Science 2016).

A pipeline will be constructed using functions from the cscplots package. This will follow these steps:

- generate a t-SNE; predict stemness in each of the cells, using HipSci iPS expression profiles (RNA-Seq); identify genes
  associated with higher stemness and very high transcription, in malignant melanoma cells; 
- use the t-SNE to select putative cancer stem cells, by using the stemness predictions and other information such as
  transcription gradients and predicted cell types;
- detect preferential expression in the selected putative cancer stem cells using a TPM Count Fractions matrix, to
  find genes driving these cells.


### Step 1: t-SNE Generation ###########################################################################################

Create the results directories, and load the melanoma expression data included with this package.

```{r message=FALSE}
library("Matrix")
library("biomaRt")
library("gtools")
library("Rtsne")
library("caret")
library("ranger")
library("e1071")
library("randomForest")
library("ggplot2")
library("ggrepel"); options(ggrepel.max.overlaps = Inf)
library("cscplots")

project_name   <- "mela_4558_TPM" 
wd             <- './' # set the working directory
saveDir        <- paste0(wd, 'save/')
resultsBaseDir <- paste0('results/')
resultsDir     <- paste0('results/', project_name, '/')
tSNEsDir       <- paste0(resultsDir, 'tSNEs/')
pCSCsDir       <- paste0(tSNEsDir, "pCSC_group_selection/")
prefExprDir    <- paste0(resultsDir, "preferential_expression/")
saveDir        <- paste0(wd, 'save/')
dataDir        <- paste0(wd, '../data/')
dir.create(saveDir, showWarnings = FALSE)
dir.create(resultsBaseDir, showWarnings = FALSE)
dir.create(resultsDir, showWarnings = FALSE)
dir.create(tSNEsDir, showWarnings = FALSE)
dir.create(pCSCsDir, showWarnings = FALSE)
dir.create(prefExprDir, showWarnings = FALSE)
dir.create(dataDir, showWarnings = FALSE)
```

Generate a t-SNE using the melanoma scRNA-Seq expression matrix. The recommended perplexity parameter value to use for the t-SNE is 70, but you can 
use a list of perplexity values to generate multiple t-SNEs, if you prefer.

```{r}
#perplexityValues     <<- c(5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130) # using
                                                                 # this range is helpful, but takes longer
perplexityValues      <<- c(70)
perplexityValuesPrint <- add_zeros_to_perplexity_values(perplexityValues)
tSNE_run_names        <- paste0("tSNE_", perplexityValuesPrint)

outDir <- tSNEsDir

set.seed(50) # for reproduction of results in the literature, but comment this out if you would like
             # unfixed results

# Data available automatically in this package is also available from
# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72056 
tTPM_mela_4558 <- rbind(tTPM_mela_1_2500, tTPM_mela_2501_4558)
tTPM_mela_1_2500 <- NULL
tTPM_mela_2501_4558 <- NULL
# tTPM_mela_4558[1:10,1:10]

mela_4558_TPM_tSNE_results <- run_all_tSNEs(perplexityValues, tTPM_mela_4558, project_name, outDir,
                                            montage <- 0) 
save(mela_4558_TPM_tSNE_results, file = paste0(saveDir, "mela_4558_TPM_tSNE_results.RData"))
#load(paste0(saveDir, "mela_4558_TPM_tSNE_results.RData"))

# Add to results to a list, which allows you to add further t-SNEs later (e.g. using subsets of the cell
# population) - this is necessary for use with some of the cscplots functions used below; a list is
# expected:
mela_tSNE_list <- list()
mela_tSNE_list[[1]] <- mela_4558_TPM_tSNE_results
names(mela_tSNE_list)[1] <- project_name
```

----------

### Step 2: Stemness Predictions #######################################################################################

In the same way as demonstrated in tutorial 1, each cell in the melanoma population of cells will be scored using its
probability of being a stem cell.

The stemness probabilities are generated using random forests, that are trained on stem cell (iPSC) RNA-Seq expression
data from the HipSci consortium. The likelihood that each of the cancer cells are stem cells, is relative to fibroblasts and
mononuclear cells. 

These stemness scores will be used later to assist the identification putative cancer stem cells.

We'll use three random forest implementations (ranger, rf, randomForest) separately, so that we can compare and contrast the results
later, e.g. comparing them within a t-SNE can give greater confidence when selecting pCSCs.

```{r}
number_of_genes_to_use <- 16000
#number_of_genes_to_use <- 10000 # can give reasonable results
#number_of_genes_to_use <- 16700 # fails on my R installation

# Put the melanoma expression matrix in a list first - this is necessary for use with some of the functions:
mela_tTPM_list <- list()
mela_tTPM_list[[1]] <- tTPM_mela_4558
names(mela_tTPM_list)[1] <- "mela_4558_TPM"

mela_stemness_prediction_DFs_lists <- list()
mela_stemness_prediction_DFs_lists[[1]] <- run_ML_stemness_predictors(model_type <- 'ranger',
                                                                      mela_tTPM_list,
                                                                      hipsci_training_data,
                                                                      stemness_type <- 'hipsci',
                                                                      number_of_genes_to_use,
                                                                      sample_data_type <- 'TPM',
                                                                      resultsDir)
names(mela_stemness_prediction_DFs_lists)[1] <- "ranger_hipsci_TPM"
mela_stemness_prediction_DFs_lists[[2]] <- run_ML_stemness_predictors(model_type <- 'rf',
                                                                      mela_tTPM_list,
                                                                      hipsci_training_data,
                                                                      stemness_type <- 'hipsci',
                                                                      number_of_genes_to_use,
                                                                      sample_data_type <- 'TPM',
                                                                      resultsDir)
names(mela_stemness_prediction_DFs_lists)[2] <- "rf_hipsci_TPM"
mela_stemness_prediction_DFs_lists[[3]] <- run_ML_stemness_predictors(model_type <- 'randomForest',
                                                                      mela_tTPM_list,
                                                                      hipsci_training_data,
                                                                      stemness_type <- 'hipsci',
                                                                      number_of_genes_to_use,
                                                                      sample_data_type <- 'TPM',
                                                                      resultsDir)
names(mela_stemness_prediction_DFs_lists)[3] <- "randomForest_hipsci_TPM"

save(mela_stemness_prediction_DFs_lists, file = paste0(saveDir, "mela_stemness_prediction_DFs_lists.RData"))
#load(paste0(saveDir, "mela_stemness_prediction_DFs_lists.RData"))
```

----------

### Step 3: Calculate Total Transcription & Total Transcribed Genes

Total transcription (all expression in all transcripts) for each cell, and total transcribed genes (sum of all expression in all transcripts) for each cell,
can be useful in evaluating the t-SNE for cells that are highly transcriptionally active. We'll collect that info now:

```{r}
mela_total_transcription_levels_list <- list()
mela_total_transcribed_genes_list    <- list()
cohort_index <- 1 # only 1 cohort being used in this project i.e. all 4558 melanoma
                  # population cells from multiple patients
df <- tTPM_mela_4558
total_transcription_levels <- rowSums(df)
total_transcribed_genes_vec <- vector()
for(i in 1:nrow(df)){
    thisCount <- sum(df[i,] >= 1)
    if(is.na(thisCount)){ cat("Is na\n") }; if(thisCount <= 0){ cat("0 or less\n") }
    total_transcribed_genes_vec[[i]] <- thisCount
}
names(total_transcribed_genes_vec) <- rownames(df)
#print(cor.test(as.numeric(total_transcribed_genes_vec), as.numeric(total_transcription_levels),
#               method="pearson")) # cor 0.9739835 
mela_total_transcription_levels_list[[cohort_index]] <- total_transcription_levels
mela_total_transcribed_genes_list[[cohort_index]] <- total_transcribed_genes_vec
names(mela_total_transcription_levels_list)[cohort_index] <- project_name
names(mela_total_transcribed_genes_list)[cohort_index] <- project_name
df <- NULL
```

----------

### Step 4: t-SNE Analysis #############################################################################################

We'll now explore the t-SNEs, with the assistance of the the stemness scores generated in step 2. Our aim is to select subsets
of cancer cells for each patient for further analysis as pCSCs; we'll look for cells that have high transcription and high
predicted stemness. This will also consider the t-SNE topology, for example if there are clusters of cells, with these
features.

It's recommended to view plots in this tutorial using the high resolution tif files that are generated, which will help you
to work precisely with small numbers of cells.

#### __4a)__: Annotate the t-SNEs

Add the cell scores that we have already, to a t-SNE list for subsequent plotting and analysis:

```{r}
mela_tSNE_list_with_annotations <- list()
stemness_score_indices_to_use <- c(2,2,2) # use 2 for HipSci IPS predictions (1 for fibro & 3 for mononuclear)

cohort_index <- 1 # only 1 cohort used in this project
tSNE_results_with_annotations_tmp <- mela_tSNE_list[[cohort_index]]

tSNE_results_with_annotations_tmp <-
    add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp, perplexityValues,
                               scores <- as.numeric(mela_total_transcription_levels_list[[cohort_index]]),
                               scoresName <- "Total_transcription")
tSNE_results_with_annotations_tmp <-
    add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp, perplexityValues,
                               as.numeric(mela_total_transcribed_genes_list[[cohort_index]]),
                               "Total_transcribed_genes")

for(i in 1:length(mela_stemness_prediction_DFs_lists)){
    tSNE_results_with_annotations_tmp <-
        add_scores_to_tSNE_results(tSNE_results_with_annotations_tmp, perplexityValues,
                                   mela_stemness_prediction_DFs_lists[[i]][[cohort_index]][,stemness_score_indices_to_use[i]],
                                   paste0(names(mela_stemness_prediction_DFs_lists[i]), "_",
                                   names(mela_stemness_prediction_DFs_lists[[i]][[cohort_index]])[stemness_score_indices_to_use[i]]))
}

mela_tSNE_list_with_annotations[[cohort_index]] <- tSNE_results_with_annotations_tmp
names(mela_tSNE_list_with_annotations)[cohort_index] <- names(mela_tSNE_list)[cohort_index]
tSNE_results_with_annotations_tmp <- NULL

#save(mela_tSNE_list_with_annotations, file = paste0(saveDir, "mela_tSNE_list_with_annotations.RData"))
```

#### __4b)__: Plot t-SNE Stemness and Transcription Gradients

We'll use the newly annotated t-SNE list, to plot the t-SNEs, overlaid with stemness scores, and transcription levels:

```{r}
cohort_index <- 1
outDir <- tSNEsDir

for(i in 1:length(perplexityValues)){
    for(j in 3:length(names(mela_tSNE_list_with_annotations[[cohort_index]][[i]]))){ # 1:2 are the Y1 & Y2
                                                                                     # values so skip them
        plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[i]], "Y1", "Y2",
                    runName <- tSNE_run_names[i],
                    scoresName <- names(mela_tSNE_list_with_annotations[[cohort_index]][[i]])[j],
                    datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index],
                    outDir, labelPositions <- NULL, colourPositions <- NULL, highestOnTop <- TRUE)
    }
}
```

#### __4c)__: Cell Type Plots and Additional Info

To further understand the composition of the t-SNE, the cell types present will be visualised. This will use cell annotations provided
by the original publication. We'll also use results from ESTIMATE to show immune scores for each cell, and also stromal scores. If you
have a preferred method of cell type prediction e.g. scPred, CIBERSORTx cell type deconvolutions, etc. please feel free to generate and
include these results.

Visualising different types of information about each of the cells helps to build up an understanding of the cell population represented in the t-SNE.
This may in turn help the selection of putative cancer stem cells for preferential expression analysis.

For example, clinically relevant annotations from the original publication can be visualised in the t-SNE e.g. sex, age, survival, drug
therapies received, etc. 

Another stemness score, generated using PluriTest (Nature Methods, 2011), is also included here - please contact the original authors for
this software, if you would like to use it on your own scRNA-Seq data. 

We'll now add this information to the annotated t-SNE list we're working with, so that it can be displayed in the t-SNE.

```{r}
#str(mela_4558_additional_info)

cohort_index <- 1

mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Combined_Pluritest_Score,
                               "Combined_PluriTest_Score")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Samples,
                               "Samples")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Cell_malignancy,
                               "Cell_malignancy")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Cell_type,
                               "Cell_type")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$ESTIMATE_stromal_scores,
                               "ESTIMATE_stromal_scores")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$ESTIMATE_immune_scores,
                               "ESTIMATE_immune_scores")
                                                                 
# For fun, if you feel like it:                                  
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Sex,
                               "Sex")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Age,
                               "Age")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Survival,
                               "Survival")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Mutation_status,
                               "Mutation_status")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$PD1_status,
                               "PD1_status")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$PDL1_status,
                               "PDL1_status")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Preop_drug_types,
                               "Preop_drug_types")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Preop_total_drugs,
                               "Preop_total_drugs")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Postop_drugs,
                               "Postop_drugs")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Sites,
                               "Sites")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Sites_concise,
                               "Sites_concise")
mela_tSNE_list_with_annotations[[cohort_index]] <-
    add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                               perplexityValues,
                               mela_4558_additional_info$Sites_summary,
                               "Sites_summary")

save(mela_tSNE_list_with_annotations, file = paste0(saveDir, "mela_tSNE_list_with_annotations.RData"))
#load(paste0(saveDir, "mela_tSNE_list_with_annotations.RData"))
```

Now print the t-SNEs with cell scores, etc. overlaid (all tifs are written to results/mela_4558_TPM/tSNEs/):

```{r}
cohort_index <- 1 # only 1 cohort used in this project
perpl_index <- 1 # if only 1 perplexity value was used, i.e. perplexity = 70, set to 1
outDir <- tSNEsDir
run_name <- tSNE_run_names[perpl_index]

plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "ESTIMATE_stromal_scores", names(mela_tSNE_list_with_annotations)[cohort_index], outDir,
            NULL, NULL, highestOnTop <- TRUE)
plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "ESTIMATE_immune_scores", names(mela_tSNE_list_with_annotations)[cohort_index], outDir,
            NULL, NULL, highestOnTop <- TRUE)

plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score",
            names(mela_tSNE_list_with_annotations)[cohort_index], outDir, NULL, NULL,
            highestOnTop <- TRUE)
```

To see the PluriTest range more easily, reset the highest PluriTest score, as it's an outlier, to the 2nd highest value:

```{r}
range(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score)
which(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score > 4.53)
mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score[4180]
mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score[4180] <- 0
range(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score)
mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]$Combined_PluriTest_Score[4180] <- 2.72
plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score", names(mela_tSNE_list_with_annotations)[cohort_index], outDir,
            NULL, 4180, highestOnTop <- TRUE) # Cell 4180 (highest PluriTest) in black
```

And all the other cell annotations also (uncomment the additional catergories below if you would like to see them also):

```{r}
plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                categoryName <- "Samples",
                datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                "Cell_malignancy", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                "Cell_type", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)

# For fun:
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Sex", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Age", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Survival", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Mutation_status", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "PD1_status", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "PDL1_status", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)

#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Preop_drug_types", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Preop_total_drugs", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Postop_drugs", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)

#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Sites", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Sites_concise", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
#plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
#                "Sites_summary", names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
```

You could also use the random forest predictors we used for the stemness scores, to predict cell types in the melanoma cohort.
To achive this, we can train them using an scRNA-Seq expression matrix that has cell type annotations. For example,
we can use an NSCLC scRNA-Seq dataset published by Lambrechts *et al.* (Nature Medicine 2018) that's available in this
package and used in tutorial 3.

__This will take time to run__ though (hours); you can skip this step if you are not interested.

Here, we use 10,000 NSLC cells to predict cell types in the melanoma, using the random forest implementations we used previously:
```{r}
number_of_cells_to_use_for_cell_type <- 10000

tCounts_NSCLC_10000_with_cell_types <-
    cbind.data.frame(NSCLC_45000_additional_info$Cell_type[1:number_of_cells_to_use_for_cell_type],
                     tCounts_NSCLC_1_15000[1:number_of_cells_to_use_for_cell_type,])
colnames(tCounts_NSCLC_10000_with_cell_types) <-
    c("Cell_type", colnames(tCounts_NSCLC_10000_with_cell_types)[2:ncol(tCounts_NSCLC_10000_with_cell_types)])
```

Run the cell type predictions:

```{r results=FALSE}
number_of_genes_to_use <- 16000

mela_cell_type_prediction_DFs_lists <- list()
mela_cell_type_prediction_DFs_lists[[1]] <-
    run_ML_cellType_predictors(model_type <- 'ranger',
                               mela_tTPM_list,
                               cellTypeData <- tCounts_NSCLC_10000_with_cell_types,
                               expressionData_type <- 'NSCLC_cell_types',
                               number_of_genes_to_use,
                               sample_data_type <- 'TPM',
                               resultsDir)
names(mela_cell_type_prediction_DFs_lists)[1] <- "ranger_NSCLC_cell_types"

# To use more random forest methods:
#mela_cell_type_prediction_DFs_lists[[2]] <-
#    run_ML_cellType_predictors(model_type <- 'rf',
#                               mela_tTPM_list,
#                               cellTypeData <- tCounts_NSCLC_10000_with_cell_types,
#                               expressionData_type <- 'NSCLC_cell_types',
#                               number_of_genes_to_use,
#                               sample_data_type <- 'TPM',
#                               resultsDir)
#names(mela_cell_type_prediction_DFs_lists)[2] <- "rf_NSCLC_cell_types"
#mela_cell_type_prediction_DFs_lists[[3]] <-
#    run_ML_cellType_predictors(model_type <- 'randomForest',
#                               mela_tTPM_list,
#                               cellTypeData <- tCounts_NSCLC_10000_with_cell_types,
#                               expressionData_type <- 'NSCLC_cell_types',
#                               number_of_genes_to_use,
#                               sample_data_type <- 'TPM',
#                               resultsDir)
#names(mela_cell_type_prediction_DFs_lists)[3] <- "randomForest_NSCLC_cell_types"

save(mela_cell_type_prediction_DFs_lists, file = paste0(saveDir, "mela_cell_type_prediction_DFs_lists.RData"))
#load(paste0(saveDir, "mela_cell_type_prediction_DFs_lists.RData"))
```

Add to the t-SNE list and plot them:

```{r}
cohort_index <- 1
outDir <- tSNEsDir

for(i in 1:length(mela_cell_type_prediction_DFs_lists)){    
    for(j in 1:ncol(mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]])){ # do for all cell types
        mela_tSNE_list_with_annotations[[cohort_index]] <-
            add_scores_to_tSNE_results(mela_tSNE_list_with_annotations[[cohort_index]],
                perplexityValues,
                mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]][,j],
                paste0(names(mela_cell_type_prediction_DFs_lists[i]), "_",
                names(mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]])[j]))
    }
}

for(i in 1:length(mela_cell_type_prediction_DFs_lists)){    
    for(j in 1:ncol(mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]])){
        if(!names(mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]])[j] == "Alveolar"){ # Skip alveolar
            plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[i]], "Y1", "Y2",
                        runName <- tSNE_run_names[i],
                        scoresName <- paste0(names(mela_cell_type_prediction_DFs_lists[i]), "_",
                                             names(mela_cell_type_prediction_DFs_lists[[i]][[cohort_index]])[j]),
                        datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index], outDir,
                                             labelPositions <- NULL, colourPositions <- NULL,
                                             highestOnTop <- TRUE)
        }
    }
}
```

#### __4d)__: Plot CDKs

Viewing CDK expression in the t-SNEs can sometimes be helpful in identifying regions of interest, such as groups of cells
that may be proliferating. I tried all of the CDK genes available in the melanoma scRNA-Seq, and three seemed useful (CDK1/3/14),
particularly CDK1:

```{r}
CDKsDir <- paste0(tSNEsDir, "CDKs/")
dir.create(CDKsDir, showWarnings = FALSE)

#CDKs available in this dataset
colnames(tTPM_mela_4558)[grep("^CDK", colnames(tTPM_mela_4558))]

gene_names_to_print <- c("CDK1", "CDK3", "CDK14")

print_tSNE_coloured_by_gene(gene_names_to_print,
                            mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                            tTPM_mela_4558, run_name, outDir <- CDKsDir, point_size <- 1,
                            all_HUGO_abbreviations_and_fullnames, dataDir)
```

#### __4e)__: Malignant Cell Clusters

#####__(i)__ Selection of Malignant Cell Clusters

We now have t-SNE plots coloured using predicted stemness scores, transcription levels, the patient each cell is from, malignancy and cell type annotations
from the original publication, and immune and stromal scores from the ESTIMATE software package. Using these different types of information to build
up an understanding of this cell population, it's apparent that there are distinct clusters of malignant cells from different patients in
the t-SNE.

We can select these clusters of malignant cells manually, using the t-SNE. For example, I selected cluster 4 by selecting a group of malignant cells
from sample 78 as follows: 

```{r}
cohort_index <- 1
perpl_index <- 1 # perplexity
outDir <- tSNEsDir
run_name <- tSNE_run_names[perpl_index]

plot_categories(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                run_name, categoryName <- "Samples",
                datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
```

Plotting with a grid in the background is helpful when selecting cells in the t-SNE:

```{r}
plot_categories_with_grid(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                          run_name, "Samples",
                          datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index], outDir)
```

The cluster to be selected, is in a box that covers the area: Y1 :-33.5 to -40, Y2 = 3 to 8.5
(the coordinates may be slightly different for you.)

```{r}
cluster_4_pos <-
    select_tSNE_box_positions(tSNE_info = mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(33.5,40), Y2_borders = c(3,8.5))
```

Display cluster 4 in black:

```{r}
plot_categories_with_grid(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                          run_name, "Samples",
                          datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index],
                          outDir, labelPositions <- NULL, colourPositions <- cluster_4_pos)
```

Here's eight tumour cell clusters each from distinct patients, that we'll analyse further.

[There's one larger malignant cluster from another patient in the middle at the top, but it's less clear if there's a coherent higher stemness and higher
transcription region within it, so we won't focus on it here.]

```{r}
str(mela_clusters_info)

mela_clusters_info <- mela_clusters_info[-9] # not going to study cluster 9 here, as it has a less obvious pCSC group

eight_clusters_tmp <- c(mela_clusters_info[[1]], mela_clusters_info[[2]], mela_clusters_info[[3]],
                        mela_clusters_info[[4]], mela_clusters_info[[5]], mela_clusters_info[[6]],
                        mela_clusters_info[[7]], mela_clusters_info[[8]])
plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score", names(mela_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, eight_clusters_tmp, highestOnTop <- TRUE)
```

These are the same groups, shown with cluster labels.

```{r}
tmpData <- mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]]

tmpData$cluster_numbers <- rep(0, nrow(tmpData))
tmpData$cluster_numbers[(mela_clusters_info[[1]])] <- '1'
tmpData$cluster_numbers[(mela_clusters_info[[2]])] <- '2'
tmpData$cluster_numbers[(mela_clusters_info[[3]])] <- '3'
tmpData$cluster_numbers[(mela_clusters_info[[4]])] <- '4'
tmpData$cluster_numbers[(mela_clusters_info[[5]])] <- '5'
tmpData$cluster_numbers[(mela_clusters_info[[6]])] <- '6'
tmpData$cluster_numbers[(mela_clusters_info[[7]])] <- '7'
tmpData$cluster_numbers[(mela_clusters_info[[8]])] <- '8'

label_positions <- c(
    mela_clusters_info[[1]],
    mela_clusters_info[[2]],
    mela_clusters_info[[3]],
    mela_clusters_info[[4]],
    mela_clusters_info[[5]],
    mela_clusters_info[[6]],
    mela_clusters_info[[7]],
    mela_clusters_info[[8]]
)

cohort_index <- 1
perpl_index <- 1

plot_categories(tmpData, "Y1", "Y2", runName <- "tSNE_070", categoryName <- "Samples",
                datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index], pCSCsDir,
                label_positions, colourPositions <- NULL, labelsVec <- NULL, label_field <- 'cluster_numbers')
tmpData <- NULL
```

This is an outlier cell according to PluriTest with much higher predicted stemness:

```{r}
plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score", names(mela_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, 4180, highestOnTop <- TRUE)
```

And this is a small cluster of 7 cells of high stemness cells:

```{r}
plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score", names(mela_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, c(4180, 2787, 2859, 2982, 2968, 2978, 3019), highestOnTop <- TRUE)
```

#####__(ii)__: Intra-cluster Correlations & Stemness Markers

Now that we have some clusters of malignant cells from different patients, let's look at the stemness correlations within each of them, to find
genes highly correlated with stemness for each of these tumours. The most postively correlated genes will be used later as stemness markers,
to assist identification of pCSCs.

```{r}
cohort_index <- 1

scores_to_use_list <- list()
for(i in 1:length(mela_stemness_prediction_DFs_lists)){
    scores_to_use_list[[i]] <- mela_stemness_prediction_DFs_lists[[i]][[cohort_index]]
}
names(scores_to_use_list) <- names(mela_stemness_prediction_DFs_lists)
```

For the three stemness predictions that we obtained using the HipSci transcriptomes, identify genes correlated with stemness
increase (genes that are activated have positive correlations; genes that are deactivated have negative correlations).

```{r}
mela_stemness_correlations_list <- list()

for(i in 1:length(mela_clusters_info)){
    cat(paste0("Cluster = ", names(mela_clusters_info)[i], "\n"))
    tExprMatr_to_use <- tTPM_mela_4558[mela_clusters_info[[i]],]
    
    for(k in 1:length(scores_to_use_list)){
        score_to_use_index <- which(names(scores_to_use_list[[k]]) == 'IPS') # for HipSci IPS predictions
                                                                             # (vs fibro & mononuclear)
        scores_to_use <- scores_to_use_list[[k]][,score_to_use_index][mela_clusters_info[[i]]]
    
        mela_stemness_correlations_list[[length(mela_stemness_correlations_list)+1]] <-
            get_expression_scores_correlations(tExprMatr_to_use, scores_to_use)
        names(mela_stemness_correlations_list)[length(mela_stemness_correlations_list)] <-
            paste0(names(mela_clusters_info)[i], "; ", names(scores_to_use_list)[k])
    
        cat(paste0(run_name, "; ", names(scores_to_use_list)[k], ":\n"))
        cat(paste0("\nPositive correlations (top 10):\n"))
        print(mela_stemness_correlations_list[[length(mela_stemness_correlations_list)]][[1]][1:10])
        cat(paste0("\nNegative correlations (top 10):\n"))
        print(mela_stemness_correlations_list[[length(mela_stemness_correlations_list)]][[2]][1:10])
        cat("\n")    
    }
}
```

Do the same for PluriTest stemness predictions:

```{r}
for(i in 1:length(mela_clusters_info)){
    cat(paste0("Cluster = ", names(mela_clusters_info)[i], "\n"))
    tExprMatr_to_use <- tTPM_mela_4558[mela_clusters_info[[i]],]
    scores_to_use <- mela_4558_additional_info$Combined_Pluritest_Score[mela_clusters_info[[i]]]

    mela_stemness_correlations_list[[length(mela_stemness_correlations_list)+1]] <-
        get_expression_scores_correlations(tExprMatr_to_use, scores_to_use)
    names(mela_stemness_correlations_list)[length(mela_stemness_correlations_list)] <-
        paste0(names(mela_clusters_info)[i], "; Combined_Pluritest_Score")

    cat(paste0("Combined_Pluritest_Score:\n"))
    cat(paste0("\nPositive correlations (top 10):\n"))
    print(mela_stemness_correlations_list[[length(mela_stemness_correlations_list)]][[1]][1:10])
    cat(paste0("\nNegative correlations (top 10):\n"))
    print(mela_stemness_correlations_list[[length(mela_stemness_correlations_list)]][[2]][1:10])
    cat("\n")    
}
```

N.B. Immune suppression and suppression of anitviral genes is seen in the negative correlations calculated using PluriTest scores,
in some of the clusters, i.e. in cluster 4: IFIT1 & IFIT3, cluster 5: HLA-C/H/B/A, cluster 6: HLA-E/A/C/B/H, TAPBP, and cluster 8: B2M & HLA-H/B/C/A.

Looking at the genes positively correlated with PluriTest scores, I thought that clusters 2, 4, 5, 6 and 8 were intriguing, so I plotted the expression
of many of the most highly correlated of these genes on the t-SNE.

Clusters 4 & 6 each have higher stemness on one side of the cluster, towards the high PluriTest stemness cluster of 7 cells that we identified previously.
These higher stemness regions are further highlighted by some of genes most correlated with the PluriTest scores when calculated for cluster 4, i.e.
SMC4, POLA1, CDK1, MCM6, NUSAP1, MCM7, FANCI & NCAPG, and for cluster 6, TYMS. We'll use these as stemness markers to help delineate some pCSCs in each
of the malignant clusters. Some of these may indicate cycling/proliferating cells in the high stemness and high transcription regions.

```{r}
gene_names_to_print <- c('SMC4', 'POLA1', 'CDK1', 'MCM6', 'NUSAP1', 'MCM7', 'FANCI', 'NCAPG', 'TYMS')

markersDir <- paste0(tSNEsDir, "high_stemness_markers/")
dir.create(markersDir, showWarnings = FALSE)
outDir <- markersDir

print_tSNE_coloured_by_gene(gene_names_to_print, mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                            tTPM_mela_4558, run_name, outDir, point_size <- 1,
                            all_HUGO_abbreviations_and_fullnames, dataDir)
```

#### __4f)__: Manual pCSC Group Selections

In step 4e, some melanoma stemness markers were selected: CDK1, FANCI, NCAPG, SMC4, NUSAP1, MCM7, MCM6, POLA1 & TYMS.

We've seen that stemness and transcription increase along gradients within the malignant clusters; and these lead to regions with high transcription and
high predicted stemness.

Cluster 4 has elevated expression of many of these stemness markers towards one side. When looking closely at this area, the highest expression of many of these
genes occurs approximately within two contiguous groups of cells near to each other. In the t-SNE I generated, these are in the location: Y1 >38 and <40; Y2 >6.7 and <8.
However, there's one extra cell in this box, which is not part of the two groups, so I don't want to select it with the others. This extra cell is located
within: Y1 >38 and <38.5; Y2 >6.7 and <7.

So, to select pCSC group 4:

```{r}
outDir <- pCSCsDir

pCSCs_box1 <-
    select_tSNE_box_positions(tSNE_info = mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(38,40), Y2_borders = c(6.8,8)) # seed 50
pCSCs_box2 <-
    select_tSNE_box_positions(tSNE_info = mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(38,38.5), Y2_borders = c(6.8,7))

plot_scores_with_grid(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      runName <- "tSNE_070", scoresName <- "Total_transcription",
                      datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index],
                      outDir, labelPositions <- NULL, colourPositions <- pCSCs_box1, highestOnTop <- TRUE)
plot_scores_with_grid(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      "tSNE_070", "Total_transcription", names(mela_tSNE_list_with_annotations)[cohort_index],
                      outDir, NULL, pCSCs_box2, highestOnTop <- TRUE)

which(pCSCs_box1 %in% pCSCs_box2) # 16

pCSCs_group_4 <- pCSCs_box1[-(which(pCSCs_box1 %in% pCSCs_box2))]
```

Using a similar methodolgy, a pCSC group was selected from each of the other malignant cell clusters, to obtain eight pCSC groups in total.

Two further pCSC groups were selected from a high stemness region in the t-SNE: pCSC group 9 is cluster of 7 cells with high predicted stemness that we looked at, and
pCSC group 10 is that same cluster of 7 cells plus some other high stemness cells located nearby.

```{r}
str(mela_pCSC_groups)

cohort_index <- 1
perpl_index <- 1

mela_pCSC_groups_positions <- c(mela_pCSC_groups[[1]], mela_pCSC_groups[[2]], mela_pCSC_groups[[3]],
                                mela_pCSC_groups[[4]], mela_pCSC_groups[[5]], mela_pCSC_groups[[6]],
                                mela_pCSC_groups[[7]], mela_pCSC_groups[[8]], mela_pCSC_groups[[9]],
                                mela_pCSC_groups[[10]])

plot_scores(mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
            runName <- "tSNE_070", scoresName <- "Total_transcription",
            datasetName <- names(mela_tSNE_list_with_annotations)[cohort_index],
            outDir, labelPositions <- NULL, colourPositions <- mela_pCSC_groups_positions,
            highestOnTop <- TRUE)
```

----------

### Step 5: Preferential Expression Plots For Selected pCSC Groups ############################################

From eight malignant cell clusters identified (step 4e), putative cancer stem cells were select from each (step 4g).
We can now look at the preferential expression in these pCSC groups, relative to other cells, e.g. compared to all of the cells in their
malignant cluster, all malignant tumour cells, or all cells including non-malignant cells.

Some info needs to be prepared for one of the preferential expression plots; HUGO gene groups (e.g. families of related genes, or genes that are
components of the same molecular complex) will be mapped to the genes in scRNA-Seq expression matrix (columns).

```{r}
#head(all_HUGO_groups_mordered)
mela_transcript_names <- colnames(tTPM_mela_4558)

# Create a list of HUGO gene groups, with each group containing the position of each group member in the
# gene expression matrix columns.
mela_all_HUGO_groups_positions <- map_gene_positions_to_HUGO_gene_groups(mela_transcript_names,
                                                                         all_HUGO_groups_mordered)
save(mela_all_HUGO_groups_positions, file = paste0(saveDir, "mela_all_HUGO_group_positions.RData"))
#load(paste0(saveDir, "mela_all_HUGO_group_positions.RData"))

# Subset to only use HUGO gene groups that have two or more members in the gene expression matrix
mela_all_HUGO_groups_positions_two_or_more <-
    mela_all_HUGO_groups_positions[which(lengths(mela_all_HUGO_groups_positions) >= 2)]
gene_group_positions_to_use <- mela_all_HUGO_groups_positions_two_or_more
```

The expression matrix of TPM values will be converted to TPM count fractions:

```{r}
expression_matrix <- count_fractions_transcriptwise(tTPM_mela_4558)
save(expression_matrix, file = paste0(saveDir, "mela_4558_expression_matrix-TPMCountFractions.RData"))
#load(paste0(saveDir, "mela_4558_expression_matrix-TPMCountFractions.RData"))
frag_count_method <- 'TPM CFs' # expression measurement type, but just a label, can be anything
```

The locations of the malignant tumour cells in the expression matrix, also need to be noted, for each sample.

```{r}
mela_4558_all_malignant_positions <- which(mela_4558_additional_info$Cell_malignancy == "Yes")

# All samples (each patient's tumour cells including non-malignant cells) in the cohort:
mela_sample_59_positions_4558 <- which(mela_4558_additional_info$Samples == 59) # Cluster 7 source
mela_sample_71_positions_4558 <- which(mela_4558_additional_info$Samples == 71) # Cluster 6 source
mela_sample_78_positions_4558 <- which(mela_4558_additional_info$Samples == 78) # Cluster 4 source
mela_sample_80_positions_4558 <- which(mela_4558_additional_info$Samples == 80) # Cluster 3 source
mela_sample_81_positions_4558 <- which(mela_4558_additional_info$Samples == 81) # Cluster 2 source
mela_sample_82_positions_4558 <- which(mela_4558_additional_info$Samples == 82) # Cluster 1 source
mela_sample_88_positions_4558 <- which(mela_4558_additional_info$Samples == 88) # Cluster 5 source
mela_sample_89_positions_4558 <- which(mela_4558_additional_info$Samples == 89) # Cluster 8 source

# Positions for each patient's malignant tumour cells
mela_sample_59_positions_tumour_only_4558 <-
    mela_sample_59_positions_4558[which(mela_sample_59_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 7 source
mela_sample_71_positions_tumour_only_4558 <-
    mela_sample_71_positions_4558[which(mela_sample_71_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 6 source
mela_sample_78_positions_tumour_only_4558 <-
    mela_sample_78_positions_4558[which(mela_sample_78_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 4 source
mela_sample_80_positions_tumour_only_4558 <-
    mela_sample_80_positions_4558[which(mela_sample_80_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 3 source
mela_sample_81_positions_tumour_only_4558 <-
    mela_sample_81_positions_4558[which(mela_sample_81_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 2 source
mela_sample_82_positions_tumour_only_4558 <-
    mela_sample_82_positions_4558[which(mela_sample_82_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 1 source
mela_sample_88_positions_tumour_only_4558 <-
    mela_sample_88_positions_4558[which(mela_sample_88_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 5 source
mela_sample_89_positions_tumour_only_4558 <-
    mela_sample_89_positions_4558[which(mela_sample_89_positions_4558 %in%
                                        mela_4558_all_malignant_positions)] # Cluster 8 source
```

To generate the preferential expression plots, set some parameters for the run:

```{r}
mela_pCSC_groups_to_run <- list()

mela_pCSC_groups_info_list_tmp <- list()
mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_1_cluster_1'
mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[1]]
mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 82'
mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_82_positions_tumour_only_4558
names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
                                           'Source patient tumour positions')
mela_pCSC_groups_to_run[[1]] <- mela_pCSC_groups_info_list_tmp

mela_pCSC_groups_info_list_tmp <- list()
mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_2_cluster_2'
mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[2]]
mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 81'
mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_81_positions_tumour_only_4558
names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
                                           'Source patient tumour positions')
mela_pCSC_groups_to_run[[2]] <- mela_pCSC_groups_info_list_tmp

tissue_type                                  <<- "melanoma"
outDir                                       <- prefExprDir
no_ribo                                      <- TRUE  # set to TRUE to remove ribosomal transcripts from
                                                      # some of the expression bar charts
separate_row_for_high_expression_transcripts <- FALSE # set to TRUE to print highest expression bars on a
                                                      # seperate row
```

We'll analyse two of the pCSC groups here for brevity, but uncomment the following code if you'd like to look at the others.

```{r}
#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_3_cluster_3'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[3]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 80'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_80_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[3]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_4_cluster_4'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[4]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 78'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_78_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[4]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_5_cluster_5'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[5]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 88'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_88_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[5]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_6_cluster_6'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[6]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 71'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_71_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[6]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_7_cluster_7'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[7]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 59'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_59_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[7]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_8_cluster_8'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[8]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Sample 89'
#mela_pCSC_groups_info_list_tmp[[4]] <- mela_sample_89_positions_tumour_only_4558
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[8]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_9'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[9]]
#mela_pCSC_groups_info_list_tmp[[3]] <- 'Samples 84 (1 cell) & 60 (6 cells)'
#mela_pCSC_groups_info_list_tmp[[4]] <- ''
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[9]] <- mela_pCSC_groups_info_list_tmp

#mela_pCSC_groups_info_list_tmp <- list()
#mela_pCSC_groups_info_list_tmp[[1]] <- 'mela_pCSC_group_10'
#mela_pCSC_groups_info_list_tmp[[2]] <- mela_pCSC_groups[[10]]
#mela_pCSC_groups_info_list_tmp[[3]] <-
#    'Samples 84 (14 cells), 60 (6 cells), 65 (2 cells), 80 (20 cells), & 94 (1 cell)'
#mela_pCSC_groups_info_list_tmp[[4]] <- ''
#names(mela_pCSC_groups_info_list_tmp) <- c('pCSC group name', 'pCSC group positions', 'Source patient name',
#                                           'Source patient tumour positions')
#mela_pCSC_groups_to_run[[10]] <- mela_pCSC_groups_info_list_tmp
```

__We can now print the preferential expression plots.__

(If the images generated look small when viewing them in the web pages, you could try opening each of them in a new tab
e.g. by using a right mouse click.)

```{r fig.width=20}
mela_all_results_ranks_list <- list() # Preferential expression ranked results will be stored in here for
                                      # further analysis

mela_4558_tumour_only_positions <- mela_4558_all_malignant_positions

for(r in 1:length(mela_pCSC_groups_to_run)){

    this_sample_name     <- mela_pCSC_groups_to_run[[r]][[3]]
    this_pCSC_group_name <- mela_pCSC_groups_to_run[[r]][[1]]
    cat(paste0(this_pCSC_group_name, "; ", this_sample_name, "\n"))

    # Subset transcription info for pCSC group
    pCSC_group_pos <- NULL    
    pCSC_group_pos <- mela_pCSC_groups_to_run[[r]][[2]]
    this_patient_tumour_only_positions <- mela_pCSC_groups_to_run[[r]][[4]]

    # Generate graphs
    mela_all_results_ranks_list[[r]] <- print_preferential_expression_bar_graphs(
        pCSC_group_name                    <- this_pCSC_group_name,
        pCSC_group_positions               <- pCSC_group_pos,
        HUGO_groups_sets_of_columnNumbers  <- gene_group_positions_to_use,
        tExprMatr                          <- expression_matrix,
        tumour_only_positions              <- mela_4558_tumour_only_positions,
        this_patient_tumour_only_positions,
        outDir,
        tissue_type,
        family_type_name                   <- 'HUGO',
        patient_source_name                <- this_sample_name,
        frag_count_method,
        no_ribo,
        separate_row_for_high_expression_transcripts,
        max_number_of_graph_rows_stacked_transcripts <- 2,
        max_number_of_graph_rows_stacked_families <- 2,
        all_HUGO_abbreviations_and_fullnames,
        dataDir
    )
    names(mela_all_results_ranks_list)[[r]] <- this_pCSC_group_name
}
save(mela_all_results_ranks_list, file = paste0(saveDir, "mela_all_results_ranks_list_raw.RData"))
#load(paste0(saveDir, "mela_all_results_ranks_list_raw.RData"))
```

We need to print cell numbers on the t-SNEs that correspond with the cell labels displayed on the stacked transcript preferential expression bar graphs,
so that graph can be related to the t-SNE:

```{r}
cohort_index <- 1
perpl_index <- 1

for(i in 1:length(mela_pCSC_groups_to_run)){
    this_pCSC_group_name <- mela_pCSC_groups_to_run[[i]][[1]]
    this_pCSC_group_pos  <- mela_pCSC_groups_to_run[[i]][[2]]
    
    this_perplexity_name <- names(mela_tSNE_list_with_annotations[[cohort_index]])[perpl_index]
    this_sample_name     <- names(mela_tSNE_list_with_annotations)[[cohort_index]]
    this_tSNE_results_with_annotations <-
        mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]] # taking first from perplexity list
    rownames(this_tSNE_results_with_annotations)[(this_pCSC_group_pos)] <-
        paste0("#", as.character(1:length(this_pCSC_group_pos)))
    plot_scores(this_tSNE_results_with_annotations, "Y1", "Y2", this_pCSC_group_name, "Total_transcription",
                this_sample_name, tSNEsDir, this_pCSC_group_pos)
    
    outFile <- paste0(tSNEsDir, this_pCSC_group_name, '-Total_transcription_Y1vsY2.tif') 
    renamed_file <- paste0(tSNEsDir, this_pCSC_group_name, "-Total_transcription_Y1vsY2-",
                           length(this_pCSC_group_pos), "_cells-labelled.tif")
    file.rename(outFile, renamed_file)
}
```

__These are the top ranked preferentially expressed genes and gene groups:__

```{r}
for(i in 1:length(mela_all_results_ranks_list)){
    cat(paste0("\n## Rankings for ", names(mela_all_results_ranks_list[i]), " ##\n\n"))
    cat(paste0("Top 100 transcripts for ", names(mela_all_results_ranks_list[i]), ":\n"))
    print(names(mela_all_results_ranks_list[[i]][[1]]))
    cat(paste0("\nTop 100 families for ", names(mela_all_results_ranks_list[i]), ":\n"))
    print(names(mela_all_results_ranks_list[[i]][[2]]))
    cat(paste0("\nTop 100 transcripts in ranked families for ", names(mela_all_results_ranks_list[i]), ":\n"))
    print(names(mela_all_results_ranks_list[[i]][[3]][1:100]))
}
```

----------

### Step 6: t-SNEs for the Highest Ranked Preferentially Expressed  Genes ############################################

We can explore expression patterns within the melanoma cell population of the genes identified as highly preferentially expressed,
to further understand their relationship with the pCSCs and the entire cell population.

Here are a few examples taken from the top ranked preferentially expressed genes and gene groups:

```{r}
names(mela_all_results_ranks_list[[1]][[3]][which(names(mela_all_results_ranks_list[[1]][[3]][1:100]) %in%
                                                  names(mela_all_results_ranks_list[[2]][[3]][1:100]))])
names(mela_all_results_ranks_list[[2]][[3]][which(names(mela_all_results_ranks_list[[2]][[3]][1:100]) %in%
                                                  names(mela_all_results_ranks_list[[1]][[3]][1:100]))])

gene_names_to_print <- c('TEKT2', 'PAGE2', 'BAGE4', 'VCX')
cohort_index <- 1
perpl_index <- 1

print_tSNE_coloured_by_gene(gene_names_to_print, mela_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                            tTPM_mela_4558, run_name, outDir <- tSNEsDir, point_size <- 1,
                            all_HUGO_abbreviations_and_fullnames, dataDir)
```

This is the end of cscplots tutorial 2.

__Now try [tutorial 3](tutorial_3.html)__, which looks at CSCs from NSCLC, and studies their preferential expression. This includes
generating Cells Per Million values for the scRNA-Seq matrix to assist preferential expression analysis.

----------
