<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cscplots Tutorial 3 • cscplots</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cscplots Tutorial 3">
<meta property="og:description" content="cscplots">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cscplots</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial_1.html">cscplots Tutorial 1</a>
    </li>
    <li>
      <a href="../articles/tutorial_2.html">cscplots Tutorial 2</a>
    </li>
    <li>
      <a href="../articles/tutorial_3.html">cscplots Tutorial 3</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tutorial_3_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>cscplots Tutorial 3</h1>
            
      
      
      <div class="hidden name"><code>tutorial_3.Rmd</code></div>

    </div>

    
    
<div id="tutorial-3-identifying-and-characterising-cancer-stem-cells-in-nsclc-single-cell-rna-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#tutorial-3-identifying-and-characterising-cancer-stem-cells-in-nsclc-single-cell-rna-seq" class="anchor"></a><strong>Tutorial 3: Identifying and Characterising Cancer Stem Cells In NSCLC Single-Cell RNA-Seq</strong>
</h3>
<p>Tutorial 3 uses an NSCLC scRNA-Seq dataset published by Lambrechts <em>et al.</em> (Nature Medicine 2018). The original dataset contains 52,698 cells, but we’ll use 45,000 of these cells (randomly sampled cells) to make some of the computational methods used more tractable. Please contact that original authors to obtain the full 52,698 cells.</p>
<p>A pipeline will be constructed using functions from the cscplots package to perform the following steps.</p>
<ul>
<li>Generate a t-SNE using read counts for the NSCLC cells; predict stemness and other cell types in each of the cells, using
    iPSC RNA-Seq and melanoma scRNA-Seq; identify genes associated with higher stemness and higher transcription, in NSCLC cells.</li>
<li>Use the t-SNE to select putative cancer stem cells (pCSCs), by using the stemness predictions and other information such as transcription gradients and predicted cell types.</li>
<li>Detect preferential expression in the selected pCSCs using a Cells Per Million matrix, to find genes driving these cells.</li>
</ul>

<p>
<h4>Tutorial 3 Pipeline Overview:</h4>
<p>
<p>
<center>
<img src="figures/cscplots_tutorial_2.png" width="900">
</center>
</div>
<div id="step-1-t-sne-generation" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-t-sne-generation" class="anchor"></a>Step 1: t-SNE Generation</h3>
<p>Create the results directories, and load the NSCLC expression data included with this package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://Matrix.R-forge.R-project.org/">"Matrix"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/r-gregmisc/gtools">"gtools"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jkrijthe/Rtsne">"Rtsne"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/topepo/caret/">"caret"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/imbs-hl/ranger">"ranger"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"e1071"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">"randomForest"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org">"ggplot2"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/slowkow/ggrepel">"ggrepel"</a></span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>ggrepel.max.overlaps <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"cscplots"</span><span class="op">)</span>

<span class="va">project_name</span>   <span class="op">&lt;-</span> <span class="st">"NSCLC_45000"</span> 
<span class="va">wd</span>             <span class="op">&lt;-</span> <span class="st">'./'</span> <span class="co"># set the working directory</span>
<span class="va">saveDir</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">wd</span>, <span class="st">'save/'</span><span class="op">)</span>
<span class="va">resultsBaseDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'results/'</span><span class="op">)</span>
<span class="va">resultsDir</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'results/'</span>, <span class="va">project_name</span>, <span class="st">'/'</span><span class="op">)</span>
<span class="va">tSNEsDir</span>       <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="st">'tSNEs/'</span><span class="op">)</span>
<span class="va">pCSCsDir</span>       <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, <span class="st">"pCSC_group_selection/"</span><span class="op">)</span>
<span class="va">prefExprDir</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="st">"preferential_expression/"</span><span class="op">)</span>
<span class="va">dataDir</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">wd</span>, <span class="st">'../data/'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">saveDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">resultsBaseDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">resultsDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">pCSCsDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">prefExprDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">dataDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Combine the NSCLC data that's available with this package.</span>

<span class="va">tCounts_NSCLC_45000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">tCounts_NSCLC_1_15000</span>, <span class="va">tCounts_NSCLC_15001_30000</span>, <span class="va">tCounts_NSCLC_30001_45000</span><span class="op">)</span>
<span class="va">tCounts_NSCLC_1_15000</span>     <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">tCounts_NSCLC_15001_30000</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">tCounts_NSCLC_30001_45000</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>

<span class="va">tTPM_mela_4558</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">tTPM_mela_1_2500</span>, <span class="va">tTPM_mela_2501_4558</span><span class="op">)</span>
<span class="va">tTPM_mela_1_2500</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">tTPM_mela_2501_4558</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></code></pre></div>
<p>We’re going to work with the raw read counts initially, but if you prefer, you can use TPMs. You can generate TPMs by uncommenting this:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#gene_lengths_NSCLC_45000[1:10] # Gene lengths for the NSCLC data are automatically available in this package</span>
<span class="co">#transcript_lengths_NSCLC_45000[1:10] # Transcript lengths are also in this package</span>

<span class="co"># Or to get the gene or transcript lengths from biomart:</span>
<span class="co">#gene_names &lt;- colnames(tCounts_NSCLC_45000)</span>
<span class="co">#gene_lengths_NSCLC_45000       &lt;- get_gene_lengths_using_gene_names(gene_names)</span>
<span class="co">#transcript_lengths_NSCLC_45000 &lt;- get_transcript_lengths_using_gene_names(gene_names)</span>

<span class="co">#no_transcript_lengths_pos &lt;- as.numeric(which(is.na(transcript_lengths_NSCLC_45000)))</span>
<span class="co">#transcript_lengths_NSCLC_45000_no_lengths_removed &lt;- transcript_lengths_NSCLC_45000[-(no_transcript_lengths_pos)]</span>
<span class="co">#lengths_to_use &lt;- transcript_lengths_NSCLC_45000_no_lengths_removed</span>
<span class="co">#counts_to_use &lt;- tCounts_NSCLC_45000[,-c(no_transcript_lengths_pos)]</span>

<span class="co">#tTPM_NSCLC_45000      &lt;- tpm_samplewise(counts_to_use, lengths_to_use)</span>
<span class="co">#save(tTPM_NSCLC_45000, file = paste0(saveDir, "tTPM_NSCLC_45000.RData"))</span>
<span class="co">##load(paste0(saveDir, "tTPM_NSCLC_45000.RData"))</span>
<span class="co">#tTPM_NSCLC_45000 &lt;- NULL # because we'll not use it here</span></code></pre></div>
<p>When generating a t-SNE using the NSCLC scRNA-Seq expression matrix, the recommended perplexity parameter value to use is 70. However, you can use a list of perplexity values to generate multiple t-SNEs if you prefer.</p>
<p>Here we’ll use the raw counts for the input, but you could use TPMs (which can be generated as shown above).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perplexityValues</span>      <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">70</span><span class="op">)</span>
<span class="va">perplexityValuesPrint</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_zeros_to_perplexity_values.html">add_zeros_to_perplexity_values</a></span><span class="op">(</span><span class="va">perplexityValues</span><span class="op">)</span>
<span class="va">tSNE_run_names</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"tSNE_"</span>, <span class="va">perplexityValuesPrint</span><span class="op">)</span>

<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span></code></pre></div>
<p>However, generating the t-SNE for the 45,000 NSCLC cells is computationally intensive. You can use a shortcut and use the following t-SNE info which is available in this package:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">NSCLC_45000_counts_tSNE_results</span><span class="op">)</span>
<span class="co">#&gt; List of 1</span>
<span class="co">#&gt;  $ Perplexity=70:'data.frame':   45000 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ Y1: num [1:45000] 27.1341 0.0347 18.164 -5.5993 0.7671 ...</span>
<span class="co">#&gt;   ..$ Y2: num [1:45000] -7.5 19 15.2 -27.5 26.2 ...</span></code></pre></div>
<p>To generate the t-SNE yourself:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="co"># for reproduction of results; comment for unfixed results</span>

<span class="co">#NSCLC_45000_counts_tSNE_results &lt;-</span>
<span class="co">#    run_all_tSNEs(perplexityValues, tCounts_NSCLC_45000, project_name, outDir, montage &lt;- 0) </span>
<span class="co">#save(NSCLC_45000_counts_tSNE_results, file = paste0(saveDir, "NSCLC_45000_counts_tSNE_results.RData"))</span>
<span class="co">#load(paste0(saveDir, "NSCLC_45000_counts_tSNE_results.RData"))</span></code></pre></div>
<p>Now add the t-SNE results to a list. This data structure is used so that it’s possible to add further cohort’s t-SNE results later for concurrent analysis - as demonstrated in tutorial 1.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_tSNE_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_tSNE_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_45000_counts_tSNE_results</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">project_name</span></code></pre></div>
<hr>
</div>
<div id="step-2-cell-stemness-and-other-cell-type-predictions" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-cell-stemness-and-other-cell-type-predictions" class="anchor"></a>Step 2: Cell Stemness and Other Cell Type Predictions</h3>
<p>In the same way as shown in tutorials 1 and 2, each cell in the NSCLC population will be given a stemness score (relative to fibroblasts and mononuclear cells). This uses random forests trained on stem cell (iPSC) RNA-Seq expression data from HipSci. Below three random forest implementations are used, which you can use compare and contrast the stemness gradients, particularly when overlaid on the t-SNEs, but you could use just one or two.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">number_of_genes_to_use</span> <span class="op">&lt;-</span> <span class="fl">16000</span>

<span class="co"># Put the NSCLC expression matrix in a list first:</span>
<span class="va">NSCLC_tCounts_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_tCounts_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tCounts_list</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NSCLC_45000_counts"</span>

<span class="va">NSCLC_stemness_prediction_DFs_lists</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'ranger'</span>,
                               <span class="va">NSCLC_tCounts_list</span>, <span class="va">hipsci_training_data</span>,
                               <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>, <span class="va">number_of_genes_to_use</span>,
                               <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'counts'</span>, <span class="va">resultsDir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ranger_hipsci_TPM"</span>
<span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'rf'</span>,
                               <span class="va">NSCLC_tCounts_list</span>, <span class="va">hipsci_training_data</span>,
                               <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>, <span class="va">number_of_genes_to_use</span>,
                               <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'counts'</span>, <span class="va">resultsDir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"rf_hipsci_TPM"</span>
<span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'randomForest'</span>,
                               <span class="va">NSCLC_tCounts_list</span>, <span class="va">hipsci_training_data</span>,
                               <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>, <span class="va">number_of_genes_to_use</span>,
                               <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'counts'</span>, <span class="va">resultsDir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"randomForest_hipsci_TPM"</span>

<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"NSCLC_stemness_prediction_DFs_lists.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "NSCLC_stemness_prediction_DFs_lists.RData"))</span></code></pre></div>
<p>Additional information available in this package for the NSCLC cells are: - cell type annotations provided by the NSCLC dataset publishers; - each cell’s patient source (sample number); - each cell’s tumour location (“patient piece”); - PluriTest predicted stemness scores; - ESTIMATE generated scores for stromal and immune cells.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span>
<span class="co">#&gt;                     Combined_Pluritest_Score Samples Cell_type Patient_piece</span>
<span class="co">#&gt; TAAGTGCAGTAGTGCG_23                 1.991112       5    Tumour        Middle</span>
<span class="co">#&gt; CACACCTCACGAAATA_24                 1.349034       5   Myeloid        Normal</span>
<span class="co">#&gt; TAGAGCTTCACATACG_23                 1.514655       5   Myeloid        Middle</span>
<span class="co">#&gt; CCCTCCTCACCTGGTG_22                 1.508521       5    T_cell          Edge</span>
<span class="co">#&gt; ATCACGATCCCTGACT_23                 2.323920       5    Tumour        Middle</span>
<span class="co">#&gt;                     ESTIMATE_stromal_scores ESTIMATE_immune_scores</span>
<span class="co">#&gt; TAAGTGCAGTAGTGCG_23                169.9282               646.2285</span>
<span class="co">#&gt; CACACCTCACGAAATA_24                740.7100              1077.7239</span>
<span class="co">#&gt; TAGAGCTTCACATACG_23                673.3327              1770.3552</span>
<span class="co">#&gt; CCCTCCTCACCTGGTG_22                522.6086              1510.5415</span>
<span class="co">#&gt; ATCACGATCCCTGACT_23                413.4537               199.4386</span></code></pre></div>
<p>However, if you have scRNA-Seq without cell type annotations, we can predict cell types using another expression matrix if it has cell types annotations available (this is very similar to what we just did with the HipSci data). You can use the cscplots function run_ML_cellType_predictors() to do this. For illustration purposes, we could use the melanoma scRNA-Seq from Tirosh <em>et al.</em> that was used in tutorial 2, to predict cell types in the NSCLC cells. If you were working with a dataset other than NSCLC, you could use the NSCLC expression matrix combined with the NSCLC cell types to predict cell types also.</p>
<p>These predictions will take a while (~a few hours) though, and you can skip these predictions if you are not interested. They may however, help you to understand the cell population further. Additionally, you could use several random forest implementations as used above for stemness predictions, to allow comparison of their results to reinforce confidence in the predictions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">number_of_genes_to_use</span> <span class="op">&lt;-</span> <span class="fl">16000</span>
<span class="co">#number_of_genes_to_use &lt;- 10000 # makes the computations slightly quicker, but 16000 is prefered</span>

<span class="co"># Additional info available for the melanoma cells:</span>
<span class="va">mela_4558_additional_info</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span>
<span class="co">#&gt;                             Combined_Pluritest_Score Samples Mutation_status</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb                     1.637538      72       NRAS_Q61R</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb                   1.444785      58       Wild_type</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb                     2.060787      71       NRAS_Q61L</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb                 2.350515      81      BRAF_V600E</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb                  2.278504      80       NRAS_Q61L</span>
<span class="co">#&gt;                             PD1_status PDL1_status    Sex Preop_drug_types</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb        Unknown     Unknown Female    Immunotherapy</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb      Unknown     Unknown Female    Immunotherapy</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb        Unknown     Unknown   Male             None</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb    Unknown     Unknown Female             None</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb     Unknown     Unknown Female             None</span>
<span class="co">#&gt;                             Preop_total_drugs Postop_drugs Survival</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb                     4         None    Alive</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb                   1         None    Alive</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb                     0         None    Alive</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb                 0         None    Alive</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb                  0         None    Alive</span>
<span class="co">#&gt;                                                 Sites Sites_concise</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb     External iliac lymph node    Lymph node</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb     Subcutaneous leg lesion  Subcutaneous</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb              Transverse colon         Bowel</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb       Axillary lymph node    Lymph node</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb        Axillary lymph node    Lymph node</span>
<span class="co">#&gt;                             Sites_summary Cell_malignancy Cell_type Age</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb          Lymphoid              No     Bcell  57</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb         Distant              No     Tcell  67</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb           Distant             Yes Malignant  79</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb      Lymphoid             Yes Malignant  43</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb       Lymphoid             Yes Malignant  86</span>
<span class="co">#&gt;                             ESTIMATE_stromal_scores ESTIMATE_immune_scores</span>
<span class="co">#&gt; Cy72_CD45_H02_S758_comb                    151.8619              1425.0562</span>
<span class="co">#&gt; CY58_1_CD45_B02_S974_comb                  155.6125              1960.3669</span>
<span class="co">#&gt; Cy71_CD45_D08_S524_comb                   -539.6290              -388.5469</span>
<span class="co">#&gt; Cy81_FNA_CD45_B01_S301_comb               -927.3360              -581.6889</span>
<span class="co">#&gt; Cy80_II_CD45_B07_S883_comb                -976.0852              -507.3353</span>

<span class="co"># Combine the melanoma cell type annotations with the melanoma expression matrix for use in predictions</span>
<span class="va">mela_4558_cell_types</span> <span class="op">&lt;-</span> <span class="va">mela_4558_additional_info</span><span class="op">$</span><span class="va">Cell_type</span>
<span class="va">tTPM_mela_4558_with_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind.data.frame</a></span><span class="op">(</span><span class="va">mela_4558_cell_types</span>, <span class="va">tTPM_mela_4558</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tTPM_mela_4558_with_cell_types</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cell_type"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tTPM_mela_4558</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Run the cell type predictions:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Cell type predictions using melanoma data</span>
<span class="va">NSCLC_cell_type_prediction_DFs_lists</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/run_ML_cellType_predictors.html">run_ML_cellType_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'ranger'</span>, <span class="va">NSCLC_tCounts_list</span>,
                               <span class="va">cellTypeData</span> <span class="op">&lt;-</span> <span class="va">tTPM_mela_4558_with_cell_types</span>,
                               <span class="va">expressionData_type</span> <span class="op">&lt;-</span> <span class="st">'mela_cell_types'</span>, <span class="va">number_of_genes_to_use</span>,
                               <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'counts'</span>, <span class="va">resultsDir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ranger_mela_cell_types"</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"NSCLC_cell_type_prediction_DFs_lists.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "NSCLC_cell_type_prediction_DFs_lists.RData"))</span>

<span class="co"># By the way, you could use this dataframe to predict cell types in other cohorts using the NSCLC data:</span>
<span class="co">#tCounts_NSCLC_45000_with_cell_types &lt;- cbind.data.frame(NSCLC_45000_additional_info$Cell_type, tCounts_NSCLC_45000)</span>
<span class="co">#colnames(tCounts_NSCLC_45000_with_cell_types) &lt;- c("Cell_type", colnames(tCounts_NSCLC_45000))</span></code></pre></div>
<p>We can also use other cell annotations in the melanoma dataset to predict similar features in the NSCLC cells. For example, we could try that with “cell malignancy” annotations of the melanoma cells to predict the probabilities of NSCLC cells being maligant. Again, skip this if you want to.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Combine the melanoma cell malignany annotations with the melanoma expression matrix for use in predictions</span>
<span class="va">mela_4558_cell_malignancy</span> <span class="op">&lt;-</span> <span class="va">mela_4558_additional_info</span><span class="op">$</span><span class="va">Cell_malignancy</span>
<span class="va">tTPM_mela_4558_with_cell_malignancy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind.data.frame</a></span><span class="op">(</span><span class="va">mela_4558_cell_malignancy</span>, <span class="va">tTPM_mela_4558</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tTPM_mela_4558_with_cell_malignancy</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cell_type"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tTPM_mela_4558</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Run the cell malignancy predictions:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Cell malignancy predictions using melanoma data:</span>
<span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/run_ML_cellType_predictors.html">run_ML_cellType_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'ranger'</span>, <span class="va">NSCLC_tCounts_list</span>,
                               <span class="va">cellTypeData</span> <span class="op">&lt;-</span> <span class="va">tTPM_mela_4558_with_cell_malignancy</span>,
                               <span class="va">expressionData_type</span> <span class="op">&lt;-</span> <span class="st">'mela_cell_malignancy'</span>, <span class="va">number_of_genes_to_use</span>,
                               <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'counts'</span>, <span class="va">resultsDir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ranger_mela_cell_malignancy"</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span>,
     file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"NSCLC_cell_malignancy_prediction_DFs_lists.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "NSCLC_cell_malignancy_prediction_DFs_lists.RData"))</span></code></pre></div>
<hr>
</div>
<div id="step-3-total-transcription-total-transcribed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-total-transcription-total-transcribed-genes" class="anchor"></a>Step 3: Total Transcription &amp; Total Transcribed Genes</h3>
<p>Calculate total transcription (all expression in all transcripts) for each cell, and total transcribed genes (sum of all expression in all transcripts) for each cell; will be used later, e.g. to help evaluate the t-SNEs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_total_transcription_levels_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_total_transcribed_genes_list</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># only 1 cohort being used in this project i.e. all 45000 NSCLC population cells</span>
                  <span class="co"># from multiple patients</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span>
<span class="co">#df &lt;- tTPM_NSCLC_45000 # If you're using TPMs</span>
<span class="va">total_transcription_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
<span class="va">total_transcribed_genes_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">thisCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">thisCount</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Is na\n"</span><span class="op">)</span> <span class="op">}</span>; <span class="kw">if</span><span class="op">(</span><span class="va">thisCount</span> <span class="op">&lt;=</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"0 or less\n"</span><span class="op">)</span> <span class="op">}</span>
    <span class="va">total_transcribed_genes_vec</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">thisCount</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">total_transcribed_genes_vec</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html">cor.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">total_transcribed_genes_vec</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">total_transcription_levels</span><span class="op">)</span>,
               method<span class="op">=</span><span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Pearson's product-moment correlation</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; data:  as.numeric(total_transcribed_genes_vec) and as.numeric(total_transcription_levels)</span>
<span class="co">#&gt; t = 448.26, df = 44998, p-value &lt; 2.2e-16</span>
<span class="co">#&gt; alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">#&gt; 95 percent confidence interval:</span>
<span class="co">#&gt;  0.9021941 0.9055753</span>
<span class="co">#&gt; sample estimates:</span>
<span class="co">#&gt;       cor </span>
<span class="co">#&gt; 0.9038988</span>
<span class="va">NSCLC_total_transcription_levels_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">total_transcription_levels</span>
<span class="va">NSCLC_total_transcribed_genes_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">total_transcribed_genes_vec</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_total_transcription_levels_list</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">project_name</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_total_transcribed_genes_list</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">project_name</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></code></pre></div>
<hr>
</div>
<div id="step-4-t-sne-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#step-4-t-sne-analysis" class="anchor"></a>Step 4: t-SNE Analysis</h3>
<p>Similarly to in tutorials 1 &amp; 2, we’ll use the stemness scores generated in step 2 to identify regions of higher predicted stemness in the t-SNEs. This will be compared against other information displayed in the t-SNEs, to select pCSCs. Identification of cells that have high transcription, high predicted stemness, and that are consistent with the t-SNE topology (e.g. cell clusters) are the aim. But depending on the tumour population that you are analysing, and the additional information available, you may adapt the features that you are using to identify cancer stem cells.</p>
<p>As always, it’s recommended to view plots in this tutorial using the high resolution tif files that are generated, which will help you to work more precisely with the cells.</p>
<div id="a-annotate-the-t-snes" class="section level4">
<h4 class="hasAnchor">
<a href="#a-annotate-the-t-snes" class="anchor"></a><strong>4a)</strong>: Annotate the t-SNEs</h4>
<p>Add the cell scores previously generated to a t-SNE list for subsequent plotting and analysis:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_tSNE_list_with_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># only 1 cohort used in this project</span>
<span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="va">NSCLC_tSNE_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>

<span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>, <span class="va">perplexityValues</span>,
                               <span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">NSCLC_total_transcription_levels_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                               <span class="va">scoresName</span> <span class="op">&lt;-</span> <span class="st">"Total_transcription"</span><span class="op">)</span>
<span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>, <span class="va">perplexityValues</span>,
                               <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">NSCLC_total_transcribed_genes_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                               <span class="st">"Total_transcribed_genes"</span><span class="op">)</span>

<span class="va">stemness_score_indices_to_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span> <span class="co"># use 2 for HipSci IPS predictions (1 for fibro &amp; 3 for mononuclear)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>, <span class="va">perplexityValues</span>,
                                   <span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">stemness_score_indices_to_use</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,
                                   <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">"_"</span>,
                                   <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">stemness_score_indices_to_use</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>    
    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="co"># do for all cell types</span>
        <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>, <span class="va">perplexityValues</span>,
                                       <span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">"_"</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">}</span>

<span class="va">cell_malignancy_score_index_to_use</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># use 3 for "Yes" malignancy predictions (1 for "Unresolved" &amp; 2 for "No")</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>, <span class="va">perplexityValues</span>,
                                   <span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">cell_malignancy_score_index_to_use</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,
                                   <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">"_"</span>,
                                   <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">cell_malignancy_score_index_to_use</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tSNE_results_with_annotations_tmp</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>
<span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="co">#save(NSCLC_tSNE_list_with_annotations, file = paste0(saveDir, "NSCLC_tSNE_list_with_annotations.RData"))</span></code></pre></div>
</div>
<div id="b-plot-t-sne-cell-scores-and-transcription-gradients" class="section level4">
<h4 class="hasAnchor">
<a href="#b-plot-t-sne-cell-scores-and-transcription-gradients" class="anchor"></a><strong>4b)</strong>: Plot t-SNE Cell Scores and Transcription Gradients</h4>
<p>Plot the t-SNEs with stemness scores, transcription levels, and also predicted cell type scores and predicted cell malignancy scores if you generated them:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">perplexityValues</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>    
    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">3</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="co"># 1:2 are the Y1 &amp; Y2</span>
                                                                                      <span class="co"># values so skip them</span>
        <span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                    <span class="va">runName</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,
                    <span class="va">scoresName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,
                    <span class="va">datasetName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>,
                    <span class="va">labelPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">}</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-16-1.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-2.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-3.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-4.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-5.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-6.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-7.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-8.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-9.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-10.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-11.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-12.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-16-13.png" width="1152"></p>
</div>
<div id="c-additional-info-t-sne-plots" class="section level4">
<h4 class="hasAnchor">
<a href="#c-additional-info-t-sne-plots" class="anchor"></a><strong>4c)</strong>: Additional Info t-SNE Plots</h4>
<p>Additional cell annotations will now be added to the annotated t-SNE list, for subsequent display on the t-SNE i.e.  sample numbers, tumour location info and cell types from the NSCLC publishers, and also ESTIMATE immune/stromal and PluriTest stemness prediction scores:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># only 1 cohort used in this project</span>

<span class="co"># Sample numbers</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">Samples</span><span class="op">)</span>, <span class="st">"Samples"</span><span class="op">)</span>
<span class="co"># Tumour location</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">Patient_piece</span>, <span class="st">"Patient_piece"</span><span class="op">)</span>
<span class="co"># Cell types, and immune and stromal scores</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">Cell_type</span>, <span class="st">"Cell_type"</span><span class="op">)</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">ESTIMATE_stromal_scores</span>, <span class="st">"ESTIMATE_stromal_scores"</span><span class="op">)</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">ESTIMATE_immune_scores</span>, <span class="st">"ESTIMATE_immune_scores"</span><span class="op">)</span>
<span class="co"># PluriTest scores</span>
<span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">perplexityValues</span>,
                               <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">Combined_Pluritest_Score</span>, <span class="st">"Combined_PluriTest_Score"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"NSCLC_tSNE_list_with_annotations.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "NSCLC_tSNE_list_with_annotations.RData"))</span></code></pre></div>
<p>Print the t-SNEs with cell scores, etc. just added (tifs are written to results/NSCLC_45000/tSNEs/):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># if only 1 perplexity value was used, i.e. perplexity = 70, set to 1</span>
<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="fu"><a href="../reference/plot_categories.html">plot_categories</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">run_name</span>,
                <span class="st">"Samples"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
<pre><code><a href="../reference/plot_categories.html">#&gt; NULL
plot_categories(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                "Patient_piece", names(NSCLC_tSNE_list_with_annotations)[cohort_index], outDir)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-2.png" width="1152"></p>
<pre><code><a href="../reference/plot_categories.html">#&gt; NULL

plot_categories(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                "Cell_type", names(NSCLC_tSNE_list_with_annotations)[cohort_index], outDir)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-3.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "ESTIMATE_stromal_scores", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NULL, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-4.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "ESTIMATE_immune_scores", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NULL, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-5.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL

plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Combined_PluriTest_Score", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NULL, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-18-6.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
</div>
<div id="d-cdk-t-snes" class="section level4">
<h4 class="hasAnchor">
<a href="#d-cdk-t-snes" class="anchor"></a><strong>4d)</strong>: CDK t-SNEs</h4>
<p>Viewing CDK associated expression in the t-SNEs can be helpful in identifying regions of interest in the NSCLC:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDKsDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, <span class="st">"CDKs/"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">CDKsDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#CDKs available in this dataset</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tCounts_NSCLC_45000</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^CDK"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tCounts_NSCLC_45000</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt;  [1] "CDK10"      "CDK1"       "CDK11A"     "CDK11B"     "CDK12"     </span>
<span class="co">#&gt;  [6] "CDK13"      "CDK14"      "CDK15"      "CDK16"      "CDK17"     </span>
<span class="co">#&gt; [11] "CDK18"      "CDK19"      "CDK20"      "CDK2"       "CDK2AP1"   </span>
<span class="co">#&gt; [16] "CDK2AP2"    "CDK4"       "CDK5"       "CDK5R1"     "CDK5R2"    </span>
<span class="co">#&gt; [21] "CDK5RAP1"   "CDK5RAP2"   "CDK5RAP3"   "CDK6"       "CDK7"      </span>
<span class="co">#&gt; [26] "CDK8"       "CDK9"       "CDKAL1"     "CDKL1"      "CDKL2"     </span>
<span class="co">#&gt; [31] "CDKL3"      "CDKL4"      "CDKL5"      "CDKN1A"     "CDKN1B"    </span>
<span class="co">#&gt; [36] "CDKN1C"     "CDKN2A"     "CDKN2AIP"   "CDKN2AIPNL" "CDKN2B"    </span>
<span class="co">#&gt; [41] "CDKN2B-AS1" "CDKN2C"     "CDKN2D"     "CDKN3"</span>

<span class="co"># For brevity, just a few are shown here, but others are also interesting.</span>
<span class="co">#gene_names_to_print &lt;- colnames(tCounts_NSCLC_45000)[grep("^CDK", colnames(tCounts_NSCLC_45000))] # To plot all CDKs available</span>
<span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CDK1"</span>, <span class="st">"CDK4"</span>, <span class="st">"CDKN2A"</span><span class="op">)</span>

<span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>, <span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                            <span class="va">tCounts_NSCLC_45000</span>, <span class="va">run_name</span>, <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">CDKsDir</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">0.1</span>,
                            <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="co">#&gt;   CDK1 [cyclin dependent kinase 1]</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-19-1.png" width="1152"></p>
<pre><code>#&gt;   CDK4 [cyclin dependent kinase 4]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-19-2.png" width="1152"></p>
<pre><code>#&gt;   CDKN2A [cyclin dependent kinase inhibitor 2A]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-19-3.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
</div>
<div id="e-selection-of-tumour-cell-groups-and-clusters" class="section level4">
<h4 class="hasAnchor">
<a href="#e-selection-of-tumour-cell-groups-and-clusters" class="anchor"></a><strong>4e)</strong>: Selection of Tumour Cell Groups and Clusters</h4>
<p>Looking at the t-SNEs that we’ve generated so far, there are tumour cells that group by sample/patient, on the right side of the t-SNE we’re using here. These groups of tumour cells each have regions within them of high transcription (total transcription and total transcribed genes) and higher predicted stemness. These areas also overlap with higher predicted malignancy if you created those scores.</p>
<p>We’ll select these groups of tumour cells, and also all of the tumour cells for each patient. These groups of tumour cells can be used later for further analysis, e.g. to compare against pCSCs selected from within them for preferential expression analysis.</p>
<p>Select the tumour cells for each patient:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_45000_tumour_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">==</span> <span class="st">'Tumour'</span><span class="op">)</span>

<span class="va">patient1_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">patient2_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">patient3_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">patient4_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">patient5_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_45000_additional_info</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">5</span><span class="op">)</span>

<span class="va">patient1_tumour_pos</span> <span class="op">&lt;-</span> <span class="va">patient1_pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">patient1_pos</span> <span class="op">%in%</span> <span class="va">NSCLC_45000_tumour_pos</span><span class="op">)</span><span class="op">]</span>
<span class="va">patient2_tumour_pos</span> <span class="op">&lt;-</span> <span class="va">patient2_pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">patient2_pos</span> <span class="op">%in%</span> <span class="va">NSCLC_45000_tumour_pos</span><span class="op">)</span><span class="op">]</span>
<span class="va">patient3_tumour_pos</span> <span class="op">&lt;-</span> <span class="va">patient3_pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">patient3_pos</span> <span class="op">%in%</span> <span class="va">NSCLC_45000_tumour_pos</span><span class="op">)</span><span class="op">]</span>
<span class="va">patient4_tumour_pos</span> <span class="op">&lt;-</span> <span class="va">patient4_pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">patient4_pos</span> <span class="op">%in%</span> <span class="va">NSCLC_45000_tumour_pos</span><span class="op">)</span><span class="op">]</span>
<span class="va">patient5_tumour_pos</span> <span class="op">&lt;-</span> <span class="va">patient5_pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">patient5_pos</span> <span class="op">%in%</span> <span class="va">NSCLC_45000_tumour_pos</span><span class="op">)</span><span class="op">]</span>

<span class="va">NSCLC_patient_tumour_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">patient1_tumour_pos</span>
<span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">patient2_tumour_pos</span>
<span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">patient3_tumour_pos</span>
<span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">patient4_tumour_pos</span>
<span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">patient5_tumour_pos</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_patient_tumour_info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'patient1_tumour_pos'</span>, <span class="st">'patient2_tumour_pos'</span>, <span class="st">'patient3_tumour_pos'</span>,
                                      <span class="st">'patient4_tumour_pos'</span>, <span class="st">'patient5_tumour_pos'</span><span class="op">)</span></code></pre></div>
<p>Manually select the groups of tumour cells for each patient/sample on the right side of the t-SNE:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="co"># All patient 1 tumour cells:</span>
<span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">run_name</span>,
                      <span class="st">"Total_transcription"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>,
                      <span class="va">outDir</span>, <span class="cn">NULL</span>, <span class="va">patient1_tumour_pos</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-1.png" width="1152"></p>
<pre><code>#&gt; NULL
# Select only those on the right side of the t-SNE:
cluster_box_pos_tmp &lt;-
    select_tSNE_box_positions(tSNE_info = NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(27,40), Y2_borders = c(7,9.1))
patient1_malignant_cluster_pos &lt;- patient1_tumour_pos[which(patient1_tumour_pos %in% cluster_box_pos_tmp)]
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, patient1_malignant_cluster_pos, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-2.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL

# Select patient 2 malignant cells group:
plot_scores_with_grid(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                      "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
                      outDir, NULL, patient2_tumour_pos, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-3.png" width="1152"></p>
<pre><code>#&gt; NULL
cluster_box_pos_tmp &lt;-
    select_tSNE_box_positions(tSNE_info = NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(27.5,31), Y2_borders = c(8,10))
patient2_malignant_cluster_pos &lt;- patient2_tumour_pos[which(patient2_tumour_pos %in% cluster_box_pos_tmp)]
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, patient2_malignant_cluster_pos, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-4.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL

# Patient 3 malignant cells group:
plot_scores_with_grid(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                      "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
                      outDir, NULL, patient3_tumour_pos, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-5.png" width="1152"></p>
<pre><code>#&gt; NULL
cluster_box_pos_tmp &lt;-
    select_tSNE_box_positions(tSNE_info = NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(31,38), Y2_borders = c(3,8))
patient3_malignant_cluster_pos &lt;- patient3_tumour_pos[which(patient3_tumour_pos %in% cluster_box_pos_tmp)]
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, patient3_malignant_cluster_pos, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-6.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL

# Patient 4 malignant cells group:
plot_scores_with_grid(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                      "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
                      outDir, NULL, patient4_tumour_pos, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-7.png" width="1152"></p>
<pre><code>#&gt; NULL
cluster_box_pos_tmp &lt;-
    select_tSNE_box_positions(tSNE_info = NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(32,39), Y2_borders = c(-5,2))
patient4_malignant_cluster_pos &lt;- patient4_tumour_pos[which(patient4_tumour_pos %in% cluster_box_pos_tmp)]
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, patient4_malignant_cluster_pos, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-8.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL

# Patient 5 malignant cells group:
plot_scores_with_grid(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
                      "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
                      outDir, NULL, patient5_tumour_pos, highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-9.png" width="1152"></p>
<pre><code>#&gt; NULL
cluster_box_pos_tmp &lt;-
    select_tSNE_box_positions(tSNE_info = NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]],
                              Y1_borders = c(21,34), Y2_borders = c(-13,6))
patient5_malignant_cluster_pos &lt;- patient5_tumour_pos[which(patient5_tumour_pos %in% cluster_box_pos_tmp)]
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, patient5_malignant_cluster_pos, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-21-10.png" width="1152"></p>
<pre><code>#&gt; NULL

NSCLC_malignant_clusters_info &lt;- list()
NSCLC_malignant_clusters_info[[1]] &lt;- patient1_malignant_cluster_pos
NSCLC_malignant_clusters_info[[2]] &lt;- patient2_malignant_cluster_pos
NSCLC_malignant_clusters_info[[3]] &lt;- patient3_malignant_cluster_pos
NSCLC_malignant_clusters_info[[4]] &lt;- patient4_malignant_cluster_pos
NSCLC_malignant_clusters_info[[5]] &lt;- patient5_malignant_cluster_pos
names(NSCLC_malignant_clusters_info) &lt;- c('NSCLC_malignant_cluster1', 'NSCLC_malignant_cluster2',
                                          'NSCLC_malignant_cluster3', 'NSCLC_malignant_cluster4',
                                          'NSCLC_malignant_cluster5')</code></pre>
<p>If you didn’t have tumour annotations in your scRNA-Seq dataset, you could use cell type predictions instead, such as using the melanoma cell types info:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_45000_predicted_malignant_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.3</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
            <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>, <span class="va">NSCLC_45000_predicted_malignant_pos</span>,
            <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-22-1.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
</div>
<div id="f-stemness-correlations" class="section level4">
<h4 class="hasAnchor">
<a href="#f-stemness-correlations" class="anchor"></a><strong>4f)</strong>: Stemness Correlations</h4>
<p>Now that we have some clusters of malignant cells from different patients, let’s look at the stemness correlations within each of them, to find genes highly correlated with predicted stemness.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="co"># Transcripts correlated with predicted stemness:</span>
<span class="va">scores_to_use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_prediction_DFs_lists</span><span class="op">)</span></code></pre></div>
<p>You can examine different subsets of cells. For example, we can look inside all tumour cells for each patient, or within each of the malignant cell groups that were selected:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#NSCLC_clusters_info &lt;- NSCLC_patient_tumour_info # Use this for all tumour cells per patient</span>
<span class="va">NSCLC_clusters_info</span> <span class="op">&lt;-</span> <span class="va">NSCLC_malignant_clusters_info</span>

<span class="va">NSCLC_stemness_correlations_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">tExprMatr_to_use</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span>
    
    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">score_to_use_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">'IPS'</span><span class="op">)</span> <span class="co"># for HipSci IPS predictions</span>
                                                                             <span class="co"># (vs fibro &amp; mononuclear)</span>
        <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">score_to_use_index</span><span class="op">]</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>
    
        <span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tExprMatr_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span>
    
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">run_name</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>    
    <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster1</span>
<span class="co">#&gt; tSNE_070; ranger_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      CMC2      RBX1    TMBIM6     PSMA1      ETFB     SRSF9      RPA2     CD151 </span>
<span class="co">#&gt; 0.6183706 0.6127188 0.6106691 0.5934751 0.5817899 0.5795623 0.5758795 0.5727046 </span>
<span class="co">#&gt;     REXO2    CYB5R3 </span>
<span class="co">#&gt; 0.5726995 0.5712639 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       KMT2D     PPP1R37       AP4M1       APOL2     SLC35D2     PCDHB10 </span>
<span class="co">#&gt;  -0.4359486  -0.3392593  -0.3357678  -0.3336006  -0.3245674  -0.3155265 </span>
<span class="co">#&gt;       PGBD4      ANKMY1       PLAUR RP6-65G23.5 </span>
<span class="co">#&gt;  -0.3155265  -0.3115638  -0.3111348  -0.3017508 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; rf_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;    UNC119    ATP5G3    MYEOV2     MRPL4     SAP18   NDUFB10     PSMB1    NDUFS7 </span>
<span class="co">#&gt; 0.6495795 0.6246321 0.5969438 0.5950545 0.5806764 0.5752335 0.5739095 0.5706296 </span>
<span class="co">#&gt;      SDHC   LAMTOR1 </span>
<span class="co">#&gt; 0.5671184 0.5615432 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      ZMIZ2    SH3KBP1      PDE4C      PLAUR      NCALD      VEGFA    ANGPTL4 </span>
<span class="co">#&gt; -0.3517708 -0.3413825 -0.3372386 -0.3358673 -0.3312764 -0.3224426 -0.3222916 </span>
<span class="co">#&gt;  LINC01123       PLAU      LARS2 </span>
<span class="co">#&gt; -0.3213768 -0.3192618 -0.3174475 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; randomForest_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     SRSF9    UQCR10   TMEM50A    ATG101     ROGDI     ACTN1      HSF1   SPATA20 </span>
<span class="co">#&gt; 0.6193439 0.5790251 0.5716361 0.5682137 0.5646422 0.5620615 0.5605324 0.5601607 </span>
<span class="co">#&gt;      LSM4      EMG1 </span>
<span class="co">#&gt; 0.5547375 0.5518750 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       ZNF579        CNNM4         LIG3        GCNT2       CRIPAK RP11-666O2.2 </span>
<span class="co">#&gt;   -0.3474763   -0.3459094   -0.3428823   -0.3217373   -0.3213768   -0.3213768 </span>
<span class="co">#&gt;       NUDT11 CTC-479C5.12        UBE3D        WDFY3 </span>
<span class="co">#&gt;   -0.3147301   -0.3056651   -0.3036198   -0.3002882 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster2</span>
<span class="co">#&gt; tSNE_070; ranger_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     MYDGF     STAP2    TIMM22      CFL1 C14orf166       CD9      SDHC     BANF1 </span>
<span class="co">#&gt; 0.5390528 0.5209031 0.5092547 0.5086818 0.5069502 0.5015589 0.4963556 0.4956891 </span>
<span class="co">#&gt;    NDFIP2   TAX1BP3 </span>
<span class="co">#&gt; 0.4944985 0.4915588 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     ZNF200     B3GNT8      TRBC2     B3GNT9   SLC16A10   PLS3-AS1   TMEM161B </span>
<span class="co">#&gt; -0.3345475 -0.3296880 -0.3158550 -0.2951054 -0.2812723 -0.2738289 -0.2709601 </span>
<span class="co">#&gt;      USP49  CTBP1-AS2     FAM43A </span>
<span class="co">#&gt; -0.2707120 -0.2666531 -0.2658919 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; rf_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;    STOML2    ARPC1A     PGAM1      YBX1    MRPL37      MLF2      PARL     EIF5A </span>
<span class="co">#&gt; 0.5965714 0.5891647 0.5841143 0.5776399 0.5762852 0.5749432 0.5707864 0.5670836 </span>
<span class="co">#&gt;      NGRN       FAH </span>
<span class="co">#&gt; 0.5638178 0.5588472 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         TNNT1        ZNF766         HCAR2         SCAF4         TYW1B </span>
<span class="co">#&gt;    -0.3740700    -0.3298526    -0.2878235    -0.2872127    -0.2844117 </span>
<span class="co">#&gt;         ALG12        FBXO33         CBLN3 CTC-444N24.11         RNF39 </span>
<span class="co">#&gt;    -0.2754779    -0.2655761    -0.2606005    -0.2546477    -0.2524535 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; randomForest_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;   TAX1BP3      DPP7    NDUFB9      RCE1     MYDGF      RAC1   ATP6V0B      CD63 </span>
<span class="co">#&gt; 0.5572696 0.5301971 0.5200110 0.5193008 0.5113409 0.5113365 0.4929049 0.4925454 </span>
<span class="co">#&gt;      RPL3     RPS4X </span>
<span class="co">#&gt; 0.4913636 0.4828680 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         ZNF200          USP33 CTD-2553L13.10       C6orf226         STXBP3 </span>
<span class="co">#&gt;     -0.3154985     -0.3091538     -0.3035045     -0.2788763     -0.2738758 </span>
<span class="co">#&gt;         CCDC67         KLHL22         TIMM21         CCHCR1          TNNT1 </span>
<span class="co">#&gt;     -0.2738289     -0.2709601     -0.2694969     -0.2611772     -0.2585714 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster3</span>
<span class="co">#&gt; tSNE_070; ranger_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      MYL6      GPX4    MT-CO2    MT-CYB      TSPO      OAZ1      RPS9     RPLP1 </span>
<span class="co">#&gt; 0.5983753 0.5890842 0.5822982 0.5808419 0.5807249 0.5805695 0.5804950 0.5795641 </span>
<span class="co">#&gt;     SERF2    MT-ND4 </span>
<span class="co">#&gt; 0.5761640 0.5748099 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      KHDC1  LINC01191 AP000695.4       HULC      PALM3      ADPRH AC009950.2 </span>
<span class="co">#&gt; -0.1409134 -0.1241711 -0.1199456 -0.1199456 -0.1189406 -0.1185291 -0.1171792 </span>
<span class="co">#&gt;      HOXA5    ZKSCAN7    ANKDD1A </span>
<span class="co">#&gt; -0.1158292 -0.1144792 -0.1123192 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; rf_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     COX6C     ATP5L     EIF3K     HINT1    MT-ND1    MT-CO2     RPS11    MT-ND4 </span>
<span class="co">#&gt; 0.4661065 0.4616056 0.4586435 0.4506175 0.4495311 0.4444114 0.4433432 0.4416296 </span>
<span class="co">#&gt;    MT-CYB    MT-ND5 </span>
<span class="co">#&gt; 0.4403065 0.4398200 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;   AC009120.11  RP11-54O7.17  RP1-239B22.5         SSC4D       FAM86B1 </span>
<span class="co">#&gt;    -0.1412090    -0.1407205    -0.1395445    -0.1358090    -0.1285237 </span>
<span class="co">#&gt;     PCSK6-AS1       CCDC157 RP11-672L10.6 RP11-380O24.1         PDE6G </span>
<span class="co">#&gt;    -0.1261988    -0.1252171    -0.1249183    -0.1247834    -0.1231191 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; randomForest_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;    RPL35A     RPS18     PFDN5      RPS9     TCEB2     RPS15       FAU     RPS14 </span>
<span class="co">#&gt; 0.4275312 0.4186493 0.4148316 0.4104216 0.4018611 0.3994986 0.3990034 0.3981494 </span>
<span class="co">#&gt;    RPL23A     SERF2 </span>
<span class="co">#&gt; 0.3974214 0.3968971 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         ADCY7 CTD-2308L22.1         SCIMP      APOBEC3A          FGD2 </span>
<span class="co">#&gt;    -0.1620053    -0.1585295    -0.1567143    -0.1459334    -0.1441047 </span>
<span class="co">#&gt;        AGPAT4        KLHL25          MNDA     LINC01431       POLR2J2 </span>
<span class="co">#&gt;    -0.1428224    -0.1400078    -0.1386733    -0.1347290    -0.1339253 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster4</span>
<span class="co">#&gt; tSNE_070; ranger_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     RPL30    RPL23A      RPL8     RPS18      IDH2    RPL37A     RPS16      RPS9 </span>
<span class="co">#&gt; 0.4645068 0.4548158 0.4544106 0.4510148 0.4505274 0.4475255 0.4475043 0.4467213 </span>
<span class="co">#&gt;      RPS2     RPL29 </span>
<span class="co">#&gt; 0.4426969 0.4412592 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         PARVG RP11-473M20.5   RP11-53I6.3         ACSL6 CTD-2653M23.3 </span>
<span class="co">#&gt;   -0.10376745   -0.10168210   -0.09939838   -0.09525425   -0.09377061 </span>
<span class="co">#&gt;    AC145110.1     LINC00637 CTD-2001C12.1          PIRT        FCGR3A </span>
<span class="co">#&gt;   -0.09348938   -0.09280896   -0.09164913   -0.09154354   -0.09131313 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; rf_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     PPDPF      DSTN      OST4      PGK1     RPS4X     CXXC5    FKBP1A      RPS7 </span>
<span class="co">#&gt; 0.3191864 0.3132218 0.3123411 0.3085210 0.3020248 0.2922232 0.2916322 0.2915593 </span>
<span class="co">#&gt;     RPS21      SDHA </span>
<span class="co">#&gt; 0.2872304 0.2866697 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      CXCL10      CXCL11       ISG20        XAF1        ZBP1       ISG15 </span>
<span class="co">#&gt;  -0.1870627  -0.1787527  -0.1638297  -0.1593232  -0.1482438  -0.1450785 </span>
<span class="co">#&gt;        CD38         CD2        VNN3 RP11-1E22.1 </span>
<span class="co">#&gt;  -0.1402588  -0.1370190  -0.1334571  -0.1310512 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; randomForest_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     ARPC5      PTMA      RPS2     H2AFY     XRCC5  TMEM106C   HNRNPA0      DSTN </span>
<span class="co">#&gt; 0.4253602 0.4217566 0.4192618 0.4147765 0.4103759 0.4059917 0.4019911 0.4013131 </span>
<span class="co">#&gt;     RPS21     RPS17 </span>
<span class="co">#&gt; 0.4003146 0.3985455 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     SAMSN1     IL10RA       TYMP     CXCL11      TRBC2       CD3E      PTPRC </span>
<span class="co">#&gt; -0.1439377 -0.1423372 -0.1378065 -0.1304724 -0.1297264 -0.1296034 -0.1277653 </span>
<span class="co">#&gt;       HGH1       LAG3       VNN1 </span>
<span class="co">#&gt; -0.1205575 -0.1200520 -0.1198005 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster5</span>
<span class="co">#&gt; tSNE_070; ranger_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     RPL13      RPS8     RPL30     RPL11       FAU    RPS27A     RPS20     RPLP2 </span>
<span class="co">#&gt; 0.6636049 0.6549729 0.6500882 0.6490674 0.6470173 0.6460950 0.6453705 0.6441644 </span>
<span class="co">#&gt;     SERF2    RPL27A </span>
<span class="co">#&gt; 0.6440229 0.6411402 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       ZSCAN23  RP11-290F5.2  RP11-798L4.1      IGLV6-57         ATCAY </span>
<span class="co">#&gt;   -0.06959515   -0.06478427   -0.05684212   -0.05671668   -0.05665846 </span>
<span class="co">#&gt;        ZNF888 RP11-367N14.2         OPRD1       COL11A2 RP11-403A21.1 </span>
<span class="co">#&gt;   -0.05656663   -0.05537264   -0.05491793   -0.05441794   -0.05440865 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; rf_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      RPS8     RPL13     SERF2     RPL29    RPS27A     RPL15     RPS16     RPLP2 </span>
<span class="co">#&gt; 0.4828221 0.4773000 0.4712413 0.4582275 0.4564988 0.4549420 0.4543811 0.4542686 </span>
<span class="co">#&gt;    RPS15A     RPL30 </span>
<span class="co">#&gt; 0.4525606 0.4518974 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         PTPN22         IFITM1    SLCO4A1-AS1     AC092580.3           ASB2 </span>
<span class="co">#&gt;    -0.09088081    -0.07676279    -0.07388051    -0.06861911    -0.06850439 </span>
<span class="co">#&gt;      LINC00473  RP11-1000B6.5 RP11-508N22.13     AC005754.8         SLC1A6 </span>
<span class="co">#&gt;    -0.06694330    -0.06682905    -0.06632872    -0.06608450    -0.06244670 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; tSNE_070; randomForest_hipsci_TPM:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     SERF2 LINC01420     SFTA2   MT-ATP6    MT-CO1    SPINT2    MT-ND4    MT-CO2 </span>
<span class="co">#&gt; 0.4435837 0.4346487 0.4210716 0.4199587 0.4188746 0.4138101 0.4129645 0.4102450 </span>
<span class="co">#&gt;     UQCRH      CST3 </span>
<span class="co">#&gt; 0.4013455 0.4001113 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;    SERPINB3   PARD3-AS1       ACRBP    SERPINB4        CST7      ZNF564 </span>
<span class="co">#&gt; -0.09443916 -0.08592393 -0.08057027 -0.07987555 -0.07793383 -0.07490313 </span>
<span class="co">#&gt;        TYMP        AQP9     UNQ6494      SLAMF1 </span>
<span class="co">#&gt; -0.07107012 -0.06870599 -0.06773399 -0.06708496</span></code></pre></div>
<p>We can also look at correlations with other features of the data, e.g. cell malignancy predictions:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transcripts correlated with predicted cell malignancy:</span>
<span class="va">scores_to_use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_malignancy_prediction_DFs_lists</span><span class="op">)</span>

<span class="co">#NSCLC_clusters_info &lt;- NSCLC_patient_tumour_info</span>
<span class="va">NSCLC_clusters_info</span> <span class="op">&lt;-</span> <span class="va">NSCLC_malignant_clusters_info</span>

<span class="va">NSCLC_predicted_malignancy_correlations_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">tExprMatr_to_use</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span>
    
    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">score_to_use_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">'Yes'</span><span class="op">)</span> <span class="co"># for mela TPM cell malignancy</span>
                                                                             <span class="co"># predictions</span>
        <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">score_to_use_index</span><span class="op">]</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>
    
        <span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tExprMatr_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span>
    
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">run_name</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_predicted_malignancy_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>    
    <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster1</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_malignancy:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;    NDUFC1     PPP4C    TMBIM6    EIF4A1    ATP1B3     COX8A     VDAC1     SRSF9 </span>
<span class="co">#&gt; 0.7562309 0.7543169 0.7370144 0.7273449 0.7253791 0.7214840 0.7179000 0.7131251 </span>
<span class="co">#&gt;    COX6A1  NIPSNAP1 </span>
<span class="co">#&gt; 0.7094218 0.7086402 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     RNF19B   C1orf198       GJB4      GSDMC    FAM208B      PLAUR    TMEM45B </span>
<span class="co">#&gt; -0.3655113 -0.3653764 -0.3545054 -0.3500558 -0.3437147 -0.3328994 -0.3213379 </span>
<span class="co">#&gt;   EIF4EBP2      S100P     STEAP4 </span>
<span class="co">#&gt; -0.3194911 -0.3188424 -0.3174293 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster2</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_malignancy:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     PDIA4      ENO1      P4HB     DDOST    HNRNPK      NUDC     PSMD2       TKT </span>
<span class="co">#&gt; 0.5449622 0.5386265 0.5385927 0.5330813 0.5118810 0.5107331 0.5098580 0.5086884 </span>
<span class="co">#&gt;   SEC61A1     ALDOA </span>
<span class="co">#&gt; 0.5075563 0.5000550 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       FTCDNL1        NFATC1       CSNK1G1 RP11-152N13.5 RP11-332H14.2 </span>
<span class="co">#&gt;    -0.3206063    -0.2880579    -0.2811790    -0.2789668    -0.2745619 </span>
<span class="co">#&gt;         FEZF1          ZXDB      KIAA1614         ERAP2          ESR2 </span>
<span class="co">#&gt;    -0.2738943    -0.2632462    -0.2606005    -0.2571885    -0.2553091 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster3</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_malignancy:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     APLP2      TPI1    NDUFS6    FAAP20      GPX4    MT-ND4     RPLP1    NDUFB7 </span>
<span class="co">#&gt; 0.6550304 0.6330454 0.6168002 0.6161378 0.6144785 0.6140127 0.6131635 0.6113396 </span>
<span class="co">#&gt;    MT-CYB    PPP1CA </span>
<span class="co">#&gt; 0.6070229 0.6034831 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     SLC4A4       MNDA       FPR1       PLEK    B4GALT6       CCR2       HRH2 </span>
<span class="co">#&gt; -0.1283334 -0.1253155 -0.1252237 -0.1227774 -0.1207741 -0.1178965 -0.1165755 </span>
<span class="co">#&gt;     AGPAT4   NAALADL1       TNXB </span>
<span class="co">#&gt; -0.1147262 -0.1138577 -0.1126126 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster4</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_malignancy:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;       RAN    SERBP1      PTMA     SNRPB     PSMB1    CHCHD2      CFL1      RPS2 </span>
<span class="co">#&gt; 0.6988943 0.6918730 0.6819027 0.6689798 0.6688533 0.6657410 0.6639542 0.6629987 </span>
<span class="co">#&gt;      CCT5     GSTP1 </span>
<span class="co">#&gt; 0.6621065 0.6590179 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     IL10RA       SRGN      PTPRC      HCLS1       CD3E  EPHA1-AS1 AC145110.1 </span>
<span class="co">#&gt; -0.1269315 -0.1243694 -0.1183214 -0.1170411 -0.1161835 -0.1109081 -0.1073699 </span>
<span class="co">#&gt;       HPGD      SCML4   HOXB-AS1 </span>
<span class="co">#&gt; -0.1068631 -0.1060479 -0.1053286 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster5</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_malignancy:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;       FAU     SRP14     RPS16      RPS8     COX5B    COX4I1     RPL27     RPS17 </span>
<span class="co">#&gt; 0.7769792 0.7717414 0.7600709 0.7577514 0.7516962 0.7506576 0.7500883 0.7497338 </span>
<span class="co">#&gt;      MYL6     RPL13 </span>
<span class="co">#&gt; 0.7466590 0.7457695 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       CCDC136         OPRD1    AL163953.2         CDHR4           LPL </span>
<span class="co">#&gt;   -0.06913926   -0.06779020   -0.06020175   -0.05874026   -0.05633706 </span>
<span class="co">#&gt;         DUOX2        GPR171          XCL2          NSG1 RP11-1057B6.1 </span>
<span class="co">#&gt;   -0.05581730   -0.05406994   -0.05133257   -0.05105689   -0.05053349</span></code></pre></div>
<p>You could look at the cell type correlations, which include a malignant cell type:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transcripts correlated with predicted cell types:</span>
<span class="va">scores_to_use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_cell_type_prediction_DFs_lists</span><span class="op">)</span>

<span class="va">NSCLC_malignant_cell_correlations_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">tExprMatr_to_use</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span>
    
    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">score_to_use_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">'Malignant'</span><span class="op">)</span> <span class="co"># for mela TPM cell</span>
                                                                                   <span class="co"># type predictions</span>
        <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">score_to_use_index</span><span class="op">]</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>
    
        <span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tExprMatr_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span>
    
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">run_name</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_malignant_cell_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>    
    <span class="op">}</span>

<span class="op">}</span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster1</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_types:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;         PLAUR        METRNL        STEAP4         KMT2D         SAMD1 </span>
<span class="co">#&gt;     0.4288615     0.3914695     0.3822108     0.3624554     0.3473013 </span>
<span class="co">#&gt;         USP53 RP11-670E13.6        MMS22L         DUOX1      C1orf198 </span>
<span class="co">#&gt;     0.3385966     0.3315092     0.3208834     0.3182443     0.3151528 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     UBE2L3      PSMB1       NME1      HMGN1       YBX1      SRSF9     DYNLL1 </span>
<span class="co">#&gt; -0.7549279 -0.7373760 -0.7130072 -0.7102197 -0.7087312 -0.7016359 -0.6938047 </span>
<span class="co">#&gt;       PTMA     TMBIM6        ERH </span>
<span class="co">#&gt; -0.6907114 -0.6825395 -0.6819897 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster2</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_types:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;          CHD7         SIMC1        SLC8A1          SOX5         ASXL2 </span>
<span class="co">#&gt;     0.3449052     0.3094297     0.3082229     0.2897031     0.2840570 </span>
<span class="co">#&gt;     LINC-PINT CTD-2517O10.6         OVOS2        KLHL22         PRKD1 </span>
<span class="co">#&gt;     0.2828121     0.2766613     0.2720209     0.2701238     0.2640791 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      ACBD6       CDK4       MLF2    HNRNPA1       CNBP      MYDGF     STOML2 </span>
<span class="co">#&gt; -0.6827153 -0.6786878 -0.6679843 -0.6679586 -0.6473618 -0.6442487 -0.6428526 </span>
<span class="co">#&gt;       MYL6      CLIC1   HSD17B10 </span>
<span class="co">#&gt; -0.6389220 -0.6284326 -0.6279204 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster3</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_types:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;       B4GALT6        SLC4A4   RP11-30K9.6      KIF9-AS1     LINC01191 </span>
<span class="co">#&gt;     0.1521189     0.1421733     0.1309491     0.1188210     0.1182267 </span>
<span class="co">#&gt;      PAN3-AS1  RP11-771K4.1  CTC-366B18.4       FAM155A RP4-545L17.11 </span>
<span class="co">#&gt;     0.1177191     0.1156627     0.1142638     0.1126126     0.1077021 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     MT-CYB      PFDN5     MT-CO1     MT-ND4     MT-CO2       MYL6     MT-ND1 </span>
<span class="co">#&gt; -0.6262167 -0.6195548 -0.6173888 -0.6159876 -0.6149009 -0.6105944 -0.6090249 </span>
<span class="co">#&gt;     MT-CO3      ATP5E      SERF2 </span>
<span class="co">#&gt; -0.6081973 -0.6062654 -0.6002627 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster4</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_types:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;  CTC-492K19.7 RP13-465B17.5     WDR11-AS1 RP11-474O21.5       PCDHGC3 </span>
<span class="co">#&gt;    0.10178544    0.09904019    0.09658648    0.08899862    0.08790995 </span>
<span class="co">#&gt;       ZDHHC11  RP11-38L15.3   RP3-449M8.9  RP11-283G6.3  RP4-563E14.1 </span>
<span class="co">#&gt;    0.08709345    0.08600478    0.08542908    0.08532437    0.08427017 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       PTMA       RPS2        RAN      PSMB1       CFL1     CHCHD2      SNRPB </span>
<span class="co">#&gt; -0.5956436 -0.5946028 -0.5905388 -0.5899269 -0.5845319 -0.5842364 -0.5838819 </span>
<span class="co">#&gt;      PSMA7     COX4I1      ATP5E </span>
<span class="co">#&gt; -0.5804039 -0.5677836 -0.5668645 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster5</span>
<span class="co">#&gt; tSNE_070; ranger_mela_cell_types:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;         CDK18 RP11-1057B6.1           LPL         FBLL1         VTCN1 </span>
<span class="co">#&gt;    0.06320914    0.05997690    0.05918375    0.05716636    0.05643562 </span>
<span class="co">#&gt;         ATCAY    AC005330.2    AL163953.2      OGFR-AS1    AP001046.5 </span>
<span class="co">#&gt;    0.05505145    0.05389479    0.05227603    0.05121172    0.04918694 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;     RPL27A        FAU      RPL27      RPL30      RPLP2     RPS27A      RPL32 </span>
<span class="co">#&gt; -0.7221707 -0.7132535 -0.7095959 -0.7086290 -0.7083522 -0.7075510 -0.7052112 </span>
<span class="co">#&gt;       RPS8     RPS15A       RPL9 </span>
<span class="co">#&gt; -0.7046815 -0.7041389 -0.7035245</span></code></pre></div>
<p>PluriTest stemness correlations:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#NSCLC_clusters_info &lt;- NSCLC_patient_tumour_info</span>
<span class="va">NSCLC_clusters_info</span> <span class="op">&lt;-</span> <span class="va">NSCLC_malignant_clusters_info</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">tExprMatr_to_use</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span>
    <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">NSCLC_45000_additional_info</span><span class="op">$</span><span class="va">Combined_Pluritest_Score</span><span class="op">[</span><span class="va">NSCLC_clusters_info</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>

    <span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tExprMatr_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_clusters_info</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; Combined_Pluritest_Score"</span><span class="op">)</span>

    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Combined_Pluritest_Score:\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>    
<span class="op">}</span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster1</span>
<span class="co">#&gt; Combined_Pluritest_Score:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      RBMX    PPP1CC     KIF5B   HNRNPA1     H2AFZ HNRNPA2B1      SKP1      SNX6 </span>
<span class="co">#&gt; 0.6868626 0.6802843 0.6738888 0.6626736 0.6621689 0.6599792 0.6563157 0.6553863 </span>
<span class="co">#&gt;     HMGB1     VDAC1 </span>
<span class="co">#&gt; 0.6533706 0.6492895 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      DUOX1     S100A9     METRNL       ECM1       CSTB   APOBEC3A   C10orf99 </span>
<span class="co">#&gt; -0.4907122 -0.4595842 -0.4536342 -0.4450080 -0.4025048 -0.3954790 -0.3921104 </span>
<span class="co">#&gt;       KRT7     S100A8        MAL </span>
<span class="co">#&gt; -0.3916499 -0.3886818 -0.3880561 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster2</span>
<span class="co">#&gt; Combined_Pluritest_Score:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      DKC1    MAD2L1    ANP32E    DNAJA1     KPNA2    TUBA1C      CYCS      GAR1 </span>
<span class="co">#&gt; 0.5617845 0.5446091 0.5445802 0.5164068 0.4976904 0.4936097 0.4915946 0.4878083 </span>
<span class="co">#&gt;     CENPF      GARS </span>
<span class="co">#&gt; 0.4877031 0.4785805 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      KRT15      NEAT1       FLII       MUC1      PGAP3     PMEPA1      CLIP2 </span>
<span class="co">#&gt; -0.3773743 -0.3546890 -0.3479728 -0.3360867 -0.3352577 -0.3317932 -0.3295204 </span>
<span class="co">#&gt;       NCK2       CUL7     IGFBP2 </span>
<span class="co">#&gt; -0.3280832 -0.3273727 -0.3229483 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster3</span>
<span class="co">#&gt; Combined_Pluritest_Score:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      PTMA      RPSA   HNRNPA1      YBX1     HMGB1   HNRNPA3    PTGES3     DDX21 </span>
<span class="co">#&gt; 0.5158258 0.4499839 0.4480380 0.4139354 0.4110993 0.4109963 0.4052812 0.4051595 </span>
<span class="co">#&gt;    EEF1B2     H2AFV </span>
<span class="co">#&gt; 0.3995643 0.3985147 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;   PDZK1IP1       MUC1     S100A9       SNCG   HLA-DRB1   HLA-DRB5       CD74 </span>
<span class="co">#&gt; -0.2664294 -0.2312881 -0.2305310 -0.2197599 -0.2134025 -0.2054983 -0.2001342 </span>
<span class="co">#&gt;    HLA-DRA       SAA2   C15orf48 </span>
<span class="co">#&gt; -0.1989418 -0.1961407 -0.1939872 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster4</span>
<span class="co">#&gt; Combined_Pluritest_Score:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     HDAC2    TUBB2B     STMN1   KHDRBS1     HMGB1    TUBA1B      TUBB   HNRNPA1 </span>
<span class="co">#&gt; 0.5107765 0.4837884 0.4504015 0.4156678 0.4018696 0.3984348 0.3971725 0.3961603 </span>
<span class="co">#&gt;     HMGB3     XRCC5 </span>
<span class="co">#&gt; 0.3960736 0.3913800 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;    TNFSF10    S100A11       GBP5     S100A6       TYMP    RARRES3       GBP4 </span>
<span class="co">#&gt; -0.4316878 -0.3943199 -0.3628083 -0.3604064 -0.3535124 -0.3510893 -0.3507356 </span>
<span class="co">#&gt;       LY6E       RHOV      IFI35 </span>
<span class="co">#&gt; -0.3481038 -0.3412743 -0.3373921 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Cluster = NSCLC_malignant_cluster5</span>
<span class="co">#&gt; Combined_Pluritest_Score:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     RPS17     RPL35     RPS15     RPL27     RPS24     RPS16      RPS8     RPLP1 </span>
<span class="co">#&gt; 0.6295285 0.6155287 0.6132456 0.6112749 0.6048911 0.6017378 0.6007828 0.6001304 </span>
<span class="co">#&gt;   C8orf59     RPLP2 </span>
<span class="co">#&gt; 0.5979423 0.5939221 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;          LCN2         KLK11      SERPINB4 CTD-2035E11.3            CP </span>
<span class="co">#&gt;   -0.12904989   -0.08678861   -0.08247904   -0.07015579   -0.06980071 </span>
<span class="co">#&gt;      SERPINB3         LPAR6          HFM1            C6          EML6 </span>
<span class="co">#&gt;   -0.06872044   -0.06831427   -0.06754019   -0.06610388   -0.06547404</span></code></pre></div>
</div>
<div id="g-manual-pcsc-group-selections" class="section level4">
<h4 class="hasAnchor">
<a href="#g-manual-pcsc-group-selections" class="anchor"></a><strong>4g)</strong>: Manual pCSC Group Selections</h4>
<p>From the correlations that we just calculated, here are some selected stemness markers. You could use many of the other highly correlated genes instead though, or even some of those used for melanoma in tutorial 2:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#gene_names_to_print &lt;- c('OAZ1', 'DSTN', 'FAU') # from NSCLC malignant clusters</span>
<span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'NACA'</span>, <span class="st">'MYL6'</span>, <span class="st">'MIF'</span>, <span class="st">'FAU'</span>, <span class="st">'JTB'</span>, <span class="st">'SPINT2'</span><span class="op">)</span> <span class="co"># from NSCLC tumour cells by patient</span>
<span class="co">#gene_names_to_print &lt;- c('SMC4', 'POLA1', 'CDK1', 'MCM6', 'NUSAP1', 'MCM7', 'FANCI',</span>
<span class="co">#                         'NCAPG', 'TYMS') # stemness markers from melanoma</span>

<span class="va">markersDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, <span class="st">"high_stemness_markers/"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">markersDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">markersDir</span>

<span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>, <span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                            <span class="va">tCounts_NSCLC_45000</span>, <span class="va">run_name</span>, <span class="va">outDir</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">0.1</span>,
                            <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="co">#&gt;   NACA [nascent polypeptide associated complex subunit alpha]</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-1.png" width="1152"></p>
<pre><code>#&gt;   MYL6 [myosin light chain 6]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-2.png" width="1152"></p>
<pre><code>#&gt;   MIF [macrophage migration inhibitory factor]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-3.png" width="1152"></p>
<pre><code>#&gt;   FAU [FAU ubiquitin like and ribosomal protein S30 fusion]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-4.png" width="1152"></p>
<pre><code>#&gt;   JTB [jumping translocation breakpoint]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-5.png" width="1152"></p>
<pre><code>#&gt;   SPINT2 [serine peptidase inhibitor, Kunitz type 2]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-28-6.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>Comparing the high resolution tif files showing t-SNEs generated above i.e. using stemness scores, transcription levels, CDKs, stemness markers, and also considering the t-SNE topology, I selected these groups of malignant cells as putative cancer stem cell groups:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_pCSC_groups</span>
<span class="co">#&gt; $NSCLC_pCSC_group_1</span>
<span class="co">#&gt;  [1]   274   606  3460  4667 13017 15153 15925 16635 18788 19083 20091 23065</span>
<span class="co">#&gt; [13] 23387 35242 36085 37568 39886 44488</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NSCLC_pCSC_group_2</span>
<span class="co">#&gt;  [1]  6268  6560  9185  9509  9826 11195 12268 13199 16250 20589 21801 23181</span>
<span class="co">#&gt; [13] 28602 28643 29539 32863 32948 35368 36066 36129 36326 36381 38550 40859</span>
<span class="co">#&gt; [25] 12733 24743</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NSCLC_pCSC_group_3</span>
<span class="co">#&gt;  [1]  2194 11965 15645 18110 19879 22708 29027 31000 31744 31804 34146 34465</span>
<span class="co">#&gt; [13] 36817 43936 44892</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NSCLC_pCSC_group_4</span>
<span class="co">#&gt;  [1]   168  2137  2666  5063  7960 10581 11178 16548 17365 20641 21571 22225</span>
<span class="co">#&gt; [13] 23783 26237 26852 28031 28547 28978 31727 31902 33955 40753 41363 42587</span>
<span class="co">#&gt; [25] 42977 44145 44317 44518  1688  1847  3998  9490  9982 10093 11647 12753</span>
<span class="co">#&gt; [37] 13232 14333 20716 20756 21193 21567 22197 23019 24783 26497 30731 31256</span>
<span class="co">#&gt; [49] 34904 37883 38392 38981 43307  2065  2264 20201 24078 27866 42053 43964</span>
<span class="co">#&gt; [61]  2203 13254 35342 42675  9170 15409 18805</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NSCLC_pCSC_group_5</span>
<span class="co">#&gt;   [1]   865  1399  2445  4777  7845  8156  9449 12323 12576 13682 13712 14319</span>
<span class="co">#&gt;  [13] 16896 16949 18726 19153 19571 20192 20801 23450 27287 29205 30062 30896</span>
<span class="co">#&gt;  [25] 31771 31829 33608 34338 34822 35013 39156 39391 40023 40026 40280 41045</span>
<span class="co">#&gt;  [37] 43147 43337 43352 43986 44301 23635  1504  1951  3193  4838  7852  8488</span>
<span class="co">#&gt;  [49] 10694 11158 16522 18203 18394 19012 19517 19628 21783 22029 22529 23918</span>
<span class="co">#&gt;  [61] 24281 24328 25168 26306 27064 29951 31594 33607 36190 36462 38316 39758</span>
<span class="co">#&gt;  [73] 40279 43187 43742 44550  5027  6617  7261  8095  8565  9324 13155 13992</span>
<span class="co">#&gt;  [85] 19071 19318 23769 26724 33659 35801 36466 39869 40644  3751  7736  9076</span>
<span class="co">#&gt;  [97]  9408 13468 15167 15712 15981 19313 20830 22015 23717 27045 27913 28432</span>
<span class="co">#&gt; [109] 30797 31467 31785 36010 36446 37673 38936 39097 40577 42128 42628 43201</span>
<span class="co">#&gt; [121] 44418</span>

<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="co"># pCSC group 1</span>
<span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">run_name</span>,
            <span class="st">"Total_transcription"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>,
            <span class="va">outDir</span>, <span class="cn">NULL</span>, <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-1.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
# pCSC group 2
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NSCLC_pCSC_groups[[2]], highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-2.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
# pCSC group 3
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NSCLC_pCSC_groups[[3]], highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-3.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
# pCSC group 4
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NSCLC_pCSC_groups[[4]], highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-4.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores.html">#&gt; NULL
# pCSC group 5
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, NSCLC_pCSC_groups[[5]], highestOnTop &lt;- TRUE)</a></code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-5.png" width="1152"></p>
<pre><code>#&gt; NULL

# pCSC groups 1-5
all_NSCLC_pCSC_groups &lt;- c(NSCLC_pCSC_groups[[1]], NSCLC_pCSC_groups[[2]], NSCLC_pCSC_groups[[3]],
                           NSCLC_pCSC_groups[[4]], NSCLC_pCSC_groups[[5]])
plot_scores(NSCLC_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2", run_name,
            "Total_transcription", names(NSCLC_tSNE_list_with_annotations)[cohort_index],
            outDir, NULL, all_NSCLC_pCSC_groups, highestOnTop &lt;- TRUE)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-29-6.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<hr>
</div>
</div>
<div id="step-5-preferential-expression-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#step-5-preferential-expression-plots" class="anchor"></a>Step 5: Preferential Expression Plots</h3>
<p>We have selected a pCSC group from tumour cells for each of the five patients.</p>
<p>We can now look at the preferential expression in these pCSC groups, relative to other cells in the cell population.</p>
<p>Because the NSCLC data is available as read counts, it can be converted to Cells Per Million values, which will help to detect preferential expression. This can use transcript lengths or gene lengths in the CellsPM calculations. We’ll use gene lengths here:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_lengths_NSCLC_45000</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="co"># Data automatically available in this package</span>
<span class="co">#&gt;     A1BG A1BG-AS1     A1CF      A2M  A2M-AS1    A2ML1   A4GALT    A4GNT </span>
<span class="co">#&gt;     8315     7738    86267    48566     3527    64381    29178     8678 </span>
<span class="co">#&gt;     AAAS     AACS </span>
<span class="co">#&gt;    17409    77900</span>

<span class="co"># To get the gene lengths from biomart:</span>
<span class="co">#gene_names &lt;- colnames(tCounts_NSCLC_45000)</span>
<span class="co">#gene_lengths_NSCLC_45000       &lt;- get_gene_lengths_using_gene_names(gene_names)</span>

<span class="va">no_gene_lengths_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gene_lengths_NSCLC_45000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">gene_lengths_NSCLC_45000_no_lengths_removed</span> <span class="op">&lt;-</span> <span class="va">gene_lengths_NSCLC_45000</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="va">no_gene_lengths_pos</span><span class="op">)</span><span class="op">]</span>
<span class="va">lengths_to_use_1</span> <span class="op">&lt;-</span> <span class="va">gene_lengths_NSCLC_45000_no_lengths_removed</span>
<span class="va">counts_to_use_1</span> <span class="op">&lt;-</span> <span class="va">tCounts_NSCLC_45000</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">no_gene_lengths_pos</span><span class="op">)</span><span class="op">]</span>

<span class="va">tCellsPM_NSCLC_45000</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_fractions_per_base_transcriptwise.html">count_fractions_per_base_transcriptwise</a></span><span class="op">(</span><span class="va">counts_to_use_1</span>, <span class="va">lengths_to_use_1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">tCellsPM_NSCLC_45000</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"tCellsPM_NSCLC_45000-gene_lengths.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "tCellsPM_NSCLC_45000-gene_lengths.RData"))</span></code></pre></div>
<p>If you’d like to use transcript lengths to generate the CellsPM matrix, you can uncomment and use this instead:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Data automatically available in this package:</span>
<span class="co">#transcript_lengths_NSCLC_45000[1:10]</span>

<span class="co">## To get the transcript lengths from biomart:</span>
<span class="co">#gene_names &lt;- colnames(tCounts_NSCLC_45000)</span>
<span class="co">#transcript_lengths_NSCLC_45000 &lt;- get_transcript_lengths_using_gene_names(gene_names)</span>

<span class="co">#no_transcript_lengths_pos &lt;- as.numeric(which(is.na(transcript_lengths_NSCLC_45000)))</span>
<span class="co">#transcript_lengths_NSCLC_45000_no_lengths_removed &lt;- transcript_lengths_NSCLC_45000[-(no_transcript_lengths_pos)]</span>
<span class="co">#lengths_to_use_2 &lt;- transcript_lengths_NSCLC_45000_no_lengths_removed</span>
<span class="co">#counts_to_use_2 &lt;- tCounts_NSCLC_45000[,-c(no_transcript_lengths_pos)]</span>

<span class="co">#tCellsPM_NSCLC_45000  &lt;- count_fractions_per_base_transcriptwise(counts_to_use_2, lengths_to_use_2)</span>
<span class="co">#save(tCellsPM_NSCLC_45000, file = paste0(saveDir, "tCellsPM_NSCLC_45000-transcript_lengths.RData"))</span>
<span class="co">#load(paste0(saveDir, "tCellsPM_NSCLC_45000-transcript_lengths.RData"))</span></code></pre></div>
<p>HUGO gene groups need to be mapped to the scRNA-Seq expression matrix genes (columns in the matrix) first:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#head(all_HUGO_groups_mordered)</span>
<span class="va">NSCLC_transcript_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tCellsPM_NSCLC_45000</span><span class="op">)</span> <span class="co"># if you're using CellsPM</span>
<span class="co">#NSCLC_transcript_names &lt;- colnames(tTPM_NSCLC_45000)    # if you're using TPMs</span>
<span class="co">#NSCLC_transcript_names &lt;- colnames(tCounts_NSCLC_45000) # if you're using read counts for the pref expr plots</span>

<span class="co"># Create a list of HUGO gene groups, with each group containing the position of each group member in the</span>
<span class="co"># gene expression matrix columns.</span>
<span class="va">NSCLC_all_HUGO_groups_positions</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/map_gene_positions_to_HUGO_gene_groups.html">map_gene_positions_to_HUGO_gene_groups</a></span><span class="op">(</span><span class="va">NSCLC_transcript_names</span>, <span class="va">all_HUGO_groups_mordered</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">NSCLC_all_HUGO_groups_positions</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"NSCLC_all_HUGO_gene_group_positions-CellsPM.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "NSCLC_all_HUGO_gene_group_positions-CellsPM.RData"))</span>

<span class="co"># Subset to only use HUGO gene groups that have two or more members in the gene expression matrix</span>
<span class="va">NSCLC_all_HUGO_groups_positions_two_or_more</span> <span class="op">&lt;-</span>
    <span class="va">NSCLC_all_HUGO_groups_positions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">NSCLC_all_HUGO_groups_positions</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="va">gene_group_positions_to_use</span> <span class="op">&lt;-</span> <span class="va">NSCLC_all_HUGO_groups_positions_two_or_more</span></code></pre></div>
<p>Setting the parameters for the preferential expression plots, we’ll use here tumour cells from each patient to compare against the pCSCs:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_pCSC_groups_to_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">NSCLC_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'NSCLC_pCSC_group_1_patient_1_tumour'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Patient 1'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name and tumour cells'</span>, <span class="st">'pCSC group positions'</span>,
                                            <span class="st">'Source patient name'</span>, <span class="st">'Source patient tumour positions'</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_info_list_tmp</span>

<span class="va">NSCLC_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'NSCLC_pCSC_group_2_patient_2_tumour'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Patient 2'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name and tumour cells'</span>, <span class="st">'pCSC group positions'</span>,
                                            <span class="st">'Source patient name'</span>, <span class="st">'Source patient tumour positions'</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_info_list_tmp</span>

<span class="va">NSCLC_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'NSCLC_pCSC_group_3_patient_3_tumour'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Patient 3'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name and tumour cells'</span>, <span class="st">'pCSC group positions'</span>,
                                            <span class="st">'Source patient name'</span>, <span class="st">'Source patient tumour positions'</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_info_list_tmp</span>

<span class="va">NSCLC_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'NSCLC_pCSC_group_4_patient_4_tumour'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Patient 4'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name and tumour cells'</span>, <span class="st">'pCSC group positions'</span>,
                                            <span class="st">'Source patient name'</span>, <span class="st">'Source patient tumour positions'</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_info_list_tmp</span>

<span class="va">NSCLC_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'NSCLC_pCSC_group_5_patient_5_tumour'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Patient 5'</span>
<span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_patient_tumour_info</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name and tumour cells'</span>, <span class="st">'pCSC group positions'</span>,
                                            <span class="st">'Source patient name'</span>, <span class="st">'Source patient tumour positions'</span><span class="op">)</span>
<span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_info_list_tmp</span></code></pre></div>
<p>If you’d also like to compare against the malignant cell clusters we selected previously, you can uncomment the following:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#NSCLC_pCSC_groups_info_list_tmp &lt;- list()</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[1]] &lt;- 'NSCLC_pCSC_group_1_patient_1_malignant_cluster'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[2]] &lt;- NSCLC_pCSC_groups[[1]]</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[3]] &lt;- 'Patient 1'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[4]] &lt;- NSCLC_malignant_clusters_info[[1]]</span>
<span class="co">#names(NSCLC_pCSC_groups_info_list_tmp) &lt;- c('pCSC group name and tumour cells', 'pCSC group positions',</span>
<span class="co">#                                            'Source patient name', 'Source patient tumour positions')</span>
<span class="co">#NSCLC_pCSC_groups_to_run[[6]] &lt;- NSCLC_pCSC_groups_info_list_tmp</span>

<span class="co">#NSCLC_pCSC_groups_info_list_tmp &lt;- list()</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[1]] &lt;- 'NSCLC_pCSC_group_2_patient_2_malignant_cluster'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[2]] &lt;- NSCLC_pCSC_groups[[2]]</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[3]] &lt;- 'Patient 2'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[4]] &lt;- NSCLC_malignant_clusters_info[[2]]</span>
<span class="co">#names(NSCLC_pCSC_groups_info_list_tmp) &lt;- c('pCSC group name and tumour cells', 'pCSC group positions',</span>
<span class="co">#                                            'Source patient name', 'Source patient tumour positions')</span>
<span class="co">#NSCLC_pCSC_groups_to_run[[7]] &lt;- NSCLC_pCSC_groups_info_list_tmp</span>

<span class="co">#NSCLC_pCSC_groups_info_list_tmp &lt;- list()</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[1]] &lt;- 'NSCLC_pCSC_group_3_patient_3_malignant_cluster'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[2]] &lt;- NSCLC_pCSC_groups[[3]]</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[3]] &lt;- 'Patient 3'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[4]] &lt;- NSCLC_malignant_clusters_info[[3]]</span>
<span class="co">#names(NSCLC_pCSC_groups_info_list_tmp) &lt;- c('pCSC group name and tumour cells', 'pCSC group positions',</span>
<span class="co">#                                            'Source patient name', 'Source patient tumour positions')</span>
<span class="co">#NSCLC_pCSC_groups_to_run[[8]] &lt;- NSCLC_pCSC_groups_info_list_tmp</span>

<span class="co">#NSCLC_pCSC_groups_info_list_tmp &lt;- list()</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[1]] &lt;- 'NSCLC_pCSC_group_4_patient_4_malignant_cluster'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[2]] &lt;- NSCLC_pCSC_groups[[4]]</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[3]] &lt;- 'Patient 4'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[4]] &lt;- NSCLC_malignant_clusters_info[[4]]</span>
<span class="co">#names(NSCLC_pCSC_groups_info_list_tmp) &lt;- c('pCSC group name and tumour cells', 'pCSC group positions',</span>
<span class="co">#                                            'Source patient name', 'Source patient tumour positions')</span>
<span class="co">#NSCLC_pCSC_groups_to_run[[9]] &lt;- NSCLC_pCSC_groups_info_list_tmp</span>

<span class="co">#NSCLC_pCSC_groups_info_list_tmp &lt;- list()</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[1]] &lt;- 'NSCLC_pCSC_group_5_patient_5_malignant_cluster'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[2]] &lt;- NSCLC_pCSC_groups[[5]]</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[3]] &lt;- 'Patient 5'</span>
<span class="co">#NSCLC_pCSC_groups_info_list_tmp[[4]] &lt;- NSCLC_malignant_clusters_info[[5]]</span>
<span class="co">#names(NSCLC_pCSC_groups_info_list_tmp) &lt;- c('pCSC group name and tumour cells', 'pCSC group positions',</span>
<span class="co">#                                            'Source patient name', 'Source patient tumour positions')</span>
<span class="co">#NSCLC_pCSC_groups_to_run[[10]] &lt;- NSCLC_pCSC_groups_info_list_tmp</span></code></pre></div>
<p>Set a few more parameters and rename the expression matrix:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tissue_type</span>                                  <span class="op">&lt;&lt;-</span> <span class="st">"NSCLC"</span>
<span class="va">outDir</span>                                       <span class="op">&lt;-</span> <span class="va">prefExprDir</span>
<span class="va">no_ribo</span>                                      <span class="op">&lt;-</span> <span class="cn">TRUE</span>  <span class="co"># set to TRUE to remove ribosomal transcripts from some</span>
                                                      <span class="co"># of the expression bar charts</span>
<span class="va">separate_row_for_high_expression_transcripts</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co"># set to TRUE to print highest expression bars on a</span>
                                                      <span class="co"># seperate row</span>
<span class="va">frag_count_method</span>                            <span class="op">&lt;-</span> <span class="st">'CellsPM'</span> <span class="co"># Using Cells Per Million values in the expression</span>
                                                          <span class="co"># matrix, but just a name, can be anything</span>

<span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="va">tCellsPM_NSCLC_45000</span>
<span class="va">tCellsPM_NSCLC_45000</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></code></pre></div>
<p><strong>We can now print the preferential expression plots.</strong></p>
<p>(If the images generated look small when viewing them in the web pages, you could try opening each of them in a new tab e.g. by using a right mouse click.)</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NSCLC_all_results_ranks_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Preferential expression ranked results stored here</span>

<span class="va">NSCLC_45000_tumour_only_positions</span> <span class="op">&lt;-</span> <span class="va">NSCLC_45000_tumour_pos</span>

<span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>

    <span class="va">this_sample_name</span>     <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_pCSC_group_name</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">this_pCSC_group_name</span>, <span class="st">"; "</span>, <span class="va">this_sample_name</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>

    <span class="co"># Subset transcription info for pCSC group</span>
    <span class="va">pCSC_group_pos</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>    
    <span class="va">pCSC_group_pos</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_patient_tumour_only_positions</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>

    <span class="co"># Generate graphs</span>
    <span class="va">NSCLC_all_results_ranks_list</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/print_preferential_expression_bar_graphs.html">print_preferential_expression_bar_graphs</a></span><span class="op">(</span>
        <span class="va">pCSC_group_name</span>                    <span class="op">&lt;-</span> <span class="va">this_pCSC_group_name</span>,
        <span class="va">pCSC_group_positions</span>               <span class="op">&lt;-</span> <span class="va">pCSC_group_pos</span>,
        <span class="va">HUGO_groups_sets_of_columnNumbers</span>  <span class="op">&lt;-</span> <span class="va">gene_group_positions_to_use</span>,
        <span class="va">tExprMatr</span>                          <span class="op">&lt;-</span> <span class="va">expression_matrix</span>,
        <span class="va">tumour_only_positions</span>              <span class="op">&lt;-</span> <span class="va">NSCLC_45000_tumour_only_positions</span>,
        <span class="va">this_patient_tumour_only_positions</span>,
        <span class="va">outDir</span>,
        <span class="va">tissue_type</span>,
        <span class="va">family_type_name</span>                   <span class="op">&lt;-</span> <span class="st">'HUGO'</span>,
        <span class="va">patient_source_name</span>                <span class="op">&lt;-</span> <span class="va">this_sample_name</span>,
        <span class="va">frag_count_method</span>,
        <span class="va">no_ribo</span>,
        <span class="va">separate_row_for_high_expression_transcripts</span>,
        <span class="va">max_number_of_graph_rows_stacked_transcripts</span> <span class="op">&lt;-</span> <span class="fl">2</span>,
        <span class="va">max_number_of_graph_rows_stacked_families</span> <span class="op">&lt;-</span> <span class="fl">2</span>,
        <span class="va">all_HUGO_abbreviations_and_fullnames</span>,
        <span class="va">dataDir</span>
    <span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">)</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">this_pCSC_group_name</span>
<span class="op">}</span>
<span class="co">#&gt; NSCLC_pCSC_group_1_patient_1_tumour; Patient 1</span>
<span class="co">#&gt; Output directory = 'results/NSCLC_45000/preferential_expression/'</span>
<span class="co">#&gt; Printing: Single transcripts barcharts (triplets)</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-1.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-2.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-3.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-4.png" width="1920"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-5.png" width="1920"></p>
<pre><code>#&gt; Expression bar charts completed
#&gt; 
#&gt; NSCLC_pCSC_group_2_patient_2_tumour; Patient 2
#&gt; Output directory = 'results/NSCLC_45000/preferential_expression/'
#&gt; Printing: Single transcripts barcharts (triplets)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-6.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-7.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-8.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-9.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-10.png" width="1920"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-11.png" width="1920"></p>
<pre><code>#&gt; Expression bar charts completed
#&gt; 
#&gt; NSCLC_pCSC_group_3_patient_3_tumour; Patient 3
#&gt; Output directory = 'results/NSCLC_45000/preferential_expression/'
#&gt; Printing: Single transcripts barcharts (triplets)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-12.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-13.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-14.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-15.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-16.png" width="1920"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-17.png" width="1920"></p>
<pre><code>#&gt; Expression bar charts completed
#&gt; 
#&gt; NSCLC_pCSC_group_4_patient_4_tumour; Patient 4
#&gt; Output directory = 'results/NSCLC_45000/preferential_expression/'
#&gt; Printing: Single transcripts barcharts (triplets)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-18.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-19.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-20.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-21.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-22.png" width="1920"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-23.png" width="1920"></p>
<pre><code>#&gt; Expression bar charts completed
#&gt; 
#&gt; NSCLC_pCSC_group_5_patient_5_tumour; Patient 5
#&gt; Output directory = 'results/NSCLC_45000/preferential_expression/'
#&gt; Printing: Single transcripts barcharts (triplets)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-24.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-25.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-26.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-27.png" width="1920"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-28.png" width="1920"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-36-29.png" width="1920"><img src="tutorial_3_files/figure-html/unnamed-chunk-36-30.png" width="1920"></p>
<pre><code><a href="https://rdrr.io/r/base/save.html">#&gt; Expression bar charts completed
save(NSCLC_all_results_ranks_list, file = paste0(saveDir, "NSCLC_all_results_ranks_list_raw.RData"))
#load(paste0(saveDir, "NSCLC_all_results_ranks_list_raw.RData"))</a></code></pre>
<p>Print cell numbers on the t-SNEs that correspond with the cell labels displayed on the stacked transcript preferential expression bar graphs, so that graph can be related to the t-SNE:</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">this_pCSC_group_name</span> <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_pCSC_group_pos</span>  <span class="op">&lt;-</span> <span class="va">NSCLC_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
    
    <span class="va">this_perplexity_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>
    <span class="va">this_sample_name</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_tSNE_results_with_annotations</span> <span class="op">&lt;-</span>
        <span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span> <span class="co"># taking first from perplexity list</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">this_tSNE_results_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"#"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">this_tSNE_results_with_annotations</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">this_pCSC_group_name</span>, <span class="st">"Total_transcription"</span>,
                <span class="va">this_sample_name</span>, <span class="va">tSNEsDir</span>, <span class="va">this_pCSC_group_pos</span><span class="op">)</span>
    
    <span class="va">outFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, <span class="va">this_pCSC_group_name</span>, <span class="st">'-Total_transcription_Y1vsY2.tif'</span><span class="op">)</span> 
    <span class="va">renamed_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDir</span>, <span class="va">this_pCSC_group_name</span>, <span class="st">"-Total_transcription_Y1vsY2-"</span>,
                           <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span>, <span class="st">"_cells-labelled.tif"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span><span class="va">outFile</span>, <span class="va">renamed_file</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-37-1.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-37-2.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-37-3.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-37-4.png" width="1152"><img src="tutorial_3_files/figure-html/unnamed-chunk-37-5.png" width="1152"></p>
<p><strong>Top ranked preferentially expressed genes and gene groups:</strong></p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\n## Rankings for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">" ##\n\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Top 100 transcripts for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nTop 100 families for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nTop 100 transcripts in ranked families for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">NSCLC_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for NSCLC_pCSC_group_1_patient_1_tumour ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for NSCLC_pCSC_group_1_patient_1_tumour:</span>
<span class="co">#&gt;   [1] "VCX3A"        "CTAG2"        "VCX3B"        "TFDP3"        "VCX"         </span>
<span class="co">#&gt;   [6] "CAPNS2"       "C5orf66-AS1"  "RPL10L"       "FOXL2"        "MID1IP1-AS1" </span>
<span class="co">#&gt;  [11] "MAGEC2"       "CALML3"       "IGKV1D-13"    "PAGE4"        "DCAF4L2"     </span>
<span class="co">#&gt;  [16] "MAGEC1"       "GAGE2A"       "LINC01419"    "WFDC10A"      "HOXD13"      </span>
<span class="co">#&gt;  [21] "IGHV4-61"     "SACS-AS1"     "HOXB1"        "HEPN1"        "TEX13C"      </span>
<span class="co">#&gt;  [26] "PAGE5"        "FAM181B"      "XAGE5"        "HOXD11"       "IFNE"        </span>
<span class="co">#&gt;  [31] "HCG15"        "URB1-AS1"     "KRTAP5-10"    "SOX15"        "IGHV3-48"    </span>
<span class="co">#&gt;  [36] "DANT1"        "KRT13"        "TCEAL6"       "KRT31"        "TOLLIP-AS1"  </span>
<span class="co">#&gt;  [41] "IGLV5-37"     "ZNF32-AS2"    "NTS"          "OR2B6"        "GAGE12J"     </span>
<span class="co">#&gt;  [46] "FAM83C"       "CYP2W1"       "GAGE12H"      "IGHV3-21"     "MRPL53"      </span>
<span class="co">#&gt;  [51] "TGFB2-AS1"    "GNG3"         "TNNI3"        "IGHV4-39"     "IGHV5-51"    </span>
<span class="co">#&gt;  [56] "IGLV3-25"     "DMRTA2"       "KRT5"         "NKX2-8"       "NHLH2"       </span>
<span class="co">#&gt;  [61] "SMIM10L1"     "HSD17B1"      "TP53TG3D"     "NUPR2"        "CMTM5"       </span>
<span class="co">#&gt;  [66] "DNASE1L2"     "BEST2"        "TP53AIP1"     "BEX2"         "CYP4A11"     </span>
<span class="co">#&gt;  [71] "ARHGAP26-AS1" "MIF"          "MAGEB1"       "HAGLROS"      "RPEL1"       </span>
<span class="co">#&gt;  [76] "PSMB1"        "IGHV1-69"     "STATH"        "S100A2"       "BRWD1-AS2"   </span>
<span class="co">#&gt;  [81] "UCN2"         "SLC45A2"      "TCEAL5"       "ADAD2"        "NPPC"        </span>
<span class="co">#&gt;  [86] "IGLV1-47"     "MYOD1"        "IGHV3-23"     "STAR"         "PHOSPHO1"    </span>
<span class="co">#&gt;  [91] "CT55"         "MAGEB2"       "ZNF280A"      "SPRR2A"       "ADH1C"       </span>
<span class="co">#&gt;  [96] "GAST"         "FGFBP1"       "LINC01023"    "USH1G"        "BOLA2B"      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for NSCLC_pCSC_group_1_patient_1_tumour:</span>
<span class="co">#&gt;   [1] "Variable charge X/Y family"                                                  </span>
<span class="co">#&gt;   [2] "Transcription factor Dp family"                                              </span>
<span class="co">#&gt;   [3] "GAGE family"                                                                 </span>
<span class="co">#&gt;   [4] "PAGE family"                                                                 </span>
<span class="co">#&gt;   [5] "XAGE family"                                                                 </span>
<span class="co">#&gt;   [6] "MAGE family"                                                                 </span>
<span class="co">#&gt;   [7] "Transcription elongation factor A like family"                               </span>
<span class="co">#&gt;   [8] "Brain expressed X-linked family"                                             </span>
<span class="co">#&gt;   [9] "WAP four-disulfide core domain containing"                                   </span>
<span class="co">#&gt;  [10] "Bestrophins"                                                                 </span>
<span class="co">#&gt;  [11] "HOXL subclass homeoboxes"                                                    </span>
<span class="co">#&gt;  [12] "Keratins, type I"                                                            </span>
<span class="co">#&gt;  [13] "Immunoglobulin heavy locus at 14q32.33"                                      </span>
<span class="co">#&gt;  [14] "Family with sequence similarity 83 "                                         </span>
<span class="co">#&gt;  [15] "Troponin complex subunits"                                                   </span>
<span class="co">#&gt;  [16] "Immunoglobulin kappa locus at 2p11.2"                                        </span>
<span class="co">#&gt;  [17] "L ribosomal proteins"                                                        </span>
<span class="co">#&gt;  [18] "HIRA histone chaperone complex subunits"                                     </span>
<span class="co">#&gt;  [19] "Forkhead boxes"                                                              </span>
<span class="co">#&gt;  [20] "SRY-box transcription factors"                                               </span>
<span class="co">#&gt;  [21] "Keratin associated proteins"                                                 </span>
<span class="co">#&gt;  [22] "CKLF like MARVEL transmembrane domain containing"                            </span>
<span class="co">#&gt;  [23] "Immunoglobulin lambda locus at 22q11.2"                                      </span>
<span class="co">#&gt;  [24] "Adenosine deaminases acting on RNA"                                          </span>
<span class="co">#&gt;  [25] "Alcohol dehydrogenases"                                                      </span>
<span class="co">#&gt;  [26] "G protein subunits gamma"                                                    </span>
<span class="co">#&gt;  [27] "Cytochrome P450 family 2"                                                    </span>
<span class="co">#&gt;  [28] "IZUMO family"                                                                </span>
<span class="co">#&gt;  [29] "Golgi associated kinase family"                                              </span>
<span class="co">#&gt;  [30] "EF-hand domain containing"                                                   </span>
<span class="co">#&gt;  [31] "WBP1/VOPP1 family"                                                           </span>
<span class="co">#&gt;  [32] "YIP family"                                                                  </span>
<span class="co">#&gt;  [33] "Keratins, type II"                                                           </span>
<span class="co">#&gt;  [34] "S100 calcium binding proteins"                                               </span>
<span class="co">#&gt;  [35] "Gap junction proteins"                                                       </span>
<span class="co">#&gt;  [36] "Neuropeptides"                                                               </span>
<span class="co">#&gt;  [37] "Small nucleolar RNA non-coding host genes"                                   </span>
<span class="co">#&gt;  [38] "Intronic transcripts"                                                        </span>
<span class="co">#&gt;  [39] "Fatty acid desaturases"                                                      </span>
<span class="co">#&gt;  [40] "Neurotrophins"                                                               </span>
<span class="co">#&gt;  [41] "HAD Asp-based non-protein phosphatases"                                      </span>
<span class="co">#&gt;  [42] "Hydroxycarboxylic acid receptors"                                            </span>
<span class="co">#&gt;  [43] "Cytochrome P450 family 4"                                                    </span>
<span class="co">#&gt;  [44] "StAR related lipid transfer domain containing"                               </span>
<span class="co">#&gt;  [45] "GDNF family ligands"                                                         </span>
<span class="co">#&gt;  [46] "CEA cell adhesion molecule family"                                           </span>
<span class="co">#&gt;  [47] "Chloride channel accessory"                                                  </span>
<span class="co">#&gt;  [48] "Zinc fingers HIT-type"                                                       </span>
<span class="co">#&gt;  [49] "Small proline rich proteins"                                                 </span>
<span class="co">#&gt;  [50] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [51] "Potassium two pore domain channel subfamily K "                              </span>
<span class="co">#&gt;  [52] "Acid sensing ion channel subunits"                                           </span>
<span class="co">#&gt;  [53] "Mitochondrial RNase P complex"                                               </span>
<span class="co">#&gt;  [54] "Chromosomal passenger complex"                                               </span>
<span class="co">#&gt;  [55] "Y-box binding proteins"                                                      </span>
<span class="co">#&gt;  [56] "Small heat shock proteins"                                                   </span>
<span class="co">#&gt;  [57] "Beta-gamma crystallins"                                                      </span>
<span class="co">#&gt;  [58] "ER membrane protein complex subunits"                                        </span>
<span class="co">#&gt;  [59] "Coiled-coil-helix-coiled-coil-helix domain containing proteins"              </span>
<span class="co">#&gt;  [60] "TIM22 complex"                                                               </span>
<span class="co">#&gt;  [61] "Leukotriene receptors"                                                       </span>
<span class="co">#&gt;  [62] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [63] "WNK lysine deficient protein kinases"                                        </span>
<span class="co">#&gt;  [64] "Mitochondrial ribosomal proteins"                                            </span>
<span class="co">#&gt;  [65] "Transforming growth factor beta superfamily"                                 </span>
<span class="co">#&gt;  [66] "SET complex"                                                                 </span>
<span class="co">#&gt;  [67] "Cystatins, type 1"                                                           </span>
<span class="co">#&gt;  [68] "Apoptosome"                                                                  </span>
<span class="co">#&gt;  [69] "Helix-turn-helix CENPB type domain containing"                               </span>
<span class="co">#&gt;  [70] "Potassium voltage-gated channel regulatory subunits"                         </span>
<span class="co">#&gt;  [71] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [72] "Zyxin family"                                                                </span>
<span class="co">#&gt;  [73] "Profilins"                                                                   </span>
<span class="co">#&gt;  [74] "PAR bZIP family"                                                             </span>
<span class="co">#&gt;  [75] "Short chain dehydrogenase/reductase superfamily"                             </span>
<span class="co">#&gt;  [76] "Flotillins"                                                                  </span>
<span class="co">#&gt;  [77] "Acyl-CoA thioesterases"                                                      </span>
<span class="co">#&gt;  [78] "VEGF family"                                                                 </span>
<span class="co">#&gt;  [79] "Nudix hydrolase family"                                                      </span>
<span class="co">#&gt;  [80] "Antisense RNAs"                                                              </span>
<span class="co">#&gt;  [81] "U11 small nuclear ribonucleoprotein"                                         </span>
<span class="co">#&gt;  [82] "Long non-coding RNAs with non-systematic symbols"                            </span>
<span class="co">#&gt;  [83] "NKL subclass homeoboxes and pseudogenes"                                     </span>
<span class="co">#&gt;  [84] "5'-nucleotidases"                                                            </span>
<span class="co">#&gt;  [85] "Small subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [86] "Thioesterase superfamily"                                                    </span>
<span class="co">#&gt;  [87] "Diphthamide biosynthesis pathway genes"                                      </span>
<span class="co">#&gt;  [88] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [89] "Dolichyl-phosphate mannosyltransferase subunits"                             </span>
<span class="co">#&gt;  [90] "14-3-3 phospho-serine/phospho-threonine binding proteins"                    </span>
<span class="co">#&gt;  [91] "Gla domain containing"                                                       </span>
<span class="co">#&gt;  [92] "Annexins"                                                                    </span>
<span class="co">#&gt;  [93] "Soluble glutathione S-transferases"                                          </span>
<span class="co">#&gt;  [94] "Serpin peptidase inhibitors"                                                 </span>
<span class="co">#&gt;  [95] "Biogenesis of lysosomal organelles complex 1 subunits"                       </span>
<span class="co">#&gt;  [96] "Guanylate binding proteins"                                                  </span>
<span class="co">#&gt;  [97] "Proline rich and Gla domain containing"                                      </span>
<span class="co">#&gt;  [98] "RISC loading complex"                                                        </span>
<span class="co">#&gt;  [99] "Insulin like growth factor binding proteins"                                 </span>
<span class="co">#&gt; [100] "SF3b complex"                                                                </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for NSCLC_pCSC_group_1_patient_1_tumour:</span>
<span class="co">#&gt;   [1] "TFDP3"     "BEST2"     "WFDC10A"   "XAGE5"     "CABIN1"    "FAM83C"   </span>
<span class="co">#&gt;   [7] "CMTM5"     "ADAD2"     "TNNI3"     "RPL10L"    "FOXL2"     "GNG3"     </span>
<span class="co">#&gt;  [13] "WBP1"      "FJX1"      "IZUMO1"    "ADH1C"     "KRTAP5-10" "SOX15"    </span>
<span class="co">#&gt;  [19] "IGKV1D-13" "CYP2W1"    "WARS2-IT1" "BEX2"      "TCEAL6"    "NTF4"     </span>
<span class="co">#&gt;  [25] "CYP4A11"   "PHOSPHO1"  "GAGE2A"    "PAGE4"     "CEACAM7"   "PRAF2"    </span>
<span class="co">#&gt;  [31] "FADS6"     "KRT5"      "STAR"      "HCAR2"     "CAPNS2"    "ARTN"     </span>
<span class="co">#&gt;  [37] "YBX2"      "VCX3A"     "MAGEC2"    "KCNK7"     "PAGE5"     "WNK4"     </span>
<span class="co">#&gt;  [43] "KRT13"     "KRT31"     "AMH"       "HSPB1"     "HSD17B10"  "VCX3B"    </span>
<span class="co">#&gt;  [49] "CLCA4"     "S100A2"    "CYC1"      "ASIC3"     "MAGEC1"    "BIRC5"    </span>
<span class="co">#&gt;  [55] "KCNE2"     "HOXD13"    "TIMM10"    "HOXB1"     "TRIP6"     "SPRR2A"   </span>
<span class="co">#&gt;  [61] "GJB5"      "TCEAL5"    "INO80B"    "IGLV5-37"  "VCX"       "CSTB"     </span>
<span class="co">#&gt;  [67] "OXER1"     "MRPL53"    "IGHV4-61"  "ALDOA"     "HOXD11"    "DBP"      </span>
<span class="co">#&gt;  [73] "FLOT1"     "CRYBA2"    "ZNHIT2"    "NTS"       "GAGE12J"   "SPRR1B"   </span>
<span class="co">#&gt;  [79] "CLCA2"     "EMC9"      "GAGE12H"   "IGLV3-25"  "PGF"       "CHCHD10"  </span>
<span class="co">#&gt;  [85] "CALML3"    "PFN4"      "GJB6"      "CRYBA4"    "BEX4"      "SNHG19"   </span>
<span class="co">#&gt;  [91] "EMC6"      "APEX1"     "ASIC1"     "SNRNP25"   "TIGD1"     "HCAR3"    </span>
<span class="co">#&gt;  [97] "GJB2"      "KRTAP19-1" "BEX5"      "ALDOC"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for NSCLC_pCSC_group_2_patient_2_tumour ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for NSCLC_pCSC_group_2_patient_2_tumour:</span>
<span class="co">#&gt;   [1] "KRTAP20-2"   "IGHV1-58"    "C6orf15"     "TRBV2"       "UFSP1"      </span>
<span class="co">#&gt;   [6] "KRTAP19-1"   "MAGEE2"      "SLC18A3"     "LGALS7"      "IGHV3-72"   </span>
<span class="co">#&gt;  [11] "PAGE2"       "SPRN"        "IGFL1"       "IGHV3-13"    "CYP26A1"    </span>
<span class="co">#&gt;  [16] "PSMB1"       "LY6G6C"      "IGHV3-49"    "MAGEA1"      "MAGEA12"    </span>
<span class="co">#&gt;  [21] "CALML3"      "SOX2"        "TRAV21"      "IGHV3-33"    "GPHB5"      </span>
<span class="co">#&gt;  [26] "SMIM10L1"    "IGLV1-41"    "SMARCA5-AS1" "IGKV1D-8"    "PTPRCAP"    </span>
<span class="co">#&gt;  [31] "IGHV2-26"    "ARHGEF35"    "SPRR1A"      "IGKV1-16"    "MAGEB2"     </span>
<span class="co">#&gt;  [36] "HOXD4"       "SOST"        "EVX1"        "IGKV3D-20"   "MAGEA2B"    </span>
<span class="co">#&gt;  [41] "KRT74"       "FOXE1"       "IGHV3-21"    "KRT34"       "IGHV2-5"    </span>
<span class="co">#&gt;  [46] "DLX5"        "IGHV1-46"    "IGHV1-18"    "MAGEA9B"     "FZD9"       </span>
<span class="co">#&gt;  [51] "SOX15"       "LINC00626"   "MAGEA6"      "MAGEA9"      "PAGE2B"     </span>
<span class="co">#&gt;  [56] "GAST"        "IGHV4-59"    "ARTN"        "CSAG1"       "IGKV2D-29"  </span>
<span class="co">#&gt;  [61] "RAET1L"      "IGLV2-14"    "DLK2"        "MAGEA3"      "HOXA2"      </span>
<span class="co">#&gt;  [66] "MCF2L-AS1"   "C5orf66-AS1" "ADRA2B"      "HOXD13"      "CORT"       </span>
<span class="co">#&gt;  [71] "AMH"         "TRAV6"       "IGKV3-20"    "IGLV7-46"    "GPRC5D"     </span>
<span class="co">#&gt;  [76] "IGHV3-23"    "RGS9BP"      "MRPL53"      "GPX2"        "GNG8"       </span>
<span class="co">#&gt;  [81] "IGKV1-27"    "SPIN2A"      "TCL1B"       "CBR1"        "IGKV6-21"   </span>
<span class="co">#&gt;  [86] "RPEL1"       "IGLV3-9"     "LINC01063"   "INHBE"       "NR0B1"      </span>
<span class="co">#&gt;  [91] "CCNA1"       "UBL4A"       "IGKV1-9"     "IGLV1-47"    "SRXN1"      </span>
<span class="co">#&gt;  [96] "MAGEF1"      "IGKV4-1"     "IGHV3-53"    "CEL"         "ALDH3A1"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for NSCLC_pCSC_group_2_patient_2_tumour:</span>
<span class="co">#&gt;   [1] "Keratin associated proteins"                                                 </span>
<span class="co">#&gt;   [2] "PAGE family"                                                                 </span>
<span class="co">#&gt;   [3] "Immunoglobulin heavy locus at 14q32.33"                                      </span>
<span class="co">#&gt;   [4] "MAGE family"                                                                 </span>
<span class="co">#&gt;   [5] "Galectins"                                                                   </span>
<span class="co">#&gt;   [6] "Glycoprotein hormone subunits"                                               </span>
<span class="co">#&gt;   [7] "T cell receptor beta locus at 7q34"                                          </span>
<span class="co">#&gt;   [8] "Immunoglobulin kappa locus at 2p11.2"                                        </span>
<span class="co">#&gt;   [9] "GDNF family ligands"                                                         </span>
<span class="co">#&gt;  [10] "DAN family"                                                                  </span>
<span class="co">#&gt;  [11] "Spindlin family"                                                             </span>
<span class="co">#&gt;  [12] "BAG6 complex"                                                                </span>
<span class="co">#&gt;  [13] "GET4-GET5 transmembrane domain recognition complex subunits"                 </span>
<span class="co">#&gt;  [14] "SRY-box transcription factors"                                               </span>
<span class="co">#&gt;  [15] "Small proline rich proteins"                                                 </span>
<span class="co">#&gt;  [16] "Immunoglobulin lambda locus at 22q11.2"                                      </span>
<span class="co">#&gt;  [17] "F2R receptors"                                                               </span>
<span class="co">#&gt;  [18] "HOXL subclass homeoboxes"                                                    </span>
<span class="co">#&gt;  [19] "Pepsinogens"                                                                 </span>
<span class="co">#&gt;  [20] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;  [21] "Inhibin subunits"                                                            </span>
<span class="co">#&gt;  [22] "G protein-coupled receptors, Class F frizzled"                               </span>
<span class="co">#&gt;  [23] "Adrenoceptors"                                                               </span>
<span class="co">#&gt;  [24] "Golgi associated kinase family"                                              </span>
<span class="co">#&gt;  [25] "Mitochondrial RNase P complex"                                               </span>
<span class="co">#&gt;  [26] "Keratins, type II"                                                           </span>
<span class="co">#&gt;  [27] "Keratins, type I"                                                            </span>
<span class="co">#&gt;  [28] "G protein-coupled receptors, Class C orphans"                                </span>
<span class="co">#&gt;  [29] "WBP1/VOPP1 family"                                                           </span>
<span class="co">#&gt;  [30] "Brain expressed X-linked family"                                             </span>
<span class="co">#&gt;  [31] "KEOPS complex"                                                               </span>
<span class="co">#&gt;  [32] "Neurotrophins"                                                               </span>
<span class="co">#&gt;  [33] "Gap junction proteins"                                                       </span>
<span class="co">#&gt;  [34] "LY6/PLAUR domain containing"                                                 </span>
<span class="co">#&gt;  [35] "Zinc fingers HIT-type"                                                       </span>
<span class="co">#&gt;  [36] "G protein subunits gamma"                                                    </span>
<span class="co">#&gt;  [37] "IZUMO family"                                                                </span>
<span class="co">#&gt;  [38] "Selenoproteins"                                                              </span>
<span class="co">#&gt;  [39] "NKL subclass homeoboxes and pseudogenes"                                     </span>
<span class="co">#&gt;  [40] "Transforming growth factor beta superfamily"                                 </span>
<span class="co">#&gt;  [41] "Ephrins"                                                                     </span>
<span class="co">#&gt;  [42] "Proteasome"                                                                  </span>
<span class="co">#&gt;  [43] "POTE ankyrin domain containing"                                              </span>
<span class="co">#&gt;  [44] "SNAG transcriptional repressors"                                             </span>
<span class="co">#&gt;  [45] "Apoptosome"                                                                  </span>
<span class="co">#&gt;  [46] "14-3-3 phospho-serine/phospho-threonine binding proteins"                    </span>
<span class="co">#&gt;  [47] "Zyxin family"                                                                </span>
<span class="co">#&gt;  [48] "Hyalectan proteoglycans"                                                     </span>
<span class="co">#&gt;  [49] "YIP family"                                                                  </span>
<span class="co">#&gt;  [50] "T cell receptor alpha locus at 14q11.2"                                      </span>
<span class="co">#&gt;  [51] "TIM22 complex"                                                               </span>
<span class="co">#&gt;  [52] "Small heat shock proteins"                                                   </span>
<span class="co">#&gt;  [53] "Vestigial like family"                                                       </span>
<span class="co">#&gt;  [54] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;  [55] "Bisphosphoglycerate phosphatases"                                            </span>
<span class="co">#&gt;  [56] "Soluble glutathione S-transferases"                                          </span>
<span class="co">#&gt;  [57] "PARN exonuclease family"                                                     </span>
<span class="co">#&gt;  [58] "Ubiquitin C-terminal hydrolases"                                             </span>
<span class="co">#&gt;  [59] "Biogenesis of lysosomal organelles complex 2 "                               </span>
<span class="co">#&gt;  [60] "Flotillins"                                                                  </span>
<span class="co">#&gt;  [61] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [62] "VEGF family"                                                                 </span>
<span class="co">#&gt;  [63] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [64] "Haloacid dehalogenase like hydrolase domain containing"                      </span>
<span class="co">#&gt;  [65] "COP9 signalosome "                                                           </span>
<span class="co">#&gt;  [66] "Mitochondrial ribosomal proteins"                                            </span>
<span class="co">#&gt;  [67] "Paired boxes"                                                                </span>
<span class="co">#&gt;  [68] "Phosphoglycerate mutases"                                                    </span>
<span class="co">#&gt;  [69] "Peptidyl arginine deiminases"                                                </span>
<span class="co">#&gt;  [70] "Alcohol dehydrogenases"                                                      </span>
<span class="co">#&gt;  [71] "Stomatin family"                                                             </span>
<span class="co">#&gt;  [72] "SF3b complex"                                                                </span>
<span class="co">#&gt;  [73] "Fascin family"                                                               </span>
<span class="co">#&gt;  [74] "HIRA histone chaperone complex subunits"                                     </span>
<span class="co">#&gt;  [75] "Nucleoredoxin family"                                                        </span>
<span class="co">#&gt;  [76] "Lipases"                                                                     </span>
<span class="co">#&gt;  [77] "Zona pellucida glycoproteins"                                                </span>
<span class="co">#&gt;  [78] "Small subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [79] "Aldo-keto reductases"                                                        </span>
<span class="co">#&gt;  [80] "Glycosylphosphatidylinositol transamidase complex"                           </span>
<span class="co">#&gt;  [81] "FKBP prolyl isomerases"                                                      </span>
<span class="co">#&gt;  [82] "Dishevelled segment polarity proteins"                                       </span>
<span class="co">#&gt;  [83] "Neuropeptides"                                                               </span>
<span class="co">#&gt;  [84] "Plakophilins"                                                                </span>
<span class="co">#&gt;  [85] "Aldehyde dehydrogenases"                                                     </span>
<span class="co">#&gt;  [86] "MPP10 subcomplex"                                                            </span>
<span class="co">#&gt;  [87] "ER membrane protein complex subunits"                                        </span>
<span class="co">#&gt;  [88] "Pannexins"                                                                   </span>
<span class="co">#&gt;  [89] "RNA pseudouridylate synthase domain containing"                              </span>
<span class="co">#&gt;  [90] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [91] "Shisa family members"                                                        </span>
<span class="co">#&gt;  [92] "Translocase of outer mitochondrial membrane complex"                         </span>
<span class="co">#&gt;  [93] "Secretoglobins"                                                              </span>
<span class="co">#&gt;  [94] "LIM class homeoboxes"                                                        </span>
<span class="co">#&gt;  [95] "Desmosomal cadherins"                                                        </span>
<span class="co">#&gt;  [96] "ARF GTPase family"                                                           </span>
<span class="co">#&gt;  [97] "Forkhead boxes"                                                              </span>
<span class="co">#&gt;  [98] "Metaxins"                                                                    </span>
<span class="co">#&gt;  [99] "Leukotriene receptors"                                                       </span>
<span class="co">#&gt; [100] "Transcription elongation factor A like family"                               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for NSCLC_pCSC_group_2_patient_2_tumour:</span>
<span class="co">#&gt;   [1] "TRBV2"     "LGALS7"    "SOST"      "UBL4A"     "PGA5"      "KRTAP20-2"</span>
<span class="co">#&gt;   [7] "ARTN"      "GPHB5"     "F2RL2"     "ADRA2B"    "INHBE"     "FJX1"     </span>
<span class="co">#&gt;  [13] "WBP1"      "MT-ATP8"   "GPRC5D"    "PAGE2"     "HSD17B10"  "LY6G6C"   </span>
<span class="co">#&gt;  [19] "FZD9"      "AMH"       "SPRR1A"    "IZUMO1"    "SPIN2A"    "BEX2"     </span>
<span class="co">#&gt;  [25] "LAGE3"     "SNAI2"     "CYC1"      "BCAN"      "IGHV1-58"  "SFN"      </span>
<span class="co">#&gt;  [31] "TRIP6"     "NTF4"      "VGLL1"     "POTEF"     "KRT74"     "SOX2"     </span>
<span class="co">#&gt;  [37] "GPX2"      "GNG8"      "PSMB1"     "PGAM1"     "ZNHIT2"    "HPS6"     </span>
<span class="co">#&gt;  [43] "FLOT1"     "UCHL1"     "KRT34"     "TRAV21"    "PGF"       "PRAF2"    </span>
<span class="co">#&gt;  [49] "SPIN2B"    "PADI3"     "MAGEE2"    "HSPB1"     "DLX5"      "EFNB3"    </span>
<span class="co">#&gt;  [55] "TOE1"      "ADH7"      "STOML2"    "GSTM3"     "CABIN1"    "PAX3"     </span>
<span class="co">#&gt;  [61] "SOX15"     "CEL"       "ALDOC"     "IGLV1-41"  "PAGE2B"    "GJB6"     </span>
<span class="co">#&gt;  [67] "NXN"       "TIMM10B"   "GPAA1"     "ZP1"       "NANP"      "HOXD4"    </span>
<span class="co">#&gt;  [73] "COPS6"     "EVX1"      "SF3B5"     "GNG13"     "KRTAP19-1" "SPRR2A"   </span>
<span class="co">#&gt;  [79] "HDHD3"     "GJB5"      "INO80B"    "KRT5"      "LHB"       "DVL2"     </span>
<span class="co">#&gt;  [85] "TRAV6"     "IGKV1D-8"  "FSCN3"     "TIMM22"    "PNLDC1"    "IGKV1-16" </span>
<span class="co">#&gt;  [91] "EFNA3"     "ALDOA"     "PKP1"      "ALDH3A1"   "IGKV3D-20" "GJA3"     </span>
<span class="co">#&gt;  [97] "TRMT10C"   "FKBPL"     "FSCN1"     "PANX2"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for NSCLC_pCSC_group_3_patient_3_tumour ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for NSCLC_pCSC_group_3_patient_3_tumour:</span>
<span class="co">#&gt;   [1] "KRTAP2-3"     "RMRP"         "SPRR1A"       "DPPA3"        "MT-ATP8"     </span>
<span class="co">#&gt;   [6] "LINC00973"    "PARD3-AS1"    "RNU11"        "MNX1-AS2"     "SNHG25"      </span>
<span class="co">#&gt;  [11] "C19orf33"     "LHB"          "JMJD1C-AS1"   "TGFB2-AS1"    "BCYRN1"      </span>
<span class="co">#&gt;  [16] "CAMK2N2"      "SNHG19"       "FRGCA"        "CAPN15"       "CLDN9"       </span>
<span class="co">#&gt;  [21] "SEMA3B-AS1"   "MT-ND4L"      "LINC01023"    "SLC25A30-AS1" "SPINK14"     </span>
<span class="co">#&gt;  [26] "PHLDA2"       "SNHG9"        "CALML6"       "TMEM249"      "TTC36"       </span>
<span class="co">#&gt;  [31] "SPRR2F"       "CSE1L-AS1"    "MTRNR2L6"     "CAHM"         "OXCT2"       </span>
<span class="co">#&gt;  [36] "LINC01556"    "LINC00449"    "HSPB9"        "REP15"        "INHBC"       </span>
<span class="co">#&gt;  [41] "RPL10L"       "FGFBP1"       "ABHD11-AS1"   "PCAT6"        "MTRNR2L10"   </span>
<span class="co">#&gt;  [46] "FOXE1"        "MT-ND3"       "METTL7B"      "CTXN1"        "PTGER1"      </span>
<span class="co">#&gt;  [51] "TTC9B"        "NKX1-2"       "FOXD2"        "HAGLROS"      "PSMB1"       </span>
<span class="co">#&gt;  [56] "DNASE1L2"     "RPL41"        "S100A6"       "LINC00853"    "ABALON"      </span>
<span class="co">#&gt;  [61] "SFTA2"        "RAB19"        "CST6"         "C2CD4C"       "HCAR1"       </span>
<span class="co">#&gt;  [66] "PACERR"       "LINC00638"    "ERICD"        "DEFB4A"       "CEACAM7"     </span>
<span class="co">#&gt;  [71] "BRWD1-AS2"    "MIF"          "SAPCD1-AS1"   "PYDC1"        "VGF"         </span>
<span class="co">#&gt;  [76] "DPM3"         "SMIM10L1"     "UCN"          "IVL"          "MRPL41"      </span>
<span class="co">#&gt;  [81] "TOLLIP-AS1"   "MT-ND1"       "IGHV1-69-2"   "ID1"          "UNC5B-AS1"   </span>
<span class="co">#&gt;  [86] "TRAV13-2"     "MRPL53"       "NHLRC1"       "AKR7A3"       "CDSN"        </span>
<span class="co">#&gt;  [91] "PLXNB3"       "RPS21"        "ZNF593"       "RHBDL1"       "CALML3-AS1"  </span>
<span class="co">#&gt;  [96] "C6orf226"     "WASIR2"       "C1GALT1C1L"   "FOXQ1"        "NAV2-AS3"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for NSCLC_pCSC_group_3_patient_3_tumour:</span>
<span class="co">#&gt;   [1] "Keratin associated proteins"                                                 </span>
<span class="co">#&gt;   [2] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;   [3] "Small nuclear RNAs"                                                          </span>
<span class="co">#&gt;   [4] "Small proline rich proteins"                                                 </span>
<span class="co">#&gt;   [5] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;   [6] "Small nucleolar RNA non-coding host genes"                                   </span>
<span class="co">#&gt;   [7] "U11/U12 di-snRNP"                                                            </span>
<span class="co">#&gt;   [8] "NADH:ubiquinone oxidoreductase core subunits"                                </span>
<span class="co">#&gt;   [9] "Hydroxycarboxylic acid receptors"                                            </span>
<span class="co">#&gt;  [10] "Alkaline phosphatases"                                                       </span>
<span class="co">#&gt;  [11] "Dolichyl-phosphate mannosyltransferase subunits"                             </span>
<span class="co">#&gt;  [12] "Inhibin subunits"                                                            </span>
<span class="co">#&gt;  [13] "Nucleoredoxin family"                                                        </span>
<span class="co">#&gt;  [14] "Serine peptidase inhibitors, Kazal type"                                     </span>
<span class="co">#&gt;  [15] "Fascin family"                                                               </span>
<span class="co">#&gt;  [16] "Defensins, beta"                                                             </span>
<span class="co">#&gt;  [17] "HIRA histone chaperone complex subunits"                                     </span>
<span class="co">#&gt;  [18] "Calpains"                                                                    </span>
<span class="co">#&gt;  [19] "Claudins"                                                                    </span>
<span class="co">#&gt;  [20] "Mannosidases endo-alpha"                                                     </span>
<span class="co">#&gt;  [21] "Small heat shock proteins"                                                   </span>
<span class="co">#&gt;  [22] "Tissue inhibitor of metallopeptidases"                                       </span>
<span class="co">#&gt;  [23] "Prostaglandin receptors"                                                     </span>
<span class="co">#&gt;  [24] "Zinc fingers RANBP2-type "                                                   </span>
<span class="co">#&gt;  [25] "S100 calcium binding proteins"                                               </span>
<span class="co">#&gt;  [26] "Interferon induced transmembrane proteins"                                   </span>
<span class="co">#&gt;  [27] "Endothelins"                                                                 </span>
<span class="co">#&gt;  [28] "NKL subclass homeoboxes and pseudogenes"                                     </span>
<span class="co">#&gt;  [29] "CCAAT/enhancer binding proteins "                                            </span>
<span class="co">#&gt;  [30] "Pannexins"                                                                   </span>
<span class="co">#&gt;  [31] "CEA cell adhesion molecule family"                                           </span>
<span class="co">#&gt;  [32] "Long pentraxins"                                                             </span>
<span class="co">#&gt;  [33] "NME/NM23 family"                                                             </span>
<span class="co">#&gt;  [34] "7BS small molecule methyltransferases"                                       </span>
<span class="co">#&gt;  [35] "Flotillins"                                                                  </span>
<span class="co">#&gt;  [36] "Granins"                                                                     </span>
<span class="co">#&gt;  [37] "Plexins"                                                                     </span>
<span class="co">#&gt;  [38] "Cystatins, type 2"                                                           </span>
<span class="co">#&gt;  [39] "S ribosomal proteins"                                                        </span>
<span class="co">#&gt;  [40] "L ribosomal proteins"                                                        </span>
<span class="co">#&gt;  [41] "Shisa family members"                                                        </span>
<span class="co">#&gt;  [42] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [43] "Mitochondrial complex IV: cytochrome c oxidase subunits"                     </span>
<span class="co">#&gt;  [44] "Metaxins"                                                                    </span>
<span class="co">#&gt;  [45] "CDC42 effector proteins"                                                     </span>
<span class="co">#&gt;  [46] "Long non-coding RNAs with non-systematic symbols"                            </span>
<span class="co">#&gt;  [47] "Forkhead boxes"                                                              </span>
<span class="co">#&gt;  [48] "Zinc fingers HIT-type"                                                       </span>
<span class="co">#&gt;  [49] "Prolyl 3-hydroxylase family"                                                 </span>
<span class="co">#&gt;  [50] "Jun transcription factor family"                                             </span>
<span class="co">#&gt;  [51] "Transforming growth factor beta superfamily"                                 </span>
<span class="co">#&gt;  [52] "Yippee like family"                                                          </span>
<span class="co">#&gt;  [53] "Gla domain containing"                                                       </span>
<span class="co">#&gt;  [54] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [55] "Polycomb group ring fingers"                                                 </span>
<span class="co">#&gt;  [56] "Cytochrome P450 family 27"                                                   </span>
<span class="co">#&gt;  [57] "TLC domain containing"                                                       </span>
<span class="co">#&gt;  [58] "Troponin complex subunits"                                                   </span>
<span class="co">#&gt;  [59] "Coiled-coil-helix-coiled-coil-helix domain containing proteins"              </span>
<span class="co">#&gt;  [60] "Helix-turn-helix CENPB type domain containing"                               </span>
<span class="co">#&gt;  [61] "Beta 3-glycosyltransferases"                                                 </span>
<span class="co">#&gt;  [62] "Ajuba family"                                                                </span>
<span class="co">#&gt;  [63] "Mitochondrial ribosomal proteins"                                            </span>
<span class="co">#&gt;  [64] "Perilipins"                                                                  </span>
<span class="co">#&gt;  [65] "Rhomboid family"                                                             </span>
<span class="co">#&gt;  [66] "Aldo-keto reductases"                                                        </span>
<span class="co">#&gt;  [67] "Antisense RNAs"                                                              </span>
<span class="co">#&gt;  [68] "Receptor ligands"                                                            </span>
<span class="co">#&gt;  [69] "Small subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [70] "TALE class homeoboxes and pseudogenes"                                       </span>
<span class="co">#&gt;  [71] "LIM class homeoboxes"                                                        </span>
<span class="co">#&gt;  [72] "Metallothioneins"                                                            </span>
<span class="co">#&gt;  [73] "Transcription factor AP-2 family"                                            </span>
<span class="co">#&gt;  [74] "XAGE family"                                                                 </span>
<span class="co">#&gt;  [75] "SINE class homeoboxes"                                                       </span>
<span class="co">#&gt;  [76] "Selenoproteins"                                                              </span>
<span class="co">#&gt;  [77] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [78] "G protein subunits beta"                                                     </span>
<span class="co">#&gt;  [79] "Fatty acid hydroxylase domain containing"                                    </span>
<span class="co">#&gt;  [80] "Oligosaccharyltransferase complex subunits"                                  </span>
<span class="co">#&gt;  [81] "Tubulins"                                                                    </span>
<span class="co">#&gt;  [82] "Tim17 family"                                                                </span>
<span class="co">#&gt;  [83] "Interleukin 6 type cytokine family"                                          </span>
<span class="co">#&gt;  [84] "Pyrin domain containing"                                                     </span>
<span class="co">#&gt;  [85] "SF3b complex"                                                                </span>
<span class="co">#&gt;  [86] "SNAP complex "                                                               </span>
<span class="co">#&gt;  [87] "C2 domain containing"                                                        </span>
<span class="co">#&gt;  [88] "Chloride intracellular channels"                                             </span>
<span class="co">#&gt;  [89] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [90] "GATOR1 subcomplex"                                                           </span>
<span class="co">#&gt;  [91] "Engulfment and cell motility proteins"                                       </span>
<span class="co">#&gt;  [92] "Keratins, type I"                                                            </span>
<span class="co">#&gt;  [93] "CDGSH iron sulfur domain containing"                                         </span>
<span class="co">#&gt;  [94] "Peroxiredoxins"                                                              </span>
<span class="co">#&gt;  [95] "Membrane bound O-acyltransferases"                                           </span>
<span class="co">#&gt;  [96] "DNA transposon derived genes"                                                </span>
<span class="co">#&gt;  [97] "Paraneoplastic Ma antigens"                                                  </span>
<span class="co">#&gt;  [98] "WBP1/VOPP1 family"                                                           </span>
<span class="co">#&gt;  [99] "Arachidonate lipoxygenases"                                                  </span>
<span class="co">#&gt; [100] "Carbonic anhydrases"                                                         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for NSCLC_pCSC_group_3_patient_3_tumour:</span>
<span class="co">#&gt;   [1] "KRTAP2-3" "MT-ATP8"  "RNU11"    "ALPP"     "HCAR1"    "NXN"     </span>
<span class="co">#&gt;   [7] "INHBC"    "FSCN2"    "CABIN1"   "DEFB4A"   "CAPN15"   "MANEAL"  </span>
<span class="co">#&gt;  [13] "DPM3"     "SPRR1A"   "PTGER1"   "TIMP4"    "SPINK14"  "PANX2"   </span>
<span class="co">#&gt;  [19] "HSPB9"    "VGF"      "CST6"     "NPTX1"    "FLOT1"    "PLXNB3"  </span>
<span class="co">#&gt;  [25] "CEACAM7"  "MTX1"     "CEBPD"    "CLDN9"    "IFITM1"   "BGLAP"   </span>
<span class="co">#&gt;  [31] "SNHG25"   "METTL7B"  "EDN1"     "JUND"     "TLCD1"    "SHISA3"  </span>
<span class="co">#&gt;  [37] "CDC42EP5" "MT-ND4L"  "PCGF1"    "NME3"     "CYP27B1"  "EDN2"    </span>
<span class="co">#&gt;  [43] "P3H3"     "AJUBA"    "S100A6"   "TNNC2"    "LEFTY2"   "RHBDL1"  </span>
<span class="co">#&gt;  [49] "AKR7A3"   "YPEL4"    "SNHG19"   "ZNHIT2"   "TFAP2D"   "MT-CYB"  </span>
<span class="co">#&gt;  [55] "PLIN5"    "SHISA4"   "GNB3"     "YPEL3"    "LHB"      "SNHG9"   </span>
<span class="co">#&gt;  [61] "NME2"     "INO80B"   "TIGD3"    "NKX1-2"   "CEBPB"    "MT-CO2"  </span>
<span class="co">#&gt;  [67] "MT-ND3"   "ISL2"     "P3H4"     "FOXE1"    "CHCHD1"   "MT-CO3"  </span>
<span class="co">#&gt;  [73] "XAGE1A"   "CH25H"    "PLIN4"    "FOXD2"    "GAMT"     "PYDC1"   </span>
<span class="co">#&gt;  [79] "ROMO1"    "MT1E"     "CDC42EP1" "RPL10L"   "HSPB1"    "SPINK1"  </span>
<span class="co">#&gt;  [85] "S100A1"   "UQCR10"   "OST4"     "SIX2"     "CLDN8"    "NOP10"   </span>
<span class="co">#&gt;  [91] "RPS21"    "MT-ND1"   "GPX4"     "CLDN3"    "SPRR2F"   "S100A14" </span>
<span class="co">#&gt;  [97] "CYP27C1"  "RPL41"    "XAGE1B"   "CLIC3"   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for NSCLC_pCSC_group_4_patient_4_tumour ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for NSCLC_pCSC_group_4_patient_4_tumour:</span>
<span class="co">#&gt;   [1] "MIR3179-1"   "PLPPR3"      "RNU6ATAC35P" "LINC01574"   "PSMB1"      </span>
<span class="co">#&gt;   [6] "TAL2"        "TAS2R19"     "SMIM10L1"    "LINC01143"   "BCYRN1"     </span>
<span class="co">#&gt;  [11] "NEUROG1"     "CEMP1"       "CST8"        "RPRML"       "SCAANT1"    </span>
<span class="co">#&gt;  [16] "SOX1"        "LINC01109"   "LINC01326"   "HOXC-AS1"    "HELT"       </span>
<span class="co">#&gt;  [21] "KRTAP5-2"    "INSM1"       "TRBV9"       "NEUROG2"     "LY6D"       </span>
<span class="co">#&gt;  [26] "KRT9"        "IFT74-AS1"   "HCG24"       "HES4"        "RPRM"       </span>
<span class="co">#&gt;  [31] "PYDC1"       "C8orf48"     "THRIL"       "SOX3"        "PAGE4"      </span>
<span class="co">#&gt;  [36] "ATOH7"       "CBLN1"       "CALML5"      "OR2B6"       "CCDC87"     </span>
<span class="co">#&gt;  [41] "MAPK4"       "TLX3"        "PI3"         "REP15"       "HOXC13-AS"  </span>
<span class="co">#&gt;  [46] "UBE2E1-AS1"  "DDX39B-AS1"  "MTRNR2L8"    "HMX2"        "LRRC70"     </span>
<span class="co">#&gt;  [51] "LRRC26"      "IGHV2-5"     "GSC"         "MAP3K14"     "IGHV3-11"   </span>
<span class="co">#&gt;  [56] "CA7"         "ORM2"        "ZNF32-AS2"   "LINC01352"   "LINC01063"  </span>
<span class="co">#&gt;  [61] "HOXC5"       "TMEM81"      "DLX1"        "SCGB1B2P"    "HSPB9"      </span>
<span class="co">#&gt;  [66] "LRRC19"      "MYCN"        "CCL27"       "PTTG2"       "HES6"       </span>
<span class="co">#&gt;  [71] "OVOL3"       "PCDHB8"      "TUSC1"       "KRTAP4-1"    "CLDN3"      </span>
<span class="co">#&gt;  [76] "HOXC12"      "FABP7"       "HPDL"        "MFSD6L"      "MNX1-AS2"   </span>
<span class="co">#&gt;  [81] "AQP5"        "UNC5B-AS1"   "NPIPB7"      "GLUD2"       "ORM1"       </span>
<span class="co">#&gt;  [86] "YEATS2-AS1"  "HAGLROS"     "MTRNR2L4"    "HOXC11"      "DUSP23"     </span>
<span class="co">#&gt;  [91] "PCDH8"       "CABIN1"      "TMEM52"      "CSTL1"       "PRR18"      </span>
<span class="co">#&gt;  [96] "RTP3"        "SCARNA2"     "CSNK1A1L"    "IGHV3-48"    "CAPN15"     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for NSCLC_pCSC_group_4_patient_4_tumour:</span>
<span class="co">#&gt;   [1] "Cystatins, type 2"                                                           </span>
<span class="co">#&gt;   [2] "PAGE family"                                                                 </span>
<span class="co">#&gt;   [3] "Taste 2 receptors"                                                           </span>
<span class="co">#&gt;   [4] "SNAG transcriptional repressors"                                             </span>
<span class="co">#&gt;   [5] "SRY-box transcription factors"                                               </span>
<span class="co">#&gt;   [6] "Gag like LTR retrotransposon derived genes"                                  </span>
<span class="co">#&gt;   [7] "Receptor transporter proteins"                                               </span>
<span class="co">#&gt;   [8] "SINE class homeoboxes"                                                       </span>
<span class="co">#&gt;   [9] "HIRA histone chaperone complex subunits"                                     </span>
<span class="co">#&gt;  [10] "Basic helix-loop-helix proteins"                                             </span>
<span class="co">#&gt;  [11] "Y-box binding proteins"                                                      </span>
<span class="co">#&gt;  [12] "Proteasome"                                                                  </span>
<span class="co">#&gt;  [13] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;  [14] "Lipocalins"                                                                  </span>
<span class="co">#&gt;  [15] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;  [16] "Proline rich proteins "                                                      </span>
<span class="co">#&gt;  [17] "Fatty acid binding protein family"                                           </span>
<span class="co">#&gt;  [18] "NKL subclass homeoboxes and pseudogenes"                                     </span>
<span class="co">#&gt;  [19] "Calcium homeostasis modulators"                                              </span>
<span class="co">#&gt;  [20] "Zinc fingers 3CxxC-type"                                                     </span>
<span class="co">#&gt;  [21] "Arylamine N-acetyltransferases"                                              </span>
<span class="co">#&gt;  [22] "Keratins, type I"                                                            </span>
<span class="co">#&gt;  [23] "Inhibin subunits"                                                            </span>
<span class="co">#&gt;  [24] "Mitogen-activated protein kinases"                                           </span>
<span class="co">#&gt;  [25] "Carbonic anhydrases"                                                         </span>
<span class="co">#&gt;  [26] "SF3b complex"                                                                </span>
<span class="co">#&gt;  [27] "Ras related GTP binding proteins"                                            </span>
<span class="co">#&gt;  [28] "Pyrin domain containing"                                                     </span>
<span class="co">#&gt;  [29] "Mitochondrial RNase P complex"                                               </span>
<span class="co">#&gt;  [30] "HOXL subclass homeoboxes"                                                    </span>
<span class="co">#&gt;  [31] "C1q domain containing"                                                       </span>
<span class="co">#&gt;  [32] "Secretoglobins"                                                              </span>
<span class="co">#&gt;  [33] "Adrenoceptors"                                                               </span>
<span class="co">#&gt;  [34] "Long pentraxins"                                                             </span>
<span class="co">#&gt;  [35] "TIM22 complex"                                                               </span>
<span class="co">#&gt;  [36] "KEOPS complex"                                                               </span>
<span class="co">#&gt;  [37] "G protein-coupled receptors, Class F frizzled"                               </span>
<span class="co">#&gt;  [38] "Long non-coding RNAs with non-systematic symbols"                            </span>
<span class="co">#&gt;  [39] "Chromosomal passenger complex"                                               </span>
<span class="co">#&gt;  [40] "SET complex"                                                                 </span>
<span class="co">#&gt;  [41] "Dolichyl-phosphate mannosyltransferase subunits"                             </span>
<span class="co">#&gt;  [42] "PRD class homeoboxes and pseudogenes"                                        </span>
<span class="co">#&gt;  [43] "Arachidonate lipoxygenases"                                                  </span>
<span class="co">#&gt;  [44] "Metallothioneins"                                                            </span>
<span class="co">#&gt;  [45] "BAG6 complex"                                                                </span>
<span class="co">#&gt;  [46] "POU class homeoboxes and pseudogenes"                                        </span>
<span class="co">#&gt;  [47] "NADH:ubiquinone oxidoreductase core subunits"                                </span>
<span class="co">#&gt;  [48] "Zinc fingers HIT-type"                                                       </span>
<span class="co">#&gt;  [49] "WAP four-disulfide core domain containing"                                   </span>
<span class="co">#&gt;  [50] "Clustered protocadherins"                                                    </span>
<span class="co">#&gt;  [51] "MPP10 subcomplex"                                                            </span>
<span class="co">#&gt;  [52] "Pregnancy specific glycoproteins"                                            </span>
<span class="co">#&gt;  [53] "S ribosomal proteins"                                                        </span>
<span class="co">#&gt;  [54] "LY6/PLAUR domain containing"                                                 </span>
<span class="co">#&gt;  [55] "Flotillins"                                                                  </span>
<span class="co">#&gt;  [56] "Helix-turn-helix CENPB type domain containing"                               </span>
<span class="co">#&gt;  [57] "S100 calcium binding proteins"                                               </span>
<span class="co">#&gt;  [58] "Keratin associated proteins"                                                 </span>
<span class="co">#&gt;  [59] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [60] "GET4-GET5 transmembrane domain recognition complex subunits"                 </span>
<span class="co">#&gt;  [61] "Tubulins"                                                                    </span>
<span class="co">#&gt;  [62] "LYR motif containing"                                                        </span>
<span class="co">#&gt;  [63] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [64] "Apoptosome"                                                                  </span>
<span class="co">#&gt;  [65] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [66] "Chromatin accessibility complex"                                             </span>
<span class="co">#&gt;  [67] "Non-clustered protocadherins"                                                </span>
<span class="co">#&gt;  [68] "Glypicans"                                                                   </span>
<span class="co">#&gt;  [69] "Endothelins"                                                                 </span>
<span class="co">#&gt;  [70] "Aquaporins"                                                                  </span>
<span class="co">#&gt;  [71] "Long intergenic non-protein coding RNAs"                                     </span>
<span class="co">#&gt;  [72] "Small heat shock proteins"                                                   </span>
<span class="co">#&gt;  [73] "Defensins, beta"                                                             </span>
<span class="co">#&gt;  [74] "Spectrin repeat containing nuclear envelope family"                          </span>
<span class="co">#&gt;  [75] "KASH domain containing"                                                      </span>
<span class="co">#&gt;  [76] "Heat shock 70kDa proteins"                                                   </span>
<span class="co">#&gt;  [77] "Atypical dual specificity phosphatases"                                      </span>
<span class="co">#&gt;  [78] "MIA family"                                                                  </span>
<span class="co">#&gt;  [79] "Spindlin family"                                                             </span>
<span class="co">#&gt;  [80] "Mitochondrial complex IV: cytochrome c oxidase subunits"                     </span>
<span class="co">#&gt;  [81] "Gla domain containing"                                                       </span>
<span class="co">#&gt;  [82] "U3 small nucleolar ribonucleoprotein"                                        </span>
<span class="co">#&gt;  [83] "Phospholipid phosphatases"                                                   </span>
<span class="co">#&gt;  [84] "Chaperonins"                                                                 </span>
<span class="co">#&gt;  [85] "TIM23 complex"                                                               </span>
<span class="co">#&gt;  [86] "Canonical high mobility group"                                               </span>
<span class="co">#&gt;  [87] "Retrotransposon Gag like"                                                    </span>
<span class="co">#&gt;  [88] "Stomatin family"                                                             </span>
<span class="co">#&gt;  [89] "Ephrins"                                                                     </span>
<span class="co">#&gt;  [90] "Leucine zipper tumor suppressor family"                                      </span>
<span class="co">#&gt;  [91] "Small subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [92] "Forkhead boxes"                                                              </span>
<span class="co">#&gt;  [93] "DNA transposon derived genes"                                                </span>
<span class="co">#&gt;  [94] "BCL2 family"                                                                 </span>
<span class="co">#&gt;  [95] "minichromosome maintenance 2-7 complex"                                      </span>
<span class="co">#&gt;  [96] "Exon junction complex"                                                       </span>
<span class="co">#&gt;  [97] "Thioesterase superfamily"                                                    </span>
<span class="co">#&gt;  [98] "Progestin and adipoQ receptor family"                                        </span>
<span class="co">#&gt;  [99] "Cysteine rich transmembrane BMP regulators"                                  </span>
<span class="co">#&gt; [100] "CREB3 transcription factor family"                                           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for NSCLC_pCSC_group_4_patient_4_tumour:</span>
<span class="co">#&gt;   [1] "CABIN1"   "TAS2R19"  "INSM1"    "PRB3"     "YBX2"     "CST8"    </span>
<span class="co">#&gt;   [7] "ASPRV1"   "NAT2"     "PAGE4"    "CALHM3"   "PSMB1"    "CBLN1"   </span>
<span class="co">#&gt;  [13] "RRAGA"    "RTP3"     "MAPK4"    "PYDC1"    "NPTX2"    "MT-ATP8" </span>
<span class="co">#&gt;  [19] "SF3B5"    "DPM3"     "CA7"      "PI3"      "SCGB1B2P" "ADRA2C"  </span>
<span class="co">#&gt;  [25] "UBL4A"    "KRT9"     "INHBB"    "FLOT1"    "SOX1"     "FABP7"   </span>
<span class="co">#&gt;  [31] "TRMT10C"  "ALOX12"   "SIX3"     "IMP4"     "ORM2"     "FZD8"    </span>
<span class="co">#&gt;  [37] "ALDOC"    "LY6D"     "INHBE"    "HSD17B10" "CYC1"     "PCDH8"   </span>
<span class="co">#&gt;  [43] "CRABP1"   "TIMM10"   "SIX2"     "GPC2"     "AURKB"    "SDHAF1"  </span>
<span class="co">#&gt;  [49] "POU3F2"   "AQP5"     "LAGE3"    "KRTAP5-2" "EDN2"     "GSC"     </span>
<span class="co">#&gt;  [55] "PSG4"     "POLE3"    "MT-ATP6"  "SYNE4"    "RTP4"     "SOX3"    </span>
<span class="co">#&gt;  [61] "ZNHIT2"   "HSPB9"    "PAGE2B"   "APEX1"    "ORM1"     "TLX3"    </span>
<span class="co">#&gt;  [67] "TIMM10B"  "HMGB2"    "DEFB126"  "MIA"      "TAL2"     "ALOX12B" </span>
<span class="co">#&gt;  [73] "SCGB2A1"  "HMX2"     "INO80B"   "MT3"      "CSTL1"    "SPIN4"   </span>
<span class="co">#&gt;  [79] "DUSP23"   "TP53RK"   "MT-ND6"   "KRT39"    "OBP2B"    "CA9"     </span>
<span class="co">#&gt;  [85] "ADRA2B"   "S100A7"   "FZD7"     "BGLAP"    "DLX1"     "NEUROG1" </span>
<span class="co">#&gt;  [91] "BCYRN1"   "ARC"      "MT1E"     "NHP2"     "PLPP6"    "TPRKB"   </span>
<span class="co">#&gt;  [97] "PEG10"    "STOML2"   "CDCA8"    "HOXC5"   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for NSCLC_pCSC_group_5_patient_5_tumour ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for NSCLC_pCSC_group_5_patient_5_tumour:</span>
<span class="co">#&gt;   [1] "TRBV6-1"    "SPATA25"    "SCGB1C2"    "MCCC1-AS1"  "OR51I1"    </span>
<span class="co">#&gt;   [6] "C11orf94"   "SPANXC"     "CRCT1"      "KRTAP4-1"   "SLC25A2"   </span>
<span class="co">#&gt;  [11] "RXFP4"      "PF4V1"      "RNF186"     "HSPB9"      "GPR52"     </span>
<span class="co">#&gt;  [16] "C7orf65"    "SPANXB1"    "PARD3-AS1"  "SETSIP"     "RNU12"     </span>
<span class="co">#&gt;  [21] "GPHA2"      "MAGEA2B"    "TRAV12-1"   "ERVV-1"     "KLLN"      </span>
<span class="co">#&gt;  [26] "PHKA2-AS1"  "IGHV3-11"   "PAPPA-AS1"  "ARL14"      "C8G"       </span>
<span class="co">#&gt;  [31] "CAPN15"     "MT-ATP8"    "TRGV8"      "MAP3K14"    "LINC01589" </span>
<span class="co">#&gt;  [36] "PSMB1"      "SEMA3B-AS1" "ZFP57"      "PPP1R35"    "LINC00543" </span>
<span class="co">#&gt;  [41] "C20orf141"  "SFTA2"      "RNF208"     "GLTPD2"     "C11orf86"  </span>
<span class="co">#&gt;  [46] "RMRP"       "SCGB1B2P"   "BCYRN1"     "IGHV1-69"   "LINC00853" </span>
<span class="co">#&gt;  [51] "LY6D"       "KHDC1L"     "UCN2"       "CT83"       "RPL10L"    </span>
<span class="co">#&gt;  [56] "FRGCA"      "HNRNPCL3"   "CORT"       "DPM3"       "IGHV3-73"  </span>
<span class="co">#&gt;  [61] "RNF224"     "MNX1-AS2"   "SLC35G5"    "SFN"        "GATA3-AS1" </span>
<span class="co">#&gt;  [66] "ARHGDIG"    "URB1-AS1"   "TSSK4"      "LINC01311"  "CCDC87"    </span>
<span class="co">#&gt;  [71] "NDN"        "CAHM"       "FOXH1"      "DANT1"      "C19orf33"  </span>
<span class="co">#&gt;  [76] "CALML5"     "MT-ND4L"    "LGALS7B"    "RHOV"       "TMEM141"   </span>
<span class="co">#&gt;  [81] "RAB40AL"    "SMIM10L1"   "PDZD3"      "TSPO2"      "IGKV1-27"  </span>
<span class="co">#&gt;  [86] "DDX53"      "SIX3-AS1"   "NUPR2"      "EHMT2-AS1"  "SATB2-AS1" </span>
<span class="co">#&gt;  [91] "MT-CO2"     "DEFB126"    "LINC01023"  "SNHG25"     "TMEM81"    </span>
<span class="co">#&gt;  [96] "ABHD11-AS1" "CRIM1"      "FRG2C"      "HOXA4"      "PCAT6"     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for NSCLC_pCSC_group_5_patient_5_tumour:</span>
<span class="co">#&gt;   [1] "SPANX family"                                                                </span>
<span class="co">#&gt;   [2] "Relaxin family peptide receptors"                                            </span>
<span class="co">#&gt;   [3] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;   [4] "Secretoglobins"                                                              </span>
<span class="co">#&gt;   [5] "Envelope ERV derived genes"                                                  </span>
<span class="co">#&gt;   [6] "Glycoprotein hormone subunits"                                               </span>
<span class="co">#&gt;   [7] "Dolichyl-phosphate mannosyltransferase subunits"                             </span>
<span class="co">#&gt;   [8] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;   [9] "Cystatins, type 4"                                                           </span>
<span class="co">#&gt;  [10] "S100 fused type protein family"                                              </span>
<span class="co">#&gt;  [11] "PTK6 family tyrosine kinases"                                                </span>
<span class="co">#&gt;  [12] "Small heat shock proteins"                                                   </span>
<span class="co">#&gt;  [13] "Histatins and statherin"                                                     </span>
<span class="co">#&gt;  [14] "T cell receptor beta locus at 7q34"                                          </span>
<span class="co">#&gt;  [15] "Protein phosphatases"                                                        </span>
<span class="co">#&gt;  [16] "NADH:ubiquinone oxidoreductase core subunits"                                </span>
<span class="co">#&gt;  [17] "Cysteine rich transmembrane BMP regulators"                                  </span>
<span class="co">#&gt;  [18] "Defensins, beta"                                                             </span>
<span class="co">#&gt;  [19] "Lipocalins"                                                                  </span>
<span class="co">#&gt;  [20] "G protein-coupled receptors, Class F frizzled"                               </span>
<span class="co">#&gt;  [21] "Taste 1 receptors"                                                           </span>
<span class="co">#&gt;  [22] "MAGE family"                                                                 </span>
<span class="co">#&gt;  [23] "Serine peptidase inhibitors, Kazal type"                                     </span>
<span class="co">#&gt;  [24] "Sp transcription factors"                                                    </span>
<span class="co">#&gt;  [25] "U11/U12 di-snRNP"                                                            </span>
<span class="co">#&gt;  [26] "Neuropeptide receptors"                                                      </span>
<span class="co">#&gt;  [27] "14-3-3 phospho-serine/phospho-threonine binding proteins"                    </span>
<span class="co">#&gt;  [28] "T cell receptor gamma locus at 7p14"                                         </span>
<span class="co">#&gt;  [29] "Alkaline phosphatases"                                                       </span>
<span class="co">#&gt;  [30] "Speedy/RINGO cell cycle regulator family"                                    </span>
<span class="co">#&gt;  [31] "IZUMO family"                                                                </span>
<span class="co">#&gt;  [32] "Gap junction proteins"                                                       </span>
<span class="co">#&gt;  [33] "CCAAT/enhancer binding proteins "                                            </span>
<span class="co">#&gt;  [34] "Intermediate filaments Type IV"                                              </span>
<span class="co">#&gt;  [35] "Synaptogyrins"                                                               </span>
<span class="co">#&gt;  [36] "SIBLING family"                                                              </span>
<span class="co">#&gt;  [37] "Pannexins"                                                                   </span>
<span class="co">#&gt;  [38] "S100 calcium binding proteins"                                               </span>
<span class="co">#&gt;  [39] "Mitochondrial complex IV: cytochrome c oxidase subunits"                     </span>
<span class="co">#&gt;  [40] "Cystatins, type 1"                                                           </span>
<span class="co">#&gt;  [41] "Kallikreins"                                                                 </span>
<span class="co">#&gt;  [42] "Glucagon receptor family"                                                    </span>
<span class="co">#&gt;  [43] "Small nucleolar RNA non-coding host genes"                                   </span>
<span class="co">#&gt;  [44] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [45] "Selenoproteins"                                                              </span>
<span class="co">#&gt;  [46] "HOXL subclass homeoboxes"                                                    </span>
<span class="co">#&gt;  [47] "Cytochrome P450 family 4"                                                    </span>
<span class="co">#&gt;  [48] "Rho family GTPases"                                                          </span>
<span class="co">#&gt;  [49] "Calpains"                                                                    </span>
<span class="co">#&gt;  [50] "Nucleoredoxin family"                                                        </span>
<span class="co">#&gt;  [51] "Proline rich proteins "                                                      </span>
<span class="co">#&gt;  [52] "LY6/PLAUR domain containing"                                                 </span>
<span class="co">#&gt;  [53] "KEOPS complex"                                                               </span>
<span class="co">#&gt;  [54] "Ubiquilin family"                                                            </span>
<span class="co">#&gt;  [55] "Gla domain containing"                                                       </span>
<span class="co">#&gt;  [56] "MARVEL domain containing"                                                    </span>
<span class="co">#&gt;  [57] "EMI domain containing"                                                       </span>
<span class="co">#&gt;  [58] "ARF GTPase family"                                                           </span>
<span class="co">#&gt;  [59] "Intronic transcripts"                                                        </span>
<span class="co">#&gt;  [60] "Biogenesis of lysosomal organelles complex 1 subunits"                       </span>
<span class="co">#&gt;  [61] "Zinc fingers RANBP2-type "                                                   </span>
<span class="co">#&gt;  [62] "Insulin like growth factor binding proteins"                                 </span>
<span class="co">#&gt;  [63] "F2R receptors"                                                               </span>
<span class="co">#&gt;  [64] "ER membrane protein complex subunits"                                        </span>
<span class="co">#&gt;  [65] "Claudins"                                                                    </span>
<span class="co">#&gt;  [66] "Neuropeptides"                                                               </span>
<span class="co">#&gt;  [67] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [68] "Serine proteases"                                                            </span>
<span class="co">#&gt;  [69] "Hemoglobin subunits"                                                         </span>
<span class="co">#&gt;  [70] "Ras association domain family"                                               </span>
<span class="co">#&gt;  [71] "Immunoglobulin heavy locus at 14q32.33"                                      </span>
<span class="co">#&gt;  [72] "BAG6 complex"                                                                </span>
<span class="co">#&gt;  [73] "Prostaglandin receptors"                                                     </span>
<span class="co">#&gt;  [74] "Peptide receptors"                                                           </span>
<span class="co">#&gt;  [75] "TLC domain containing"                                                       </span>
<span class="co">#&gt;  [76] "LEM domain containing"                                                       </span>
<span class="co">#&gt;  [77] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [78] "Sulfotransferases, cytosolic"                                                </span>
<span class="co">#&gt;  [79] "Coiled-coil-helix-coiled-coil-helix domain containing proteins"              </span>
<span class="co">#&gt;  [80] "Beta 3-glycosyltransferases"                                                 </span>
<span class="co">#&gt;  [81] "Homocysteine methyltransferases"                                             </span>
<span class="co">#&gt;  [82] "Antisense RNAs"                                                              </span>
<span class="co">#&gt;  [83] "Granins"                                                                     </span>
<span class="co">#&gt;  [84] "XAGE family"                                                                 </span>
<span class="co">#&gt;  [85] "Torsins"                                                                     </span>
<span class="co">#&gt;  [86] "Galectins"                                                                   </span>
<span class="co">#&gt;  [87] "Zinc fingers HIT-type"                                                       </span>
<span class="co">#&gt;  [88] "Ectonucleoside triphosphate diphosphohydrolase family"                       </span>
<span class="co">#&gt;  [89] "HIRA histone chaperone complex subunits"                                     </span>
<span class="co">#&gt;  [90] "GET4-GET5 transmembrane domain recognition complex subunits"                 </span>
<span class="co">#&gt;  [91] "L ribosomal proteins"                                                        </span>
<span class="co">#&gt;  [92] "G protein-coupled receptors, Class A orphans"                                </span>
<span class="co">#&gt;  [93] "Flotillins"                                                                  </span>
<span class="co">#&gt;  [94] "Receptor ligands"                                                            </span>
<span class="co">#&gt;  [95] "Tim17 family"                                                                </span>
<span class="co">#&gt;  [96] "YIP family"                                                                  </span>
<span class="co">#&gt;  [97] "Neuraminidases"                                                              </span>
<span class="co">#&gt;  [98] "Mitogen-activated protein kinase kinase kinases"                             </span>
<span class="co">#&gt;  [99] "Golgi associated kinase family"                                              </span>
<span class="co">#&gt; [100] "Mucins"                                                                      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for NSCLC_pCSC_group_5_patient_5_tumour:</span>
<span class="co">#&gt;   [1] "RXFP4"     "ERVV-1"    "TRBV6-1"   "PHPT1"     "CRIM1"     "SCGB1C2"  </span>
<span class="co">#&gt;   [7] "HSPB9"     "DPM3"      "TAS1R3"    "MT-ATP8"   "TRGV8"     "ALPP"     </span>
<span class="co">#&gt;  [13] "DEFB126"   "HTN1"      "SPDYC"     "SFN"       "GPHA2"     "SRMS"     </span>
<span class="co">#&gt;  [19] "RNU12"     "NEFM"      "NPY5R"     "SPANXC"    "FLG2"      "PANX2"    </span>
<span class="co">#&gt;  [25] "CSTB"      "PRB3"      "GCGR"      "CAPN15"    "AHSG"      "SYNGR4"   </span>
<span class="co">#&gt;  [31] "NXN"       "SPP1"      "FETUB"     "EMILIN3"   "CEBPD"     "LAGE3"    </span>
<span class="co">#&gt;  [37] "C8G"       "IZUMO2"    "SPANXB1"   "IZUMO4"    "BGLAP"     "SPINK7"   </span>
<span class="co">#&gt;  [43] "UBQLNL"    "FZD9"      "HBQ1"      "MAGEA2B"   "NOP10"     "SP5"      </span>
<span class="co">#&gt;  [49] "SP9"       "MT-ATP6"   "ARL14"     "LHB"       "MYADML2"   "KISS1R"   </span>
<span class="co">#&gt;  [55] "FZD10"     "PTGER1"    "GPX2"      "F2RL3"     "CYP4F2"    "UBL4A"    </span>
<span class="co">#&gt;  [61] "RHOV"      "MT-ND4L"   "SPINK14"   "SND1-IT1"  "PTK6"      "EMC6"     </span>
<span class="co">#&gt;  [67] "BLOC1S4"   "MT-CO2"    "GJB4"      "SNHG25"    "SP8"       "TLCD1"    </span>
<span class="co">#&gt;  [73] "BACH1-IT2" "DPM2"      "LY6D"      "SCGB1B2P"  "SYNGR2"    "NHP2"     </span>
<span class="co">#&gt;  [79] "IGFBP1"    "GJB5"      "CEBPB"     "STATH"     "MT-ND3"    "FLG"      </span>
<span class="co">#&gt;  [85] "S100P"     "IBSP"      "RASSF7"    "NDN"       "LGALS7B"   "RASSF10"  </span>
<span class="co">#&gt;  [91] "BHMT2"     "MT-ND6"    "IGFBP3"    "SNHG19"    "KLK8"      "GJB3"     </span>
<span class="co">#&gt;  [97] "EMD"       "SNAPIN"    "RAC3"      "S100A14"</span></code></pre></div>
<hr>
</div>
<div id="step-6-t-snes-for-the-highest-ranked-preferentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#step-6-t-snes-for-the-highest-ranked-preferentially-expressed-genes" class="anchor"></a>Step 6: t-SNEs for the Highest Ranked Preferentially Expressed Genes</h3>
<p>Finally, let’s look at the expression patterns of the most preferentially expressed genes, in the t-SNE.</p>
<p>Here are a few examples taken from the top ranked preferentially expressed genes and gene groups, but there are many more of interest:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
    <span class="st">'VCX3A'</span>, <span class="st">'VCX3B'</span>, <span class="st">'VCX'</span>, <span class="st">'GAGE2A'</span>, <span class="st">'PAGE4'</span>, <span class="co"># patient 1</span>
    <span class="st">'PAGE2'</span>, <span class="st">'PAGE2B'</span>, <span class="st">'PAGE1'</span>,   <span class="co"># patient 2</span>
    <span class="st">'INHBC'</span>,                      <span class="co"># patient 3</span>
   <span class="co">#'PAGE4', 'PAGE2B',            # patient 4</span>
    <span class="st">'SIX3'</span>,                       <span class="co"># patient 4</span>
    <span class="st">'SPANXC'</span>, <span class="st">'SPANXB1'</span>, <span class="st">'ERVV-1'</span> <span class="co"># patient 5</span>

<span class="op">)</span>
<span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>, <span class="va">NSCLC_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                            <span class="va">tCounts_NSCLC_45000</span>, <span class="va">run_name</span>, <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDir</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">0.5</span>,
                            <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="co">#&gt;   VCX3A [variable charge X-linked 3A]</span></code></pre></div>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-1.png" width="1152"></p>
<pre><code>#&gt;   VCX3B [variable charge X-linked 3B]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-2.png" width="1152"></p>
<pre><code>#&gt;   VCX [variable charge X-linked]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-3.png" width="1152"></p>
<pre><code>#&gt;   GAGE2A [G antigen 2A]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-4.png" width="1152"></p>
<pre><code>#&gt;   PAGE4 [PAGE family member 4]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-5.png" width="1152"></p>
<pre><code>#&gt;   PAGE2 [PAGE family member 2]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-6.png" width="1152"></p>
<pre><code>#&gt;   PAGE2B [PAGE family member 2B]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-7.png" width="1152"></p>
<pre><code>#&gt;   PAGE1 [PAGE family member 1]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-8.png" width="1152"></p>
<pre><code>#&gt;   INHBC [inhibin subunit beta C]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-9.png" width="1152"></p>
<pre><code>#&gt;   SIX3 [SIX homeobox 3]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-10.png" width="1152"></p>
<pre><code>#&gt;   SPANXC [SPANX family member C]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-11.png" width="1152"></p>
<pre><code>#&gt;   SPANXB1 [SPANX family member B1]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-12.png" width="1152"></p>
<pre><code>#&gt;   ERVV-1 [endogenous retrovirus group V member 1, envelope]</code></pre>
<p><img src="tutorial_3_files/figure-html/unnamed-chunk-39-13.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>Well done, if you got this far. This is the end of the cscplots tutorials.</p>
<p>If you have questions or comments, please feel free to contact me at <a href="mailto:phil.b.carter@gmail.com" class="email">phil.b.carter@gmail.com</a>.</p>
<p>Good luck with your single-cell cancer research!</p>
<hr>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Phil Carter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
