<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CSCPlots Tutorial 1 • CSCPlots</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CSCPlots Tutorial 1">
<meta property="og:description" content="cscplots">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CSCPlots</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial_1.html">CSCPlots Tutorial 1</a>
    </li>
    <li>
      <a href="../articles/tutorial_2.html">CSCPlots Tutorial 2</a>
    </li>
    <li>
      <a href="../articles/tutorial_3.html">CSCPlots Tutorial 3</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tutorial_1_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CSCPlots Tutorial 1</h1>
            
      
      
      <div class="hidden name"><code>tutorial_1.Rmd</code></div>

    </div>

    
    
<div id="tutorial-1-identifying-and-characterising-cancer-stem-cells-in-sclc-cdx-single-cell-rna-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#tutorial-1-identifying-and-characterising-cancer-stem-cells-in-sclc-cdx-single-cell-rna-seq" class="anchor"></a><strong>Tutorial 1: Identifying and Characterising Cancer Stem Cells In SCLC CDX Single-Cell RNA-Seq</strong>
</h3>
<p>CSCPlots is a collection of functions that can be combined to create a pipeline, to achieve the objective of finding and characterising putative cancer stem cells.</p>
<p>Results from applying this method to scRNA-Seq from several cancer types, will be published in the literature later this year. The tutorials shown here replicate some of these results, while trying to familiarise the reader with this pipeline, so that you may generate similar results with your own scRNA-Seq data.</p>
<p>During the construction and use of the pipeline, we will use t-SNE plots to identify high transcription cells with predicted high stemmess compared to the other cells from their cell population. This will require prediction of stemness in each cancer cell, using several similar machine learning methods - different implementations of random forests. We’ll use this information to identify a subset of cells as putative stem cells, and then analyse them for preferential expression, relative to the other cells in the cell population they are selected from. This will aim to identifty genes that are driving these cells, which are in turn driving the tumour they are part of. In addition, understanding which genes are elevated and suppressed in cancer stem cells (CSCs), can reveal how the tumour is resisting drug treatments, and achieving immune evasion, by giving CSCs exceptionally hardy molecular characteristics.</p>
<p>Tutorial 1 shows how to apply this methodology to SCLC CDX scRNA-Seq expression matrices published by Byers et al. (Nature Cancer 2020). This example uses only raw read counts, and we’ll focus on just two of the samples from this dataset, for brevity. A CDX is a tumour produced from patient circulating tumour cells (CTCs).</p>
<p>Tutorial 2 will use a similar pipeline, this time applied to melanoma data published by Tirosh et al. (Science 2016). This is a more complex dataset in that the 4558 cells in this population are of different types, i.e. malginant cancer cells, immune cells, fibroblasts, etc., and is sourced from 19 patients. We’ll focus on CSCs from eight of these patients’ tumours.</p>
<p>OK, let’s get started!</p>
</div>
<div id="step-1-download-sclc-cdx-scrna-seq-expression-matrices" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-download-sclc-cdx-scrna-seq-expression-matrices" class="anchor"></a>Step 1: Download SCLC CDX scRNA-Seq Expression Matrices</h3>
<p>Firstly, prepare the CDX expression data included data with this package.</p>
<p>The two samples they’re from, are taken from tumours created from CTCs, from the same SCLC patient.</p>
<p>GSM4104152_SC4.LB17009 is from a CDX created before the patient was treated with Talazoparib; GSM4104161_SC4_Talazoparib.LB17011 is from another CDX generated after this patient received clinical treatment with Talazoparib. (Talazoparib is a PARP inhibitor that interferes with DNA repair in the cancer cells.)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://Matrix.R-forge.R-project.org/">"Matrix"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/r-gregmisc/gtools">"gtools"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jkrijthe/Rtsne">"Rtsne"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/topepo/caret/">"caret"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/imbs-hl/ranger">"ranger"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"e1071"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">"randomForest"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org">"ggplot2"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/slowkow/ggrepel">"ggrepel"</a></span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>ggrepel.max.overlaps <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span>   
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"cscplots"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">project_name</span>   <span class="op">&lt;-</span> <span class="st">"SCLC_CDXs_project"</span> 
<span class="va">wd</span>             <span class="op">&lt;-</span> <span class="st">'./'</span> <span class="co"># set the working directory</span>
<span class="va">saveDir</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">wd</span>, <span class="st">'save/'</span><span class="op">)</span>
<span class="va">resultsBaseDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'results/'</span><span class="op">)</span>
<span class="va">resultsDir</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'results/'</span>, <span class="va">project_name</span>, <span class="st">'/'</span><span class="op">)</span>
<span class="va">dataDir</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">wd</span>, <span class="st">'../data/'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">saveDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">resultsBaseDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">resultsDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">dataDir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">CDX_sample_IDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
    <span class="st">'GSM4104161_SC4_Talazoparib.LB17011'</span>,
    <span class="st">'GSM4104152_SC4.LB17009'</span>
<span class="op">)</span>
<span class="va">number_of_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_sample_IDs</span><span class="op">)</span>

<span class="va">resultsDirs</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">tSNEsDirs</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">pCSCsDirs</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">prefExprDirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_sample_IDs</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">resultsDirs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="va">CDX_sample_IDs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">'/'</span><span class="op">)</span>
    <span class="va">tSNEsDirs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="va">CDX_sample_IDs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">'/tSNEs/'</span><span class="op">)</span>
    <span class="va">pCSCsDirs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="va">CDX_sample_IDs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">'/tSNEs/pCSC_group_selection/'</span><span class="op">)</span>
    <span class="va">prefExprDirs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">resultsDir</span>, <span class="va">CDX_sample_IDs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">'/preferential_expression/'</span><span class="op">)</span>
    
    <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">resultsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">pCSCsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">prefExprDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>The CDX data we’ll work with is available in this package, but if you wanted to download it from the NIH, you could uncomment the following code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#CDX_urls &lt;- c(</span>
<span class="co">#'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4104nnn/GSM4104161/suppl/GSM4104161_SC4_Talazoparib.LB17011.matrix.tar.gz',</span>
<span class="co">#'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4104nnn/GSM4104152/suppl/GSM4104152_SC4.LB17009.matrix.tar.gz')</span>
<span class="co">#        </span>
<span class="co">#for(i in 1:length(CDX_sample_IDs)){</span>
<span class="co">#    sample_ID &lt;- CDX_sample_IDs[[i]]</span>
<span class="co">#    #cat(paste0(sample_ID, "\n"))</span>
<span class="co">#    gz_file_name &lt;- paste0(sample_ID, '.matrix.tar.gz') </span>
<span class="co">#    gz_path   &lt;- paste0(wd, sample_ID, '/', gz_file_name)</span>
<span class="co">#    dir.create(file.path(wd, sample_ID), showWarnings = FALSE)</span>
<span class="co">#    download.file(CDX_urls[[i]], gz_path) # TEMP TURN OFF</span>
<span class="co">#    untar(gz_path, list=TRUE)</span>
<span class="co">#    untar(gz_path, exdir=paste0(wd, sample_ID))</span>
<span class="co">#    #unlink(gz_file_name)</span>
<span class="co">#}</span>
<span class="co">#</span>
<span class="co">#number_of_samples &lt;- length(CDX_sample_IDs)</span>
<span class="co">#CDX_tRawCounts_list &lt;- list()</span>
<span class="co">#</span>
<span class="co">#for(i in 1:number_of_samples){</span>
<span class="co">#    this_sample_ID &lt;- CDX_sample_IDs[[i]]</span>
<span class="co">#    matrix_dir &lt;- paste0(wd, this_sample_ID, '/')</span>
<span class="co">#    barcode_path  &lt;- paste0(matrix_dir, "barcodes.tsv")</span>
<span class="co">#    gene_path     &lt;- paste0(matrix_dir, "genes.tsv")</span>
<span class="co">#    matrix_path   &lt;- paste0(matrix_dir, "matrix.mtx")</span>
<span class="co">#    this_matrix &lt;- readMM(file = matrix_path)</span>
<span class="co">#    gene_names = read.delim(gene_path, </span>
<span class="co">#                            header = FALSE,</span>
<span class="co">#                            stringsAsFactors = FALSE)</span>
<span class="co">#    barcode_names = read.delim(barcode_path, </span>
<span class="co">#                               header = FALSE,</span>
<span class="co">#                               stringsAsFactors = FALSE)</span>
<span class="co">#    colnames(this_matrix) = barcode_names[[1]]</span>
<span class="co">#    rownames(this_matrix) = gene_names[[2]]</span>
<span class="co">#    this_matrix_no_duplicates &lt;- this_matrix[which(!duplicated(rownames(this_matrix))),]</span>
<span class="co">#    t_matrix &lt;- t(this_matrix_no_duplicates)</span>
<span class="co">#    tRawCounts &lt;- as.matrix(t_matrix)</span>
<span class="co">#    tRawCounts_reorder_pos &lt;- order(colnames(tRawCounts))</span>
<span class="co">#    tRawCounts_reordered &lt;- tRawCounts[,tRawCounts_reorder_pos]</span>
<span class="co">#    this_tRawCounts &lt;- tRawCounts_reordered</span>
<span class="co">#    CDX_tRawCounts_list[[i]] &lt;- this_tRawCounts</span>
<span class="co">#    names(CDX_tRawCounts_list)[[i]] &lt;- this_sample_ID</span>
<span class="co">#}</span>
<span class="co">#save(CDX_tRawCounts_list, file = paste0(saveDir, "CDX_tRawCounts_list.RData"))</span>
<span class="co">##load(paste0(saveDir, "CDX_tRawCounts_list.RData"))</span></code></pre></div>
<p>We’ll use the data already in the package though (which is the same):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_tRawCounts_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># data automatically available in this R package:</span>
<span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_GSM4104161_SC4_Talazoparib.LB17011</span> 
<span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_GSM4104152_SC4.LB17009</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">CDX_sample_IDs</span></code></pre></div>
<hr>
</div>
<div id="step-2-t-sne-generation" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-t-sne-generation" class="anchor"></a>Step 2: t-SNE Generation</h3>
<p>Here, we produce t-SNEs using the SCLC CDX scRNA-Seq matrices, just stored in CDX_tRawCounts_list.</p>
<p>The recommended perplexity value used for the t-SNE is 70, but you can use another value, or a list of values, if you prefer.</p>
<p>This may take some time to run (hours).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_tSNE_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">#perplexityValues &lt;&lt;- c(5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130) # Full range</span>
<span class="co">#perplexityValues &lt;&lt;- c(30,70)</span>
<span class="va">perplexityValues</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">70</span><span class="op">)</span> <span class="co"># Recommended</span>
<span class="va">perplexityValuesPrint</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_zeros_to_perplexity_values.html">add_zeros_to_perplexity_values</a></span><span class="op">(</span><span class="va">perplexityValues</span><span class="op">)</span>
<span class="va">tSNE_run_names</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"tSNE_"</span>, <span class="va">perplexityValuesPrint</span><span class="op">)</span>

<span class="va">CDX_tRawCounts_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_GSM4104152_SC4.LB17009</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">CDX_sample_IDs</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="co"># for reproducibility</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">number_of_samples</span><span class="op">)</span><span class="op">{</span>
    <span class="va">this_sample_ID</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_tRawCounts</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>

    <span class="va">this_raw_tSNE_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_all_tSNEs.html">run_all_tSNEs</a></span><span class="op">(</span><span class="va">perplexityValues</span>, <span class="va">this_tRawCounts</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"rawCounts_"</span>,
                                           <span class="va">this_sample_ID</span><span class="op">)</span>, <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">montage</span> <span class="op">&lt;-</span> <span class="fl">0</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">this_raw_tSNE_results</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"CDX_tSNE_"</span>, <span class="va">this_sample_ID</span>, <span class="st">".RData"</span><span class="op">)</span><span class="op">)</span>

    <span class="va">CDX_tSNE_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">this_raw_tSNE_results</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">this_sample_ID</span>
<span class="op">}</span>
<span class="co">#&gt; Perplexity = 70</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-6-1.png" width="1152"></p>
<pre><code>#&gt; Perplexity = 70</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-6-2.png" width="1152"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">CDX_tSNE_list</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"CDX_tSNE_list.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "CDX_tSNE_list.RData"))</span></code></pre></div>
<hr>
</div>
<div id="step-3-stemness-predictions" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-stemness-predictions" class="anchor"></a>Step 3: Stemness Predictions</h3>
<p>Each cell, in the SCLC CDX population of cells, will be scored using the predicted probability of it being a stem cell.</p>
<p>These stemness scores will be later used, to assist the identification putative cancer stem cells.</p>
<p>These scores are produced using machine learning methods - in this example they are random forests (RFs). These are trained on stem cell RNA-Seq expression data from induced pluripotent stem cells from the HipSci consortium. After the RFs are trained on the HipSci data, they are used to predict the likelihood that each of the cancer cells are stem cells.</p>
<p>The HipSci data is an expression matrix describing 196 samples; some of the samples and fibroblasts or mononuclear cells, but most are from iPSC bulk RNA-Seq samples.</p>
<div id="run-the-machine-learning-methods-to-predict-stemness-in-each-of-the-cancer-cells" class="section level4">
<h4 class="hasAnchor">
<a href="#run-the-machine-learning-methods-to-predict-stemness-in-each-of-the-cancer-cells" class="anchor"></a>Run the Machine Learning Methods to Predict Stemness in Each of the Cancer Cells</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">hipsci_training_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span> <span class="co"># data automatically available in this R package</span>
<span class="co">#&gt;                                                                                       Cell_type</span>
<span class="co">#&gt; HPSI0114i.bezi_1..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions       IPS</span>
<span class="co">#&gt; HPSI0114i.bezi_3..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions       IPS</span>
<span class="co">#&gt;                                                                                       A1BG</span>
<span class="co">#&gt; HPSI0114i.bezi_1..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions  0.7</span>
<span class="co">#&gt; HPSI0114i.bezi_3..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions  0.7</span>
<span class="co">#&gt;                                                                                       A1BG-AS1</span>
<span class="co">#&gt; HPSI0114i.bezi_1..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions        1</span>
<span class="co">#&gt; HPSI0114i.bezi_3..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions        2</span>
<span class="co">#&gt;                                                                                       A1CF</span>
<span class="co">#&gt; HPSI0114i.bezi_1..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions  0.1</span>
<span class="co">#&gt; HPSI0114i.bezi_3..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions  0.0</span>
<span class="co">#&gt;                                                                                       A2M</span>
<span class="co">#&gt; HPSI0114i.bezi_1..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions 0.1</span>
<span class="co">#&gt; HPSI0114i.bezi_3..induced.pluripotent.stem.cell..skin.derived..feeder.free.conditions 0.4</span></code></pre></div>
<p>Now we’ll use the run_ML_stemness_predictors function in the CSCPlots package, to use three random forest implementations (ranger, rf, randomForest) to predict stemness in each of the cancer cells in the two samples we will analyse (GSM4104161_SC4_Talazoparib.LB17011 &amp; GSM4104152_SC4.LB17009). The number of genes compared, between the HipSci and cancer expression datasets, will be limited to 16,000. If your computational resources are limited, or you would like the predictions to run more quickly, you can try using less, but lower than 10,000 is not recommeded.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">number_of_genes_to_use</span> <span class="op">&lt;-</span> <span class="fl">16000</span>
<span class="co">#number_of_genes_to_use &lt;- 10000 # can give reasonable results</span>
<span class="co">#number_of_genes_to_use &lt;- 16700 # fails on my R installation</span>

<span class="va">CDX_stemness_prediction_DFs_lists</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'ranger'</span>,
                                                                     <span class="va">CDX_tRawCounts_list</span>,
                                                                     <span class="va">hipsci_training_data</span>,
                                                                     <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>,
                                                                     <span class="va">number_of_genes_to_use</span>,
                                                                     <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'raw'</span>,
                                                                     <span class="va">resultsDir</span><span class="op">)</span>
<span class="co">#&gt; [1]:GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Probabilities: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; [2]:GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Probabilities: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; ML stemness predictions completed</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ranger_hipsci_raw"</span>
<span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'rf'</span>,
                                                                     <span class="va">CDX_tRawCounts_list</span>,
                                                                     <span class="va">hipsci_training_data</span>,
                                                                     <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>,
                                                                     <span class="va">number_of_genes_to_use</span>,
                                                                     <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'raw'</span>,
                                                                     <span class="va">resultsDir</span><span class="op">)</span>
<span class="co">#&gt; [1]:GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Probabilities: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; [2]:GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Probabilities: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; ML stemness predictions completed</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"rf_hipsci_raw"</span>
<span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_ML_stemness_predictors.html">run_ML_stemness_predictors</a></span><span class="op">(</span><span class="va">model_type</span> <span class="op">&lt;-</span> <span class="st">'randomForest'</span>,
                                                                     <span class="va">CDX_tRawCounts_list</span>,
                                                                     <span class="va">hipsci_training_data</span>,
                                                                     <span class="va">stemness_type</span> <span class="op">&lt;-</span> <span class="st">'hipsci'</span>,
                                                                     <span class="va">number_of_genes_to_use</span>,
                                                                     <span class="va">sample_data_type</span> <span class="op">&lt;-</span> <span class="st">'raw'</span>,
                                                                     <span class="va">resultsDir</span><span class="op">)</span>
<span class="co">#&gt; [1]:GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Probabilities: GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; [2]:GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Overlapping genes between sample and stemness data = 19784</span>
<span class="co">#&gt; Start of predictions</span>
<span class="co">#&gt; Predictions: testing </span>
<span class="co">#&gt; Probabilities: testing</span>
<span class="co">#&gt; Predictions: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; Probabilities: GSM4104152_SC4.LB17009</span>
<span class="co">#&gt; ML stemness predictions completed</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"randomForest_hipsci_raw"</span>

<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"CDX_stemness_prediction_DFs_lists.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "CDX_stemness_prediction_DFs_lists.RData"))</span></code></pre></div>
<hr>
</div>
</div>
<div id="step-4-genes-correlated-with-elevated-stemness-andor-high-transcription" class="section level3">
<h3 class="hasAnchor">
<a href="#step-4-genes-correlated-with-elevated-stemness-andor-high-transcription" class="anchor"></a>Step 4: Genes Correlated with Elevated Stemness and/or High Transcription</h3>
<p>Now that we have predicted stemness for each cancer cell, we can use that to see which genes/transcripts are most positively or negatively correlated with stemness in the two cancer cell populations, i.e. genes that are expressed more, or less, as stemness goes up</p>
<p>In a mixed cell population, or a population composed of cells from multiple samples and/or patients, this could be calculated in subsets of the cell population. The subsets could be for separate tumours, or subsets by cell type, or a combination of both, e.g. groups of malignant cancer cells separate patients.</p>
<p>However, here we will calculate for each of the two samples being studied correlations for all cells in each sample. The most postively correlated will later be used as stemness markers, to assist identification of cancer stem cells.</p>
<p>Genes correlated with transcription increases (total transcription and total transcribed genes) in the cancer cell populations will also be calculated for both samples, in step 4b.</p>
<div id="a-calculate-stemness-correlations" class="section level4">
<h4 class="hasAnchor">
<a href="#a-calculate-stemness-correlations" class="anchor"></a><strong>4a)</strong>: Calculate Stemness Correlations</h4>
<p>Display the top genes most correlated with stemness, for each of the three predictions of stemness that we performed in step 3:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_stemness_correlations_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">tCounts_to_use</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>

    <span class="va">scores_to_use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
    <span class="op">}</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span>

    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">score_to_use_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">'IPS'</span><span class="op">)</span> <span class="co"># for HipSci</span>
                                                                             <span class="co"># IPS predictions</span>
                                                                             <span class="co"># (vs fibro &amp; mononuclear)</span>
        <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">score_to_use_index</span><span class="op">]</span>
        
        <span class="va">CDX_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tCounts_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span>

        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>    
    <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; GSM4104161_SC4_Talazoparib.LB17011; ranger_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A     RPLP1     GAPDH      RPS3    CHCHD2     RPS4X      RPS2 </span>
<span class="co">#&gt; 0.6178536 0.6117993 0.6107854 0.6025732 0.6010813 0.6007260 0.6003718 0.5956794 </span>
<span class="co">#&gt;     RPLP0      PPIA </span>
<span class="co">#&gt; 0.5924952 0.5921076 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt; WI2-85898F10.1  RP11-337C18.9        DYNLRB2   RP11-769O8.2          POTEE </span>
<span class="co">#&gt;    -0.04325570    -0.03408901    -0.03373516    -0.03235263    -0.03178537 </span>
<span class="co">#&gt;    RP3-512E2.2      IL12A-AS1  RP11-293M10.2       HLA-DQB1      LINC00880 </span>
<span class="co">#&gt;    -0.03177451    -0.03173617    -0.03047567    -0.02987709    -0.02949740 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104161_SC4_Talazoparib.LB17011; rf_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     RPLP1     GAPDH     H3F3A      RPS3     RPS4X     RPLP0    CHCHD2 </span>
<span class="co">#&gt; 0.5461697 0.5367144 0.5332663 0.5314123 0.5297523 0.5284654 0.5230932 0.5223603 </span>
<span class="co">#&gt;      RPS2      RPS5 </span>
<span class="co">#&gt; 0.5218277 0.5199641 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;          PIGR        LANCL3 RP11-1105G2.4       C9orf50          IL33 </span>
<span class="co">#&gt;   -0.03864796   -0.03506294   -0.03357330   -0.03295053   -0.03286570 </span>
<span class="co">#&gt;          RGS1         ERP27         PEG10    AC012456.4 RP11-380L11.4 </span>
<span class="co">#&gt;   -0.03250763   -0.03235044   -0.03197755   -0.03065440   -0.03044279 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104161_SC4_Talazoparib.LB17011; randomForest_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A     RPLP1    CHCHD2     RPS4X      RPS3     RPLP0      PPIA </span>
<span class="co">#&gt; 0.6539281 0.6426318 0.6387235 0.6347425 0.6316140 0.6278591 0.6278085 0.6272227 </span>
<span class="co">#&gt;      RPS5   NGFRAP1 </span>
<span class="co">#&gt; 0.6187547 0.6186770 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;         RGPD8      HLA-DQB1        MARCH1 RP11-1105G2.4        SEL1L2 </span>
<span class="co">#&gt;   -0.03361935   -0.03191341   -0.03164974   -0.03150374   -0.03058192 </span>
<span class="co">#&gt;  RP11-769O8.2      HLA-DQA2        POU6F2   RP11-80H5.7           CKM </span>
<span class="co">#&gt;   -0.03032575   -0.02902365   -0.02837492   -0.02774004   -0.02743757 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104152_SC4.LB17009; ranger_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A      RPS2     RPS4X      PPIA      RPS3     GAPDH    CHCHD2 </span>
<span class="co">#&gt; 0.6306042 0.6228120 0.6202316 0.6159585 0.6126695 0.6109338 0.6105199 0.6099754 </span>
<span class="co">#&gt;     RPLP1     RPLP0 </span>
<span class="co">#&gt; 0.6092387 0.6068148 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;        C1QB        AIF1       POTEE        TLR2       ITLN2     HLA-DMB </span>
<span class="co">#&gt; -0.04459674 -0.04316701 -0.04237465 -0.04078755 -0.03827235 -0.03793694 </span>
<span class="co">#&gt;        GDF3        MYL9        APOE  AP000255.6 </span>
<span class="co">#&gt; -0.03778818 -0.03751757 -0.03747210 -0.03645387 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104152_SC4.LB17009; rf_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;     GAPDH      RPS2     RPS4X     RPLP1      RPS5      RPS3     H3F3A     RPLP0 </span>
<span class="co">#&gt; 0.5677822 0.5662191 0.5548749 0.5508787 0.5502041 0.5496894 0.5494387 0.5486337 </span>
<span class="co">#&gt;      YBX1    CHCHD2 </span>
<span class="co">#&gt; 0.5469834 0.5448181 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       NCKAP1L        EPSTI1          TLR2         F2RL3 RP11-168P13.1 </span>
<span class="co">#&gt;   -0.03507331   -0.03383230   -0.03362693   -0.03253707   -0.03191236 </span>
<span class="co">#&gt;       SCGB3A2       PSTPIP1    AC074391.1         RAB43  RP11-138H8.7 </span>
<span class="co">#&gt;   -0.03117861   -0.03039226   -0.03027810   -0.03023461   -0.02963887 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104152_SC4.LB17009; randomForest_hipsci_raw:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A      RPS2     RPS4X      RPS3      RPS5      PPIA    CHCHD2 </span>
<span class="co">#&gt; 0.6935580 0.6877813 0.6817751 0.6792575 0.6711148 0.6694220 0.6677912 0.6657743 </span>
<span class="co">#&gt;     RPLP0     RPLP1 </span>
<span class="co">#&gt; 0.6657181 0.6629313 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;        C1QB       POTEE        AIF1     HLA-DMB      BCL2A1      RPL10L </span>
<span class="co">#&gt; -0.05520017 -0.05405304 -0.05265054 -0.04515927 -0.04444980 -0.04306157 </span>
<span class="co">#&gt;       ITLN2        TLR2        APOE    HLA-DRB1 </span>
<span class="co">#&gt; -0.04269871 -0.04134138 -0.04064711 -0.04039440</span></code></pre></div>
</div>
<div id="b-increased-transcription-correlations" class="section level4">
<h4 class="hasAnchor">
<a href="#b-increased-transcription-correlations" class="anchor"></a><strong>4b)</strong>: Increased Transcription Correlations</h4>
<p>Similarly, we’ll now find the genes positively or negatively correlated with increase in transcription in the cell populations from each sample. We will correlate against total transcription increase as measured by the count of all reads in all genes, for each cell, and separately against total transcribed genes in each cell (number of genes with &gt;= 1 read).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_total_transcription_levels_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">CDX_total_transcribed_genes_list</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">df</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
    <span class="va">total_transcription_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
    <span class="va">total_transcribed_genes_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="op">)</span>
    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">thisCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span>
        <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">thisCount</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Is na\n"</span><span class="op">)</span> <span class="op">}</span>; <span class="kw">if</span><span class="op">(</span><span class="va">thisCount</span> <span class="op">&lt;=</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"0 or less\n"</span><span class="op">)</span> <span class="op">}</span>
        <span class="va">total_transcribed_genes_vec</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">thisCount</span>
    <span class="op">}</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">total_transcribed_genes_vec</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.test.html">cor.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">total_transcribed_genes_vec</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">total_transcription_levels</span><span class="op">)</span>,
                   method<span class="op">=</span><span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span> <span class="co"># cor ~0.95</span>
    <span class="va">CDX_total_transcription_levels_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">total_transcription_levels</span>
    <span class="va">CDX_total_transcribed_genes_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">total_transcribed_genes_vec</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_total_transcription_levels_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_total_transcribed_genes_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
    <span class="va">df</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>    
<span class="op">}</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Pearson's product-moment correlation</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; data:  as.numeric(total_transcribed_genes_vec) and as.numeric(total_transcription_levels)</span>
<span class="co">#&gt; t = 242.07, df = 5849, p-value &lt; 2.2e-16</span>
<span class="co">#&gt; alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">#&gt; 95 percent confidence interval:</span>
<span class="co">#&gt;  0.9511595 0.9558133</span>
<span class="co">#&gt; sample estimates:</span>
<span class="co">#&gt;       cor </span>
<span class="co">#&gt; 0.9535433 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Pearson's product-moment correlation</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; data:  as.numeric(total_transcribed_genes_vec) and as.numeric(total_transcription_levels)</span>
<span class="co">#&gt; t = 244.36, df = 5603, p-value &lt; 2.2e-16</span>
<span class="co">#&gt; alternative hypothesis: true correlation is not equal to 0</span>
<span class="co">#&gt; 95 percent confidence interval:</span>
<span class="co">#&gt;  0.9538414 0.9583360</span>
<span class="co">#&gt; sample estimates:</span>
<span class="co">#&gt;      cor </span>
<span class="co">#&gt; 0.956145</span>

<span class="va">CDX_transcription_correlations_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>

    <span class="va">tCounts_to_use</span> <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
    <span class="va">scores_to_use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">CDX_total_transcription_levels_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,
                               <span class="va">CDX_total_transcribed_genes_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Total transcription"</span>, <span class="st">"Total transcribed genes"</span><span class="op">)</span>

    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">scores_to_use</span> <span class="op">&lt;-</span> <span class="va">scores_to_use_list</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>        

        <span class="va">CDX_transcription_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="../reference/get_expression_scores_correlations.html">get_expression_scores_correlations</a></span><span class="op">(</span><span class="va">tCounts_to_use</span>, <span class="va">scores_to_use</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span>
        
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"; "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scores_to_use_list</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nPositive correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nNegative correlations (top 10):\n"</span><span class="op">)</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; GSM4104161_SC4_Talazoparib.LB17011; Total transcription:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A     RPLP1     RPLP0      RPS3     RPS4X      RPS2     RPL41 </span>
<span class="co">#&gt; 0.8095793 0.7907966 0.7875901 0.7873283 0.7818366 0.7806190 0.7775307 0.7755684 </span>
<span class="co">#&gt;      PTMA    EEF1A1 </span>
<span class="co">#&gt; 0.7721216 0.7679796 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;      HLA-DMB RP11-769O8.2     HLA-DQB1    EDRF1-AS1      ABHD16B   AC097381.1 </span>
<span class="co">#&gt;  -0.03477511  -0.03311174  -0.03194626  -0.03154397  -0.03132941  -0.03008679 </span>
<span class="co">#&gt;       SLC5A5   RP11-2H8.2         F8A2         MYOT </span>
<span class="co">#&gt;  -0.02932421  -0.02897103  -0.02861259  -0.02734263 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104161_SC4_Talazoparib.LB17011; Total transcribed genes:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A      PTMA    CHCHD2     RPLP1      PPIA     RPLP0       RAN </span>
<span class="co">#&gt; 0.7994971 0.7811395 0.7746597 0.7625355 0.7594703 0.7576302 0.7554167 0.7520918 </span>
<span class="co">#&gt;      RPS3     HINT1 </span>
<span class="co">#&gt; 0.7490380 0.7479833 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       HLA-DMB  RP11-769O8.2      HLA-DQB1     EDRF1-AS1    RP11-2H8.2 </span>
<span class="co">#&gt;   -0.03399386   -0.03250755   -0.03193531   -0.02987670   -0.02960475 </span>
<span class="co">#&gt;       ABHD16B    AC097381.1        SHISA7          F8A2 RP11-293M10.2 </span>
<span class="co">#&gt;   -0.02958616   -0.02871900   -0.02864544   -0.02672954   -0.02668968 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104152_SC4.LB17009; Total transcription:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A      RPS2      RPS3     RPLP0     RPS4X     RPLP1      RPL8 </span>
<span class="co">#&gt; 0.8370683 0.8314243 0.8254465 0.8240677 0.8193297 0.8175826 0.8122321 0.8081716 </span>
<span class="co">#&gt;      PPIA      RPS5 </span>
<span class="co">#&gt; 0.8081188 0.8070491 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;        C1QB        AIF1       POTEE        TLR2     HLA-DMB        MYL9 </span>
<span class="co">#&gt; -0.04455281 -0.04438695 -0.04415277 -0.04067521 -0.04029594 -0.03895785 </span>
<span class="co">#&gt;        GDF3      BCL2A1       SEPP1        PLD4 </span>
<span class="co">#&gt; -0.03890038 -0.03754954 -0.03715973 -0.03515968 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; GSM4104152_SC4.LB17009; Total transcribed genes:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Positive correlations (top 10):</span>
<span class="co">#&gt;      YBX1     H3F3A      RPS2      PPIA     RPLP0      PTMA    CHCHD2     RPS4X </span>
<span class="co">#&gt; 0.8287282 0.8230463 0.7990960 0.7955016 0.7933622 0.7902260 0.7872681 0.7872677 </span>
<span class="co">#&gt;      RPS3     RPLP1 </span>
<span class="co">#&gt; 0.7866180 0.7843388 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Negative correlations (top 10):</span>
<span class="co">#&gt;       POTEE        C1QB        AIF1        TLR2     HLA-DMB      BCL2A1 </span>
<span class="co">#&gt; -0.05665014 -0.05582511 -0.05488106 -0.05046652 -0.04845672 -0.04627461 </span>
<span class="co">#&gt;      RPL10L        GDF3        APOE    HLA-DRB1 </span>
<span class="co">#&gt; -0.04604057 -0.04303213 -0.04241209 -0.04232433</span></code></pre></div>
<p>Info about full HUGO gene names is available within this R package, and will be used to help understand the correlations:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">all_HUGO_abbreviations_and_fullnames</span><span class="op">)</span>
<span class="co">#&gt;     symbol                           name</span>
<span class="co">#&gt; 1     A1BG         alpha-1-B glycoprotein</span>
<span class="co">#&gt; 2 A1BG-AS1           A1BG antisense RNA 1</span>
<span class="co">#&gt; 3     A1CF APOBEC1 complementation factor</span>
<span class="co">#&gt; 4      A2M          alpha-2-macroglobulin</span>
<span class="co">#&gt; 5  A2M-AS1            A2M antisense RNA 1</span>
<span class="co">#&gt; 6    A2ML1   alpha-2-macroglobulin like 1</span></code></pre></div>
<p>However, HUGO continually updates this info. To get the most recent data for HUGO full names, you can uncomment the following code:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">use_biomart</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co"># Set to TRUE to use Biomart to get the full names, when FALSE downloads from ftp.ebi.ac.uk</span>
<span class="kw">if</span><span class="op">(</span><span class="va">use_biomart</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">{</span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>  <span class="op">}</span>

<span class="co">#if(use_biomart == FALSE){</span>
<span class="co">#    HUGO_ftp_address &lt;- 'ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt'</span>
<span class="co">#    download.file(HUGO_ftp_address, "hgnc_complete_set.txt")</span>
<span class="co">#    all_HUGO_info &lt;- read.csv(paste0("hgnc_complete_set.txt"), header=TRUE, sep="\t",</span>
<span class="co">#                              stringsAsFactors = FALSE)</span>
<span class="co">#    all_HUGO_abbreviations_and_fullnames &lt;- all_HUGO_info[,c(2,3)]</span>
<span class="co">#    save(all_HUGO_abbreviations_and_fullnames, file = paste0(saveDir,</span>
<span class="co">#        "all_HUGO_abbreviations_and_fullnames.RData"))</span>
<span class="co">#} else {</span>
<span class="co">#    mart &lt;- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")</span>
<span class="co">#}</span></code></pre></div>
<p>You may have noticed that there are a lot of ribosomal genes in the most positivley correlated genes, both for stemness and transcription. The gene IDs may also not have a lot of meaning for you, if you are not familiar with them. For these reasons, we’ll look up a full gene name/description for each of them, and remove those from the mostly highly correlated that have “ribosomal” in the full gene name.</p>
<p>First we’ll do that for the genes correlated with stemness increase, in the cancer cell populations. We’ll just do it for some of the stemness correlations (Ranger generated), but you could modify the stemness_correlations_indices_to_use value in the code to look at the others:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">number_of_correlations</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">stemness_correlations_indices_to_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span> <span class="co"># Manually set for</span>
                                               <span class="co"># 1="GSM4104161_SC4_Talazoparib.LB17011; ranger_hipsci_raw"</span>
                                               <span class="co"># 4="GSM4104152_SC4.LB17009; ranger_hipsci_raw"</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">stemness_correlations_indices_to_use</span><span class="op">)</span><span class="op">{</span>
<span class="co">#for(i in 1:length(CDX_stemness_correlations_list)){ # Use this for all</span>
    <span class="va">gene_names_to_get</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">number_of_correlations</span><span class="op">]</span><span class="op">)</span>

    <span class="kw">if</span><span class="op">(</span><span class="va">use_biomart</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span>
        <span class="va">match_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">gene_names_to_get</span>, <span class="va">all_HUGO_abbreviations_and_fullnames</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>
        <span class="va">gene_names_and_functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
            hgnc_symbol <span class="op">=</span> <span class="va">gene_names_to_get</span>,
            description <span class="op">=</span> <span class="va">all_HUGO_abbreviations_and_fullnames</span><span class="op">[</span><span class="va">match_pos</span>,<span class="fl">2</span><span class="op">]</span>,
            stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
        <span class="op">)</span>
        <span class="va">gene_names_and_functions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gene_names_and_functions</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'-'</span>
    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
        <span class="va">gene_names_and_functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hgnc_symbol"</span>, <span class="st">"description"</span><span class="op">)</span>,
                                          filters <span class="op">=</span> <span class="st">'hgnc_symbol'</span>, <span class="va">gene_names_to_get</span>, mart <span class="op">=</span> <span class="va">mart</span><span class="op">)</span>
    <span class="op">}</span>

    <span class="co">#print(gene_names_and_functions)</span>
    <span class="va">correlation_results</span> <span class="op">&lt;-</span> <span class="va">CDX_stemness_correlations_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">number_of_correlations</span><span class="op">]</span>
    <span class="va">correlations_with_functions_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
      <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">correlation_results</span><span class="op">)</span>,
      <span class="va">correlations</span> <span class="op">&lt;-</span> <span class="va">correlation_results</span>,
      <span class="va">functions</span> <span class="op">&lt;-</span> <span class="va">gene_names_and_functions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">correlation_results</span><span class="op">)</span>,
                   <span class="va">gene_names_and_functions</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">2</span><span class="op">]</span>,
      stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Genes'</span>, <span class="st">'Correlation'</span>, <span class="st">'Function'</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
    <span class="va">correlations_with_functions_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'-'</span>
    <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">' \\[Source.*'</span>, <span class="st">""</span>,
                                               <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>
    <span class="va">to_remove_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"ribosomal"</span>,  <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>
    <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">to_remove_pos</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">correlations_with_functions_df_no_ribo</span> <span class="op">&lt;-</span> <span class="va">correlations_with_functions_df</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">to_remove_pos</span><span class="op">)</span>,<span class="op">]</span>
    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
        <span class="va">correlations_with_functions_df_no_ribo</span> <span class="op">&lt;-</span> <span class="va">correlations_with_functions_df</span>
    <span class="op">}</span>
    <span class="co">#cat(paste0('Correlations (top ', number_of_correlations, '):\n'))</span>
    <span class="co">#print(correlations_with_functions_df)</span>
    <span class="co">#cat(paste0('Correlations (ribosomal removed):\n'))</span>
    <span class="co">#print(correlations_with_functions_df_no_ribo)</span>

    <span class="va">top_10_stemness_position</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>
    <span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span><span class="op">[[</span><span class="op">(</span><span class="va">top_10_stemness_position</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="va">correlations_with_functions_df_no_ribo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span><span class="op">)</span><span class="op">[</span><span class="va">top_10_stemness_position</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
<span class="op">}</span>

<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Top 10 positive correlations with ribosomal removed:\n"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Top 10 positive correlations with ribosomal removed:</span>
<span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span>
<span class="co">#&gt; $`GSM4104161_SC4_Talazoparib.LB17011; ranger_hipsci_raw`</span>
<span class="co">#&gt;  [1] "YBX1"    "H3F3A"   "GAPDH"   "CHCHD2"  "PPIA"    "PTMA"    "HINT1"  </span>
<span class="co">#&gt;  [8] "EEF1B2"  "NGFRAP1" "EEF1A1" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`GSM4104152_SC4.LB17009; ranger_hipsci_raw`</span>
<span class="co">#&gt;  [1] "YBX1"    "H3F3A"   "PPIA"    "GAPDH"   "CHCHD2"  "PTMA"    "EEF1B2" </span>
<span class="co">#&gt;  [8] "HINT1"   "SLC25A5" "PPDPF"</span></code></pre></div>
<p>And we’ll do the same, for the high transcription associated genes:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">number_of_correlations</span> <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">transcription_correlations_indices_to_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span> <span class="co"># Manually set for</span>
                                                    <span class="co"># 1="GSM4104161_SC4_Talazoparib.LB17011; Total transcription"</span>
                                                    <span class="co"># 3="GSM4104152_SC4.LB17009; Total transcription"</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">transcription_correlations_indices_to_use</span><span class="op">)</span><span class="op">{</span>
<span class="co">#for(i in 1:length(CDX_transcription_correlations_list)){ # Use for all</span>
    <span class="va">gene_names_to_get</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">number_of_correlations</span><span class="op">]</span><span class="op">)</span>
    <span class="kw">if</span><span class="op">(</span><span class="va">use_biomart</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span>
        <span class="va">match_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">gene_names_to_get</span>, <span class="va">all_HUGO_abbreviations_and_fullnames</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>
        <span class="va">gene_names_and_functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
            hgnc_symbol <span class="op">=</span> <span class="va">gene_names_to_get</span>,
            description <span class="op">=</span> <span class="va">all_HUGO_abbreviations_and_fullnames</span><span class="op">[</span><span class="va">match_pos</span>,<span class="fl">2</span><span class="op">]</span>,
            stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
        <span class="op">)</span>
        <span class="va">gene_names_and_functions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gene_names_and_functions</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'-'</span>
    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
        <span class="va">gene_names_and_functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hgnc_symbol"</span>, <span class="st">"description"</span><span class="op">)</span>,
                                          filters <span class="op">=</span> <span class="st">'hgnc_symbol'</span>, <span class="va">gene_names_to_get</span>, mart <span class="op">=</span> <span class="va">mart</span><span class="op">)</span>
    <span class="op">}</span>
    <span class="co">#print(gene_names_and_functions)</span>
    <span class="va">correlation_results</span> <span class="op">&lt;-</span> <span class="va">CDX_transcription_correlations_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">number_of_correlations</span><span class="op">]</span>
    <span class="va">correlations_with_functions_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
      <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">correlation_results</span><span class="op">)</span>,
      <span class="va">correlations</span> <span class="op">&lt;-</span> <span class="va">correlation_results</span>,
      <span class="va">functions</span> <span class="op">&lt;-</span> <span class="va">gene_names_and_functions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">correlation_results</span><span class="op">)</span>,
                   <span class="va">gene_names_and_functions</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">2</span><span class="op">]</span>,
      stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Genes'</span>, <span class="st">'Correlation'</span>, <span class="st">'Function'</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
    <span class="va">correlations_with_functions_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'-'</span>
    <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">' \\[Source.*'</span>, <span class="st">""</span>,
                                               <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>
    <span class="va">correlations_with_functions_df_no_ribo</span> <span class="op">&lt;-</span>
        <span class="va">correlations_with_functions_df</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"ribosomal"</span>,  <span class="va">correlations_with_functions_df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span>

    <span class="co">#cat(paste0('Correlations (top ', number_of_correlations, '):\n'))</span>
    <span class="co">#print(correlations_with_functions_df)</span>
    <span class="co">#cat(paste0('Correlations (ribosomal removed):\n'))</span>
    <span class="co">#print(correlations_with_functions_df_no_ribo)</span>

    <span class="va">top_10_transcription_position</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>
    <span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">[[</span><span class="op">(</span><span class="va">top_10_transcription_position</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="va">correlations_with_functions_df_no_ribo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">)</span><span class="op">[</span><span class="va">top_10_transcription_position</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_transcription_correlations_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
<span class="op">}</span>

<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Top 10 positive correlations with ribosomal removed:\n"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Top 10 positive correlations with ribosomal removed:</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">)</span>
<span class="co">#&gt; $`GSM4104161_SC4_Talazoparib.LB17011; Total transcription`</span>
<span class="co">#&gt;  [1] "YBX1"   "H3F3A"  "PTMA"   "EEF1A1" "PPIA"   "CHCHD2" "HINT1"  "FTL"   </span>
<span class="co">#&gt;  [9] "RAN"    "GAPDH" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`GSM4104152_SC4.LB17009; Total transcription`</span>
<span class="co">#&gt;  [1] "YBX1"    "H3F3A"   "PPIA"    "HINT1"   "CHCHD2"  "PTMA"    "HNRNPA1"</span>
<span class="co">#&gt;  [8] "EEF1A1"  "EEF1B2"  "RAN"</span></code></pre></div>
<hr>
</div>
</div>
<div id="step-5-t-sne-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#step-5-t-sne-analysis" class="anchor"></a>Step 5: t-SNE Analysis</h3>
<p>We’ll now look at the t-SNEs that were generated in step 2, and explore their topologies. The stemness scores that were produced in step 3, will be overlaid onto each of the cells represented in the t-SNEs, to look for any stemness gradients that are apparent. Similarly, we’ll alse look at expression patterns of the genes most correlated with stemness and transcription that we identified in step 4. This is with the aim of identifying a subset of cells that appear to have high transcription, high stemness, and that are interesting in the context of the shape of the t-SNE.</p>
<p>The two samples that we are analysing are quite homogenous, both in terms of cell type i.e. all are expected to be malignant cancer cells from a CDX, and that the two cell populations are each from a separate tumour.</p>
<p>If you are following this tutorial on the CSCPlots web pages at Github, the generated plots that are shown may appear small, and are harder to study in detail. It’s recommended to use the high resolution tif files that are printed by the code in this tutorial. We’ll be looking at the relationships between small numbers of cells in the t-SNEs, so it’s important to see them at high resolution. This is also true for other plots generated in this tutorial - it’s much better to look at the tif files.</p>
<div id="a-annotate-the-t-snes" class="section level4">
<h4 class="hasAnchor">
<a href="#a-annotate-the-t-snes" class="anchor"></a><strong>5a)</strong>: Annotate the t-SNEs</h4>
<p>Add the scores to a list of t-SNEs, for later plotting and analysis:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use the perplexity values you set previously, or a subset of them, if appropriate:</span>
<span class="co">#perplexityValues &lt;&lt;- c(5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130)</span>
<span class="co">#perplexityValues &lt;&lt;- 70</span>

<span class="va">CDX_tSNE_list_with_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">stemness_score_indices_to_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span> <span class="co"># manually set chosen scores to add: 2 for HipSci IPS predictions (fibro = 1; mononuclear = 3)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_tSNE_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>

    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="va">CDX_tSNE_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>

    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>,
                                                                    <span class="va">perplexityValues</span>,
                                                                    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">CDX_total_transcription_levels_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                                                                    <span class="st">"Total_transcription"</span><span class="op">)</span>
    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>,
                                                                    <span class="va">perplexityValues</span>,
                                                                    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">CDX_total_transcribed_genes_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                                                                    <span class="st">"Total_transcribed_genes"</span><span class="op">)</span>

    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_scores_to_tSNE_results.html">add_scores_to_tSNE_results</a></span><span class="op">(</span><span class="va">tSNE_results_with_annotations_tmp</span>,
                                                                        <span class="va">perplexityValues</span>,
                                                                        <span class="va">scores</span> <span class="op">&lt;-</span> <span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">stemness_score_indices_to_use</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>,
                                                                        <span class="va">scoresName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span>,
                                                                        <span class="st">"_"</span>,
                                                                        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_stemness_prediction_DFs_lists</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">stemness_score_indices_to_use</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span>

    <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tSNE_results_with_annotations_tmp</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
    <span class="va">tSNE_results_with_annotations_tmp</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"CDX_tSNE_list_with_annotations.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "CDX_tSNE_list_with_annotations.RData"))</span></code></pre></div>
</div>
<div id="b-plot-t-sne-stemness-and-transcription-gradients" class="section level4">
<h4 class="hasAnchor">
<a href="#b-plot-t-sne-stemness-and-transcription-gradients" class="anchor"></a><strong>5b)</strong>: Plot t-SNE Stemness and Transcription Gradients</h4>
<p>Use the newly annotated tSNE list (CDX_tSNE_list_with_annotations), to plot the t-SNEs overlaid with stemness scores and transcription levels:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Print all the annotated scores previously added i.e. Total_transcription, Total_transcribed_genes,</span>
<span class="co"># ranger_hipsci_raw_IPS, etc.</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="co">#j &lt;- 1 # perplexity values index</span>
    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">perplexityValues</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
        <span class="co">#outDir &lt;- paste0(tSNEsDir, names(CDX_tSNE_list_with_annotations)[i], "/")</span>
        <span class="co">#dir.create(file.path(outDir), showWarnings = FALSE)</span>
        <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
        
        <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">3</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="co"># 1:2 are the Y1 &amp; Y2</span>
                                                                             <span class="co"># values so skip them</span>
            <span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                        <span class="va">runName</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,
                        <span class="va">scoresName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>,
                        <span class="va">datasetName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">outDir</span>,
                        <span class="va">labelPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span>
        <span class="op">}</span>
    <span class="op">}</span>
<span class="op">}</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-16-1.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-2.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-3.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-4.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-5.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-6.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-7.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-8.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-9.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-16-10.png" width="1152"></p>
</div>
<div id="c-plot-t-snes-with-expression-of-highly-stemness-and-transcription-associated-genes" class="section level4">
<h4 class="hasAnchor">
<a href="#c-plot-t-snes-with-expression-of-highly-stemness-and-transcription-associated-genes" class="anchor"></a><strong>5c)</strong>: Plot t-SNEs with Expression of Highly Stemness and Transcription Associated Genes</h4>
<p>To continue building up a picture of which areas of the t-SNEs may contain putative cancer stem cells, genes correlated with high stemness and/or high transcription, will now be displayed.</p>
<p>We will select the genes to print their expression within the t-SNE; here we’ll use genes in the top 10 of stemness <strong>and</strong> transcription positive correlations, in <strong>either</strong> sample:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_names_to_print_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                           <span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">gene_names_to_print_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">CDX_top_10_stemness_positive_correlations_no_ribo</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,
                           <span class="va">CDX_top_10_transcription_positive_correlations_no_ribo</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">gene_names_to_print_1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">gene_names_to_print_1</span> <span class="op">%in%</span>
                                                            <span class="va">gene_names_to_print_2</span><span class="op">)</span><span class="op">]</span>,
                                <span class="va">gene_names_to_print_2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">gene_names_to_print_2</span> <span class="op">%in%</span>
                                                            <span class="va">gene_names_to_print_1</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>

<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># if multiple perplexity values were used for t-SNEs, this number defines the</span>
                 <span class="co"># index in the list of t-SNEs to use</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span></code></pre></div>
<p>Then we print the t-SNEs showing expression of the chosen genes, for manual inspection:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span><span class="op">(</span><span class="va">cohort_index</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>,
                                <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">run_name</span>, <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>,
                                <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt;   YBX1 [Y-box binding protein 1]</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
<pre><code>#&gt;   H3F3A [H3 Histone Family Member 3A]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-2.png" width="1152"></p>
<pre><code>#&gt;   GAPDH [glyceraldehyde-3-phosphate dehydrogenase]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-3.png" width="1152"></p>
<pre><code>#&gt;   CHCHD2 [coiled-coil-helix-coiled-coil-helix domain containing 2]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-4.png" width="1152"></p>
<pre><code>#&gt;   PPIA [peptidylprolyl isomerase A]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-5.png" width="1152"></p>
<pre><code>#&gt;   PTMA [prothymosin alpha]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-6.png" width="1152"></p>
<pre><code>#&gt;   HINT1 [histidine triad nucleotide binding protein 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-7.png" width="1152"></p>
<pre><code>#&gt;   EEF1B2 [eukaryotic translation elongation factor 1 beta 2]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-8.png" width="1152"></p>
<pre><code>#&gt;   EEF1A1 [eukaryotic translation elongation factor 1 alpha 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-9.png" width="1152"></p>
<pre><code>#&gt;   RAN [RAN, member RAS oncogene family]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-10.png" width="1152"></p>
<pre><code>#&gt;   YBX1 [Y-box binding protein 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-11.png" width="1152"></p>
<pre><code>#&gt;   H3F3A [H3 Histone Family Member 3A]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-12.png" width="1152"></p>
<pre><code>#&gt;   GAPDH [glyceraldehyde-3-phosphate dehydrogenase]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-13.png" width="1152"></p>
<pre><code>#&gt;   CHCHD2 [coiled-coil-helix-coiled-coil-helix domain containing 2]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-14.png" width="1152"></p>
<pre><code>#&gt;   PPIA [peptidylprolyl isomerase A]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-15.png" width="1152"></p>
<pre><code>#&gt;   PTMA [prothymosin alpha]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-16.png" width="1152"></p>
<pre><code>#&gt;   HINT1 [histidine triad nucleotide binding protein 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-17.png" width="1152"></p>
<pre><code>#&gt;   EEF1B2 [eukaryotic translation elongation factor 1 beta 2]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-18.png" width="1152"></p>
<pre><code>#&gt;   EEF1A1 [eukaryotic translation elongation factor 1 alpha 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-19.png" width="1152"></p>
<pre><code>#&gt;   RAN [RAN, member RAS oncogene family]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-18-20.png" width="1152"></p>
<p>We can also introduce biological and/or clinical information at this stage, to further assist identification of the cancer stem cells. SCLC is thought to have genes associated with tumour subtypes: these are ASCL1, NEUROD1, YAP1 and POU2F3. Looking at the expression of these genes in the t-SNEs indicates that ASCL1 is influential in the samples that we are studying:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="st">'ASCL1'</span>
<span class="kw">for</span><span class="op">(</span><span class="va">cohort_index</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>, <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">run_name</span>, <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>, <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt;   ASCL1 [achaete-scute family bHLH transcription factor 1]</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-19-1.png" width="1152"></p>
<pre><code>#&gt;   ASCL1 [achaete-scute family bHLH transcription factor 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-19-2.png" width="1152"></p>
<p>N.B. In addition, comparison of PARP1 expression between the Talazoparib PARP inhibitor treated and untreated samples i.e. GSM4104161_SC4_Talazoparib.LB17011 vs GSM4104152_SC4.LB17009, shows that PARP1 expression has a higher maximum after treatment (in GSM4104161_SC4_Talazoparib.LB17011). This corresponds to the area of the t-SNE that has cells with high transcription and predicted stemness. This may be of clinical interest, because post-therapy, the drug’s target is elevated in a subset of cells that may be driving the tumour.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="st">'PARP1'</span>
<span class="kw">for</span><span class="op">(</span><span class="va">cohort_index</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>,
                                <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">run_name</span>, <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>,
                                <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt;   PARP1 [poly(ADP-ribose) polymerase 1]</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-20-1.png" width="1152"></p>
<pre><code>#&gt;   PARP1 [poly(ADP-ribose) polymerase 1]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-20-2.png" width="1152"></p>
</div>
</div>
<div id="step-6-pcsc-group-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#step-6-pcsc-group-selection" class="anchor"></a>Step 6: pCSC Group Selection</h3>
<p>Step 5 compared the topologies of the t-SNEs with elevated predicted stemness and very high transcription. This indicated that there are regions of the t-SNEs where the cells appear to have increased stemness and aberrantly high transcription. These cells are good candidates for further study as putative cancer stem cells. We will select a group of cells from these areas, which can then be used to create some preferential expression plots.</p>
<p>To identify a group of cells to examine is may be arbitrary to some extent, as we are imposing a cutoff on the gradients of transcription and stemness, However, we’ll look for clusters of cells, that are located in the region of interest.</p>
<p>Generally I try to select somewhere in the region of between 2 and 30 cells, but it depends on the topology of the t-SNE and characteristics of the cell population.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">pCSCsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="va">scoresName</span> <span class="op">&lt;-</span> <span class="st">"Total_transcription"</span>,
                      <span class="va">datasetName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>,
                      <span class="va">labelPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-21-1.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>In the plot that was just generated, on the Y2 axis, in my plot (yours may be slightly different) cells located above Y2 = 39 appear interesting. In the plots that we previously created for this sample, this area of the t-SNE has high total transcribed genes, high total transcription, high predicted stemness using HipSci (ranger, rf and randomForest predictions). The genes highly correlated with stemness and transcriptiion are also highly expressed in this region, e.g. YBX1, CHCHD2, H3F3A, PPIA, HINT1 &amp; RAN. You could select cells slightly above or below this cutoff, and see how the preferential results generated in steps 7 &amp; 8 change, but they likely will be very similar.</p>
<p>To assist pCSC group selection, we will use t-SNEs with total transcription displayed. We’ll identify cells of interest by colouring them black or labelling them, and then noting their positions (row numbers) in the melanoma expression matrix. We’ll use plots with grids in the background, to enable this. Generated tif files will are written to the relevant sample’s results directory under “tSNEs/pCSC_group_selection/”, where you can view them.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">found_pos1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Y2</span> <span class="op">&gt;</span> <span class="fl">39</span><span class="op">)</span>
<span class="co">#found_pos1 &lt;- c(165, 889, 893, 1127, 1186, 1277, 1303, 3448, 4175, 4203, 4609, 4874, 5295, 5297, 5425, 5429) # In my plot these are the cells noted in found_pos1</span>

<span class="co"># reorder by total transcription</span>
<span class="va">found_pos1_ordered</span> <span class="op">&lt;-</span> <span class="va">found_pos1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos1</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>To check the selection, plot with highlighted or labelled selected cells - output is in a file ending in “with_grid-selected_cells.tif” or “with_grid-annotated.tif”:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>,
                      <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="va">found_pos1_ordered</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># selected cells in black</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-23-1.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL

plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions &lt;- found_pos1_ordered, NULL, highestOnTop &lt;- TRUE) # selected cell names labelled</a></code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-23-2.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>These are the cell IDs for this pCSC group, and we’ll store their expression matrix positions for later use:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011</span> <span class="op">&lt;-</span> <span class="va">found_pos1_ordered</span>

<span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011_df</span> <span class="op">&lt;-</span>
    <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos1_ordered</span>,<span class="op">]</span>
<span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011_cell_names</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011_cell_names</span>, sep <span class="op">=</span> <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; TGGCCAGTCGAGCCCA-1</span>
<span class="co">#&gt; TGGCGCAAGGACCACA-1</span>
<span class="co">#&gt; TCAGCAAGTTCGGCAC-1</span>
<span class="co">#&gt; ATTTCTGCACGCGAAA-1</span>
<span class="co">#&gt; TACCTTACATGTTCCC-1</span>
<span class="co">#&gt; AGGGTGAGTAAGAGGA-1</span>
<span class="co">#&gt; GATGCTAGTGCATCTA-1</span>
<span class="co">#&gt; CAACCTCGTAAACCTC-1</span>
<span class="co">#&gt; GTACTTTAGTATCGAA-1</span>
<span class="co">#&gt; GTAGTCACAGGGTACA-1</span>
<span class="co">#&gt; TGCACCTCAGTGGGAT-1</span>
<span class="co">#&gt; ATGGGAGTCGTCCAGG-1</span>
<span class="co">#&gt; AGGGTGAGTTCAGCGC-1</span>
<span class="co">#&gt; ATCGAGTTCCTCAATT-1</span>
<span class="co">#&gt; AACTTTCTCTCCCTGA-1</span>
<span class="co">#&gt; TGCACCTGTAAAGTCA-1</span></code></pre></div>
<p>We will now select some pCSCs from the 2nd sample, in the same way.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">pCSCsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>, <span class="cn">NULL</span>,
                      <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-25-1.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>In my version of the t-SNE for this sample, I selected cells below -14 on the Y2 axis due to their exceptionally high transcription levels, and because these cells were predicted to have high stemness, including high expression of the stemness markers that we found using the stemness correlations, as shown in previous plots.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">found_pos2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Y2</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">40</span><span class="op">)</span>
<span class="co">#found_pos2 &lt;- c(186, 266, 560, 1213, 2108, 2490, 3474, 3604, 4071, 4141, 5038, 5064, 5364) # these are the cells I selected</span>

<span class="co"># reorder by total transcription</span>
<span class="va">found_pos2_ordered</span> <span class="op">&lt;-</span>
    <span class="va">found_pos2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos2</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Highlighted cells in black:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>,
                      <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="va">found_pos2_ordered</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-27-1.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>And with cell labels:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>,
                      <span class="va">labelPositions</span> <span class="op">&lt;-</span> <span class="va">found_pos2_ordered</span>, <span class="cn">NULL</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-28-1.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>Store the pCSC group for this sample for later use, and show the cell names:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_pCSC_group_2_SC4.LB17009</span> <span class="op">&lt;-</span> <span class="va">found_pos2_ordered</span>

<span class="va">CDX_pCSC_group_2_SC4.LB17009_df</span> <span class="op">&lt;-</span>
    <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos2_ordered</span>,<span class="op">]</span>
<span class="va">CDX_pCSC_group_2_SC4.LB17009_cell_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">CDX_pCSC_group_2_SC4.LB17009_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="va">CDX_pCSC_group_2_SC4.LB17009_cell_names</span>, sep <span class="op">=</span> <span class="st">'\n'</span><span class="op">)</span>
<span class="co">#&gt; GCGCCAAGTCAAACTC-1</span>
<span class="co">#&gt; GTCACGGCAGACGCTC-1</span>
<span class="co">#&gt; TGAGCATCACTCAGGC-1</span>
<span class="co">#&gt; TTAGGACGTTTGTTGG-1</span>
<span class="co">#&gt; CGACTTCCAGGTCCAC-1</span>
<span class="co">#&gt; CGTTCTGGTACGCACC-1</span>
<span class="co">#&gt; ACTATCTCACCAGGTC-1</span>
<span class="co">#&gt; AATCGGTTCACCAGGC-1</span>
<span class="co">#&gt; ATTGGACGTCGCTTTC-1</span>
<span class="co">#&gt; TGAGGGAGTGTTCGAT-1</span>
<span class="co">#&gt; GGAAAGCGTAGCGTAG-1</span>
<span class="co">#&gt; AAGACCTTCTCGGACG-1</span>
<span class="co">#&gt; GTCTCGTTCGATCCCT-1</span></code></pre></div>
<p>We could have done something similar, automatically, by selecting the top n genes using total transcription as follows. However, I prefer not to do this, as it’s not taking into consideration the topology of the t-SNE or the stemness scores. It’s better to look at all of this information in combination.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co"># take the top 10 genes automatically</span>

<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">pCSCsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="va">found_pos3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Total_transcription</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span>
<span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos3</span>,<span class="fl">3</span><span class="op">]</span>
<span class="co">#&gt;  [1] 74861 65951 64345 57190 55632 55151 52247 50957 50054 49136</span>
<span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>,
                      <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="va">found_pos3</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># selected cells in black</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-30-1.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions &lt;- found_pos3, NULL, highestOnTop &lt;- TRUE) # labelled cells</a></code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-30-2.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>And now, automatic selection for the second sample:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span>

<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">pCSCsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="va">found_pos4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Total_transcription</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span>
<span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">found_pos4</span>,<span class="fl">3</span><span class="op">]</span>
<span class="co">#&gt;  [1] 68032 64060 62809 57939 57417 55718 54833 52730 52100 50286</span>
<span class="fu"><a href="../reference/plot_scores_with_grid.html">plot_scores_with_grid</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>,
                      <span class="va">run_name</span>, <span class="st">"Total_transcription"</span>,
                      <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">cohort_index</span><span class="op">]</span>, <span class="va">outDir</span>, <span class="cn">NULL</span>,
                      <span class="va">colourPositions</span> <span class="op">&lt;-</span> <span class="va">found_pos4</span>, <span class="va">highestOnTop</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># black cells</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-31-1.png" width="1152"></p>
<pre><code><a href="../reference/plot_scores_with_grid.html">#&gt; NULL
plot_scores_with_grid(CDX_tSNE_list_with_annotations[[cohort_index]][[perpl_index]], "Y1", "Y2",
                      run_name, "Total_transcription",
                      names(CDX_tSNE_list_with_annotations)[cohort_index], outDir,
                      labelPositions &lt;- found_pos4, NULL, highestOnTop &lt;- TRUE) # cell labels</a></code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-31-2.png" width="1152"></p>
<pre><code>#&gt; NULL</code></pre>
<p>… but looking at the t-SNE, I would refine it to the top 8, as this is more consistent with the tip of the t-SNE shape.</p>
</div>
<div id="step-7-preferential-expression-plots-forselected-pcsc-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#step-7-preferential-expression-plots-forselected-pcsc-groups" class="anchor"></a>Step 7: Preferential Expression Plots ForSelected pCSC Groups</h3>
<p>Now that we have some groups of pCSCs selected for both samples, we’ll look at the preferential expression in these cells compared to all of the cells in the cell populations they were selected from.</p>
<p>Some info needs to be prepared for preferential expression plot of HUGO gene groups in the pCSCs (e.g. families of related genes, or components of a molecular complex).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">all_HUGO_groups_mordered</span><span class="op">)</span> <span class="co"># automatically available in this package - you can get the most recent info from the HUGO website</span>
<span class="co">#&gt; $`1-acylglycerol-3-phosphate O-acyltransferases`</span>
<span class="co">#&gt;  [1] "AGPAT1" "LPCAT1" "LPCAT2" "AGPAT2" "AGPAT3" "AGPAT4" "AGPAT5" "GPAT4" </span>
<span class="co">#&gt;  [9] "LPCAT4" "LCLAT1" "GPAT3" </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`3-hydroxyacyl-CoA dehydratase family`</span>
<span class="co">#&gt; [1] "HACD1" "HACD2" "HACD3" "HACD4"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`5-hydroxytryptamine receptors, G protein-coupled`</span>
<span class="co">#&gt;  [1] "HTR1A"  "HTR1B"  "HTR1D"  "HTR1E"  "HTR1F"  "HTR2A"  "HTR2B"  "HTR2C" </span>
<span class="co">#&gt;  [9] "HTR4"   "HTR5A"  "HTR5BP" "HTR6"   "HTR7"  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`5-hydroxytryptamine receptors, ionotropic `</span>
<span class="co">#&gt; [1] "HTR3A" "HTR3B" "HTR3C" "HTR3D" "HTR3E"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`5'-nucleotidases`</span>
<span class="co">#&gt; [1] "NT5C"   "NT5C1A" "NT5C1B" "NT5C2"  "NT5C3A" "NT5E"   "NT5M"  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`5S ribosomal RNAs`</span>
<span class="co">#&gt;  [1] "RNA5S1"  "RNA5S10" "RNA5S11" "RNA5S12" "RNA5S13" "RNA5S14" "RNA5S15"</span>
<span class="co">#&gt;  [8] "RNA5S16" "RNA5S17" "RNA5S2"  "RNA5S3"  "RNA5S4"  "RNA5S5"  "RNA5S6" </span>
<span class="co">#&gt; [15] "RNA5S7"  "RNA5S8"  "RNA5S9"</span>

<span class="va">cohort_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">CDX_transcript_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>

<span class="co"># Create a list of HUGO gene groups, with each group containing the position of each group member in the expression matrix columns </span>
<span class="va">CDX_all_HUGO_groups_positions</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="../reference/map_gene_positions_to_HUGO_gene_groups.html">map_gene_positions_to_HUGO_gene_groups</a></span><span class="op">(</span><span class="va">CDX_transcript_names</span>, <span class="va">all_HUGO_groups_mordered</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">CDX_all_HUGO_groups_positions</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">saveDir</span>, <span class="st">"CDX_all_HUGO_group_positions.RData"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#load(paste0(saveDir, "CDX_all_HUGO_group_positions.RData"))</span>

<span class="co"># Subset to use HUGO gene groups that have two or more members in the gene expression matrix</span>
<span class="va">CDX_all_HUGO_groups_positions_two_or_more</span> <span class="op">&lt;-</span>
    <span class="va">CDX_all_HUGO_groups_positions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">CDX_all_HUGO_groups_positions</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="va">gene_group_positions_to_use</span> <span class="op">&lt;-</span> <span class="va">CDX_all_HUGO_groups_positions_two_or_more</span></code></pre></div>
<p>To run the preferential expression plots, set some parameters:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_pCSC_groups_to_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">CDX_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'pCSC_group_1_SC4_Talazoparib.LB17011'</span>   <span class="co"># pCSC group name</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_group_1_SC4_Talazoparib.LB17011</span> <span class="co"># pCSC group positions</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span>                                        <span class="co"># matrix position in CDX_tRawCounts_list</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name'</span>, <span class="st">'pCSC group positions'</span>,
                                          <span class="st">'Expression matrix list index'</span><span class="op">)</span>
<span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_info_list_tmp</span>

<span class="va">CDX_pCSC_groups_info_list_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'pCSC_group_2_SC4.LB17009'</span>          <span class="co"># pCSC group name</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_group_2_SC4.LB17009</span>        <span class="co"># pCSC group positions</span>
<span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># position in CDX_tRawCounts_list # matrix position in CDX_tRawCounts_list</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_pCSC_groups_info_list_tmp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pCSC group name'</span>, <span class="st">'pCSC group positions'</span>,
                                          <span class="st">'Expression matrix list index'</span><span class="op">)</span>
<span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_info_list_tmp</span>

<span class="va">tissue_type</span>       <span class="op">&lt;&lt;-</span> <span class="st">"CDX"</span>
<span class="va">frag_count_method</span> <span class="op">&lt;-</span> <span class="st">'raw'</span>  <span class="co"># expression measurement type, but just a name, can be anything</span>

<span class="va">no_ribo</span>                                      <span class="op">&lt;-</span> <span class="cn">TRUE</span>  <span class="co"># set to TRUE to remove ribosomal transcripts</span>
                                                      <span class="co"># from some of the expression bar charts</span>
<span class="va">separate_row_for_high_expression_transcripts</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co"># set to TRUE to print highest expression bars</span>
                                                      <span class="co"># on a seperate row</span>
<span class="va">number_of_graph_rows_stacked_transcripts</span>     <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">number_of_graph_rows_stacked_families</span>        <span class="op">&lt;-</span> <span class="fl">1</span></code></pre></div>
<p>Now that preparations are complete, we can <strong>print the preferential expression plots</strong>.</p>
<p>(If the images generated look small when viewing them in the web pages, you could try opening each of them in a new tab e.g. by using a right mouse click.)</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CDX_all_results_ranks_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># preferential expression ranks will be stored in here</span>

<span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_pCSC_groups_to_run</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    
    <span class="va">expression_matrix</span>    <span class="op">&lt;-</span> <span class="cn">NULL</span>
    <span class="va">pCSC_group_pos</span>       <span class="op">&lt;-</span> <span class="cn">NULL</span>
    
    <span class="va">this_matrix_index</span>    <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
    <span class="va">expression_matrix</span>    <span class="op">&lt;-</span> <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">this_matrix_index</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_CDX_name</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tRawCounts_list</span><span class="op">[</span><span class="va">this_matrix_index</span><span class="op">]</span><span class="op">)</span>
    <span class="va">this_pCSC_group_name</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">this_pCSC_group_name</span>, <span class="st">"; "</span>, <span class="va">this_CDX_name</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>

    <span class="va">CDX_tumour_only_positions</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span> <span class="co"># can be restricted to a subset,</span>
                                                           <span class="co"># not necessary for CDXs</span>

    <span class="co"># Subset transcription info for pCSC group</span>
    <span class="va">pCSC_group_pos</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>

    <span class="co"># Generate graphs</span>
    <span class="va">CDX_all_results_ranks_list</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/print_preferential_expression_bar_graphs.html">print_preferential_expression_bar_graphs</a></span><span class="op">(</span>
        <span class="va">pCSC_group_name</span>                    <span class="op">&lt;-</span> <span class="va">this_pCSC_group_name</span>,
        <span class="va">pCSC_group_positions</span>               <span class="op">&lt;-</span> <span class="va">pCSC_group_pos</span>,
        <span class="va">HUGO_groups_sets_of_columnNumbers</span>  <span class="op">&lt;-</span> <span class="va">gene_group_positions_to_use</span>,
        <span class="va">tExprMatr</span>                          <span class="op">&lt;-</span> <span class="va">expression_matrix</span>,
        <span class="va">tumour_only_positions</span>              <span class="op">&lt;-</span> <span class="va">CDX_tumour_only_positions</span>,
        <span class="va">this_patient_tumour_only_positions</span> <span class="op">&lt;-</span> <span class="va">CDX_tumour_only_positions</span>,
        <span class="va">outDir</span>                             <span class="op">&lt;-</span> <span class="va">prefExprDirs</span><span class="op">[[</span><span class="op">(</span><span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>,
        <span class="va">tissue_type</span>,
        <span class="va">family_type_name</span>                   <span class="op">&lt;-</span> <span class="st">'HUGO'</span>,
        <span class="va">patient_source_name</span>                <span class="op">&lt;-</span> <span class="va">this_CDX_name</span>,
        <span class="va">frag_count_method</span>,
        <span class="va">no_ribo</span>,
        <span class="va">separate_row_for_high_expression_transcripts</span>,
        <span class="va">number_of_graph_rows_stacked_transcripts</span>,
        <span class="va">number_of_graph_rows_stacked_families</span>,
        <span class="va">all_HUGO_abbreviations_and_fullnames</span>,
        <span class="va">dataDir</span>
    <span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">)</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">this_pCSC_group_name</span> <span class="co">#names(CDX_pCSC_groups)[r]</span>
<span class="op">}</span>
<span class="co">#&gt; pCSC_group_1_SC4_Talazoparib.LB17011; GSM4104161_SC4_Talazoparib.LB17011</span>
<span class="co">#&gt; Output directory = 'results/SCLC_CDXs_project/GSM4104161_SC4_Talazoparib.LB17011/preferential_expression/'</span>
<span class="co">#&gt; Printing: Single transcripts barcharts (triplets)</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-1.png" width="2880"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-2.png" width="2880"><img src="tutorial_1_files/figure-html/unnamed-chunk-34-3.png" width="2880"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-4.png" width="2880"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-5.png" width="2880"></p>
<pre><code>#&gt; Expression bar charts completed
#&gt; 
#&gt; pCSC_group_2_SC4.LB17009; GSM4104152_SC4.LB17009
#&gt; Output directory = 'results/SCLC_CDXs_project/GSM4104152_SC4.LB17009/preferential_expression/'
#&gt; Printing: Single transcripts barcharts (triplets)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-6.png" width="2880"><img src="tutorial_1_files/figure-html/unnamed-chunk-34-7.png" width="2880"></p>
<pre><code>#&gt; Printing: Single transcripts barcharts (no stacking)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-8.png" width="2880"><img src="tutorial_1_files/figure-html/unnamed-chunk-34-9.png" width="2880"></p>
<pre><code>#&gt; Printing: Single transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-10.png" width="2880"></p>
<pre><code>#&gt; Printing: Family transcripts barchart (stacked)</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-34-11.png" width="2880"><img src="tutorial_1_files/figure-html/unnamed-chunk-34-12.png" width="2880"></p>
<pre><code><a href="https://rdrr.io/r/base/save.html">#&gt; Expression bar charts completed
save(CDX_all_results_ranks_list, file = paste0(saveDir, "CDX_all_results_ranks_list_raw.RData"))
#load(paste0(saveDir, "CDX_all_results_ranks_list_raw.RData"))</a></code></pre>
<p>We need to print the t-SNEs with the cell numbers of the pCSC group used in the stacked transcripts graphs. (The tif file is written to the results/tSNEs directory under the relevant samples directory, and ends in “n”_cells-labelled.tif, where n is the number of cells.) This can be used to compare these preferential expression graphs with the t-SNEs, to understand the pCSCs in great detail.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_pCSC_groups_to_run</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">exprMatr_list_index</span>  <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_pCSC_group_name</span> <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_pCSC_group_pos</span>  <span class="op">&lt;-</span> <span class="va">CDX_pCSC_groups_to_run</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_perplexity_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="op">(</span><span class="va">exprMatr_list_index</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>
    <span class="va">this_sample_name</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[[</span><span class="op">(</span><span class="va">exprMatr_list_index</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>
    <span class="va">this_tSNE_results_with_annotations</span> <span class="op">&lt;-</span>
        <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="op">(</span><span class="va">exprMatr_list_index</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">this_tSNE_results_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span>
        <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"#"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="../reference/plot_scores.html">plot_scores</a></span><span class="op">(</span><span class="va">this_tSNE_results_with_annotations</span>, <span class="st">"Y1"</span>, <span class="st">"Y2"</span>, <span class="va">this_pCSC_group_name</span>,
                <span class="st">"Total_transcription"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_tSNE_list_with_annotations</span><span class="op">)</span><span class="op">[</span><span class="va">exprMatr_list_index</span><span class="op">]</span>,
                <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">exprMatr_list_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">this_pCSC_group_pos</span><span class="op">)</span>
    
    <span class="va">outFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">this_pCSC_group_name</span>, <span class="st">'-Total_transcription_Y1vsY2.tif'</span><span class="op">)</span> 
    <span class="va">renamed_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">this_pCSC_group_name</span>, <span class="st">"-Total_transcription_Y1vsY2-"</span>,
                           <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">this_pCSC_group_pos</span><span class="op">)</span>, <span class="st">"_cells-labelled.tif"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span><span class="va">outFile</span>, <span class="va">renamed_file</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-35-1.png" width="1152"><img src="tutorial_1_files/figure-html/unnamed-chunk-35-2.png" width="1152"></p>
<p>These are the <strong>top ranked preferential expressed genes and gene groups</strong> from the preferential expression plots, for each sample.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\n## Rankings for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">" ##\n\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Top 100 transcripts for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nTop 100 families for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"\nTop 100 transcripts in ranked families for "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="st">":\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">CDX_all_results_ranks_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for pCSC_group_1_SC4_Talazoparib.LB17011 ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for pCSC_group_1_SC4_Talazoparib.LB17011:</span>
<span class="co">#&gt;   [1] "RPLP1"    "MT-CO1"   "MT-CO3"   "EEF1A1"   "MT-CO2"   "RPL10"   </span>
<span class="co">#&gt;   [7] "RPS2"     "MALAT1"   "RPL13A"   "RPS19"    "RPS18"    "MT-ND4"  </span>
<span class="co">#&gt;  [13] "RPS8"     "RPL3"     "RPS3"     "RPS6"     "RPL41"    "RPL18A"  </span>
<span class="co">#&gt;  [19] "RPS4X"    "RPS27"    "MT-CYB"   "RPL35A"   "RPS12"    "RPL34"   </span>
<span class="co">#&gt;  [25] "FTL"      "RPL7"     "RPS13"    "RPS15A"   "RPL5"     "RPL19"   </span>
<span class="co">#&gt;  [31] "RPS27A"   "RPS14"    "FTH1"     "RPS23"    "RPS15"    "RPS7"    </span>
<span class="co">#&gt;  [37] "RPLP0"    "RPLP2"    "GNB2L1"   "RPL26"    "RPL18"    "GAPDH"   </span>
<span class="co">#&gt;  [43] "RPL6"     "RPS5"     "SEC11C"   "RPL27A"   "RPL8"     "RPL13"   </span>
<span class="co">#&gt;  [49] "TMSB4X"   "RPL15"    "RPL31"    "MT-ND2"   "RPL28"    "RPL30"   </span>
<span class="co">#&gt;  [55] "RPL23A"   "RPL9"     "RPL7A"    "RPS9"     "RPS16"    "ACTB"    </span>
<span class="co">#&gt;  [61] "RPS24"    "RPL36"    "RPL37A"   "RPL24"    "ACTG1"    "RPS28"   </span>
<span class="co">#&gt;  [67] "PTMA"     "YBX1"     "RPL39"    "RPS25"    "RPL21"    "RPS3A"   </span>
<span class="co">#&gt;  [73] "TFF3"     "RPL32"    "RPS20"    "H3F3A"    "MT-ND1"   "RPL12"   </span>
<span class="co">#&gt;  [79] "RPL10A"   "RPL4"     "MT-ATP6"  "HNRNPA1"  "RPL17"    "SMS"     </span>
<span class="co">#&gt;  [85] "RPL35"    "BTF3"     "KRT18"    "TUBA1B"   "EEF2"     "RPL11"   </span>
<span class="co">#&gt;  [91] "MARCKSL1" "RPL37"    "SLC25A6"  "RPS11"    "UBA52"    "TUBA1A"  </span>
<span class="co">#&gt;  [97] "RPS29"    "EIF1"     "RPL29"    "NACA"    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for pCSC_group_1_SC4_Talazoparib.LB17011:</span>
<span class="co">#&gt;   [1] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;   [2] "S ribosomal proteins"                                                        </span>
<span class="co">#&gt;   [3] "L ribosomal proteins"                                                        </span>
<span class="co">#&gt;   [4] "Ferritin subunits"                                                           </span>
<span class="co">#&gt;   [5] "Mitochondrial complex IV: cytochrome c oxidase subunits"                     </span>
<span class="co">#&gt;   [6] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;   [7] "NADH:ubiquinone oxidoreductase core subunits"                                </span>
<span class="co">#&gt;   [8] "Y-box binding proteins"                                                      </span>
<span class="co">#&gt;   [9] "Small nucleolar RNA protein coding host genes"                               </span>
<span class="co">#&gt;  [10] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [11] "Actins"                                                                      </span>
<span class="co">#&gt;  [12] "Seven-beta-strand methyltransferase motif containing"                        </span>
<span class="co">#&gt;  [13] "U11 small nuclear ribonucleoprotein"                                         </span>
<span class="co">#&gt;  [14] "Heat shock 90kDa proteins"                                                   </span>
<span class="co">#&gt;  [15] "Peroxiredoxins"                                                              </span>
<span class="co">#&gt;  [16] "INHAT complex"                                                               </span>
<span class="co">#&gt;  [17] "Jun transcription factor family"                                             </span>
<span class="co">#&gt;  [18] "Profilins"                                                                   </span>
<span class="co">#&gt;  [19] "Heterogeneous nuclear ribonucleoproteins"                                    </span>
<span class="co">#&gt;  [20] "Sm spliceosomal proteins"                                                    </span>
<span class="co">#&gt;  [21] "GBAF complex"                                                                </span>
<span class="co">#&gt;  [22] "SCF complex"                                                                 </span>
<span class="co">#&gt;  [23] "Eukaryotic translation initiation factor 3"                                  </span>
<span class="co">#&gt;  [24] "Myosin light chains, class 1"                                                </span>
<span class="co">#&gt;  [25] "14-3-3 phospho-serine/phospho-threonine binding proteins"                    </span>
<span class="co">#&gt;  [26] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [27] "Brain expressed X-linked family"                                             </span>
<span class="co">#&gt;  [28] "SET complex"                                                                 </span>
<span class="co">#&gt;  [29] "U11/U12 di-snRNP"                                                            </span>
<span class="co">#&gt;  [30] "Histidine triad superfamily"                                                 </span>
<span class="co">#&gt;  [31] "Ubiquitin C-terminal hydrolases"                                             </span>
<span class="co">#&gt;  [32] "Canonical high mobility group"                                               </span>
<span class="co">#&gt;  [33] "NADH:ubiquinone oxidoreductase supernumerary subunits"                       </span>
<span class="co">#&gt;  [34] "PBAF complex"                                                                </span>
<span class="co">#&gt;  [35] "Microsomal glutathione S-transferases"                                       </span>
<span class="co">#&gt;  [36] "Serine and arginine rich splicing factors"                                   </span>
<span class="co">#&gt;  [37] "Tubulins"                                                                    </span>
<span class="co">#&gt;  [38] "Spliceosomal A complex"                                                      </span>
<span class="co">#&gt;  [39] "Voltage dependent anion channels"                                            </span>
<span class="co">#&gt;  [40] "Stathmins"                                                                   </span>
<span class="co">#&gt;  [41] "Coiled-coil-helix-coiled-coil-helix domain containing proteins"              </span>
<span class="co">#&gt;  [42] "Prefoldin subunits"                                                          </span>
<span class="co">#&gt;  [43] "Chaperonins"                                                                 </span>
<span class="co">#&gt;  [44] "Oligosaccharyltransferase complex subunits"                                  </span>
<span class="co">#&gt;  [45] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [46] "Superoxide dismutases"                                                       </span>
<span class="co">#&gt;  [47] "Spliceosomal E complex"                                                      </span>
<span class="co">#&gt;  [48] "Ragulator complex "                                                          </span>
<span class="co">#&gt;  [49] "Signal recognition particle"                                                 </span>
<span class="co">#&gt;  [50] "EMSY complex"                                                                </span>
<span class="co">#&gt;  [51] "ASAP complex"                                                                </span>
<span class="co">#&gt;  [52] "Adaptor related protein complex 2"                                           </span>
<span class="co">#&gt;  [53] "Dynein 2 complex subunits"                                                   </span>
<span class="co">#&gt;  [54] "Translocase of outer mitochondrial membrane complex"                         </span>
<span class="co">#&gt;  [55] "SIN3 histone deacetylase complex subunits"                                   </span>
<span class="co">#&gt;  [56] "Actin related protein 2/3 complex subunits"                                  </span>
<span class="co">#&gt;  [57] "Proteasome"                                                                  </span>
<span class="co">#&gt;  [58] "U1 small nuclear ribonucleoprotein"                                          </span>
<span class="co">#&gt;  [59] "NURF complex"                                                                </span>
<span class="co">#&gt;  [60] "Exon junction complex"                                                       </span>
<span class="co">#&gt;  [61] "Enolases"                                                                    </span>
<span class="co">#&gt;  [62] "Histone deacetylases, class I"                                               </span>
<span class="co">#&gt;  [63] "Mitochondrial RNase P complex"                                               </span>
<span class="co">#&gt;  [64] "BAF complex"                                                                 </span>
<span class="co">#&gt;  [65] "Signal transduction and activation of RNA metabolism family"                 </span>
<span class="co">#&gt;  [66] "Cyclophilin peptidylprolyl isomerases"                                       </span>
<span class="co">#&gt;  [67] "Metallothioneins"                                                            </span>
<span class="co">#&gt;  [68] "Polycomb repressive complex 2"                                               </span>
<span class="co">#&gt;  [69] "LSm proteins"                                                                </span>
<span class="co">#&gt;  [70] "Dynein 1 complex subunits"                                                   </span>
<span class="co">#&gt;  [71] "CDGSH iron sulfur domain containing"                                         </span>
<span class="co">#&gt;  [72] "ANP32 acidic nuclear phosphoproteins"                                        </span>
<span class="co">#&gt;  [73] "Heparin binding growth factor family"                                        </span>
<span class="co">#&gt;  [74] "SF3b complex"                                                                </span>
<span class="co">#&gt;  [75] "Transmembrane p24 trafficking proteins"                                      </span>
<span class="co">#&gt;  [76] "Zinc fingers PARP-type"                                                      </span>
<span class="co">#&gt;  [77] "U3 small nucleolar ribonucleoprotein"                                        </span>
<span class="co">#&gt;  [78] "Membrane associated progesterone receptor family"                            </span>
<span class="co">#&gt;  [79] "Parvulins"                                                                   </span>
<span class="co">#&gt;  [80] "Mitochondrial complex II: succinate dehydrogenase subunits"                  </span>
<span class="co">#&gt;  [81] "R2TP complex"                                                                </span>
<span class="co">#&gt;  [82] "Protein phosphatases"                                                        </span>
<span class="co">#&gt;  [83] "BTG/Tob family"                                                              </span>
<span class="co">#&gt;  [84] "WRAD complex"                                                                </span>
<span class="co">#&gt;  [85] "Nuclear factor I family"                                                     </span>
<span class="co">#&gt;  [86] "Granins"                                                                     </span>
<span class="co">#&gt;  [87] "NuRD complex"                                                                </span>
<span class="co">#&gt;  [88] "Insulin like growth factor binding proteins"                                 </span>
<span class="co">#&gt;  [89] "Tropomyosins"                                                                </span>
<span class="co">#&gt;  [90] "Heat shock 70kDa proteins"                                                   </span>
<span class="co">#&gt;  [91] "LYR motif containing"                                                        </span>
<span class="co">#&gt;  [92] "U2 small nuclear ribonucleoprotein "                                         </span>
<span class="co">#&gt;  [93] "minichromosome maintenance 2-7 complex"                                      </span>
<span class="co">#&gt;  [94] "SF3a complex"                                                                </span>
<span class="co">#&gt;  [95] "Pseudoautosomal region 1"                                                    </span>
<span class="co">#&gt;  [96] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [97] "Isocitrate dehydrogenases"                                                   </span>
<span class="co">#&gt;  [98] "G protein subunits alpha, group s"                                           </span>
<span class="co">#&gt;  [99] "Spliceosomal P complex"                                                      </span>
<span class="co">#&gt; [100] "Mitochondrial ribosomal proteins"                                            </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for pCSC_group_1_SC4_Talazoparib.LB17011:</span>
<span class="co">#&gt;   [1] "YBX1"     "MT-ATP6"  "SMS"      "ALDOA"    "SET"      "ACTB"    </span>
<span class="co">#&gt;   [7] "JUN"      "MYL6"     "PFN1"     "UCHL1"    "HINT1"    "MT-CYB"  </span>
<span class="co">#&gt;  [13] "STMN1"    "FTL"      "SOD1"     "SKP1"     "FTH1"     "ACTG1"   </span>
<span class="co">#&gt;  [19] "MT-ND4"   "CHCHD2"   "U2AF1"    "BEX1"     "KHDRBS1"  "HSP90AB1"</span>
<span class="co">#&gt;  [25] "HSP90AA1" "ENO1"     "MGST1"    "HNRNPA1"  "RBX1"     "MT-CO1"  </span>
<span class="co">#&gt;  [31] "PFDN5"    "NHP2"     "PRDX2"    "MGST3"    "VDAC2"    "MT-CO3"  </span>
<span class="co">#&gt;  [37] "AP2M1"    "MT-CO2"   "PARP1"    "RBBP4"    "RNPS1"    "BEX2"    </span>
<span class="co">#&gt;  [43] "TUBA1B"   "PRDX1"    "YWHAE"    "SRP9"     "PSIP1"    "SAP18"   </span>
<span class="co">#&gt;  [49] "ANP32B"   "HDAC2"    "LAMTOR5"  "TUBA1A"   "HSD17B10" "MT-ND2"  </span>
<span class="co">#&gt;  [55] "TRMT10C"  "SRP14"    "DYNLL1"   "PHPT1"    "DYNLT1"   "VDAC3"   </span>
<span class="co">#&gt;  [61] "PPIA"     "YWHAQ"    "SNRNP70"  "VDAC1"    "PFN2"     "TOMM7"   </span>
<span class="co">#&gt;  [67] "PIN1"     "SNRPD2"   "SNRPC"    "HMGB1"    "APEX1"    "ANP32A"  </span>
<span class="co">#&gt;  [73] "SNRPE"    "MAGOH"    "NOP10"    "HDAC1"    "AP2S1"    "PRDX5"   </span>
<span class="co">#&gt;  [79] "FBL"      "HMGN2"    "EIF3L"    "DPY30"    "NENF"     "HSP90B1" </span>
<span class="co">#&gt;  [85] "MT-ND1"   "YWHAB"    "EIF3I"    "RBBP7"    "JUNB"     "NME1"    </span>
<span class="co">#&gt;  [91] "RBM8A"    "CISD2"    "UQCRH"    "MT2A"     "TUBB"     "CISD1"   </span>
<span class="co">#&gt;  [97] "IGFBP5"   "SNRPG"    "SNRPD1"   "ARPC3"   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Rankings for pCSC_group_2_SC4.LB17009 ##</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts for pCSC_group_2_SC4.LB17009:</span>
<span class="co">#&gt;   [1] "RPLP1"    "MALAT1"   "MT-CO1"   "MT-CO3"   "EEF1A1"   "MT-CO2"  </span>
<span class="co">#&gt;   [7] "RPL10"    "RPS2"     "RPS19"    "RPL13A"   "RPS27"    "RPS18"   </span>
<span class="co">#&gt;  [13] "RPL41"    "RPS8"     "MT-ND4"   "RPS3"     "FTL"      "RPL34"   </span>
<span class="co">#&gt;  [19] "RPL3"     "FTH1"     "RPL18A"   "RPS12"    "RPS4X"    "RPL35A"  </span>
<span class="co">#&gt;  [25] "GAPDH"    "RPLP2"    "RPS6"     "RPS15A"   "RPS14"    "MT-CYB"  </span>
<span class="co">#&gt;  [31] "RPS15"    "RPL37A"   "RPL7"     "RPS23"    "RPS7"     "TMSB4X"  </span>
<span class="co">#&gt;  [37] "RPL19"    "RPS27A"   "RPS28"    "RPL39"    "RPLP0"    "RPS13"   </span>
<span class="co">#&gt;  [43] "RPL36"    "RPL5"     "RPL26"    "RPL28"    "RPL18"    "SEC11C"  </span>
<span class="co">#&gt;  [49] "RPL31"    "RPL27A"   "RPL13"    "RPS5"     "GNB2L1"   "ACTB"    </span>
<span class="co">#&gt;  [55] "RPL6"     "PTMA"     "RPL8"     "MT-ND2"   "RPS16"    "RPL30"   </span>
<span class="co">#&gt;  [61] "RPL15"    "RPL23A"   "TUBA1B"   "ACTG1"    "RPL9"     "RPS24"   </span>
<span class="co">#&gt;  [67] "RPS25"    "RPS20"    "RPL7A"    "RPS9"     "RPL35"    "MT-ND1"  </span>
<span class="co">#&gt;  [73] "RPL37"    "RPL24"    "RPL21"    "YBX1"     "RPL32"    "TUBA1A"  </span>
<span class="co">#&gt;  [79] "H3F3A"    "EIF1"     "RPS3A"    "JUN"      "MIF"      "RPL12"   </span>
<span class="co">#&gt;  [85] "RPS29"    "MT-ATP6"  "RPL10A"   "HNRNPA1"  "TMSB10"   "RPL4"    </span>
<span class="co">#&gt;  [91] "MT-ND3"   "RPS11"    "FAU"      "UBA52"    "BTF3"     "KRT18"   </span>
<span class="co">#&gt;  [97] "RPL38"    "RPL11"    "RPL17"    "HIST1H4C"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 families for pCSC_group_2_SC4.LB17009:</span>
<span class="co">#&gt;   [1] "Mitochondrially encoded protein coding genes"                                </span>
<span class="co">#&gt;   [2] "S ribosomal proteins"                                                        </span>
<span class="co">#&gt;   [3] "L ribosomal proteins"                                                        </span>
<span class="co">#&gt;   [4] "Ferritin subunits"                                                           </span>
<span class="co">#&gt;   [5] "Mitochondrial complex IV: cytochrome c oxidase subunits"                     </span>
<span class="co">#&gt;   [6] "NADH:ubiquinone oxidoreductase core subunits"                                </span>
<span class="co">#&gt;   [7] "Mitochondrial complex V: ATP synthase subunits"                              </span>
<span class="co">#&gt;   [8] "Jun transcription factor family"                                             </span>
<span class="co">#&gt;   [9] "Small nucleolar RNA protein coding host genes"                               </span>
<span class="co">#&gt;  [10] "Actins"                                                                      </span>
<span class="co">#&gt;  [11] "Y-box binding proteins"                                                      </span>
<span class="co">#&gt;  [12] "Mitochondrial complex III: ubiquinol-cytochrome c reductase complex subunits"</span>
<span class="co">#&gt;  [13] "Heat shock 90kDa proteins"                                                   </span>
<span class="co">#&gt;  [14] "Peroxiredoxins"                                                              </span>
<span class="co">#&gt;  [15] "U11 small nuclear ribonucleoprotein"                                         </span>
<span class="co">#&gt;  [16] "Seven-beta-strand methyltransferase motif containing"                        </span>
<span class="co">#&gt;  [17] "Aldolases"                                                                   </span>
<span class="co">#&gt;  [18] "Sm spliceosomal proteins"                                                    </span>
<span class="co">#&gt;  [19] "Tubulins"                                                                    </span>
<span class="co">#&gt;  [20] "Heterogeneous nuclear ribonucleoproteins"                                    </span>
<span class="co">#&gt;  [21] "Profilins"                                                                   </span>
<span class="co">#&gt;  [22] "GBAF complex"                                                                </span>
<span class="co">#&gt;  [23] "Canonical high mobility group"                                               </span>
<span class="co">#&gt;  [24] "INHAT complex"                                                               </span>
<span class="co">#&gt;  [25] "Brain expressed X-linked family"                                             </span>
<span class="co">#&gt;  [26] "Histidine triad superfamily"                                                 </span>
<span class="co">#&gt;  [27] "Myosin light chains, class 1"                                                </span>
<span class="co">#&gt;  [28] "Fos transcription factor family"                                             </span>
<span class="co">#&gt;  [29] "SET complex"                                                                 </span>
<span class="co">#&gt;  [30] "Enolases"                                                                    </span>
<span class="co">#&gt;  [31] "Voltage dependent anion channels"                                            </span>
<span class="co">#&gt;  [32] "14-3-3 phospho-serine/phospho-threonine binding proteins"                    </span>
<span class="co">#&gt;  [33] "NADH:ubiquinone oxidoreductase supernumerary subunits"                       </span>
<span class="co">#&gt;  [34] "Serine and arginine rich splicing factors"                                   </span>
<span class="co">#&gt;  [35] "SCF complex"                                                                 </span>
<span class="co">#&gt;  [36] "Metallothioneins"                                                            </span>
<span class="co">#&gt;  [37] "Stathmins"                                                                   </span>
<span class="co">#&gt;  [38] "PBAF complex"                                                                </span>
<span class="co">#&gt;  [39] "Eukaryotic translation initiation factor 3"                                  </span>
<span class="co">#&gt;  [40] "Oligosaccharyltransferase complex subunits"                                  </span>
<span class="co">#&gt;  [41] "U11/U12 di-snRNP"                                                            </span>
<span class="co">#&gt;  [42] "Coiled-coil-helix-coiled-coil-helix domain containing proteins"              </span>
<span class="co">#&gt;  [43] "Ubiquitin C-terminal hydrolases"                                             </span>
<span class="co">#&gt;  [44] "Spliceosomal A complex"                                                      </span>
<span class="co">#&gt;  [45] "Microsomal glutathione S-transferases"                                       </span>
<span class="co">#&gt;  [46] "Spliceosomal E complex"                                                      </span>
<span class="co">#&gt;  [47] "Chaperonins"                                                                 </span>
<span class="co">#&gt;  [48] "Prefoldin subunits"                                                          </span>
<span class="co">#&gt;  [49] "Exon junction complex"                                                       </span>
<span class="co">#&gt;  [50] "ASAP complex"                                                                </span>
<span class="co">#&gt;  [51] "Dynein 2 complex subunits"                                                   </span>
<span class="co">#&gt;  [52] "Translocase of outer mitochondrial membrane complex"                         </span>
<span class="co">#&gt;  [53] "Superoxide dismutases"                                                       </span>
<span class="co">#&gt;  [54] "ANP32 acidic nuclear phosphoproteins"                                        </span>
<span class="co">#&gt;  [55] "U1 small nuclear ribonucleoprotein"                                          </span>
<span class="co">#&gt;  [56] "Signal recognition particle"                                                 </span>
<span class="co">#&gt;  [57] "Actin related protein 2/3 complex subunits"                                  </span>
<span class="co">#&gt;  [58] "H/ACA ribonucleoprotein complex"                                             </span>
<span class="co">#&gt;  [59] "Parvulins"                                                                   </span>
<span class="co">#&gt;  [60] "Cyclophilin peptidylprolyl isomerases"                                       </span>
<span class="co">#&gt;  [61] "EMSY complex"                                                                </span>
<span class="co">#&gt;  [62] "SF3b complex"                                                                </span>
<span class="co">#&gt;  [63] "Ragulator complex "                                                          </span>
<span class="co">#&gt;  [64] "SIN3 histone deacetylase complex subunits"                                   </span>
<span class="co">#&gt;  [65] "BAF complex"                                                                 </span>
<span class="co">#&gt;  [66] "LSm proteins"                                                                </span>
<span class="co">#&gt;  [67] "Adaptor related protein complex 2"                                           </span>
<span class="co">#&gt;  [68] "Proteasome"                                                                  </span>
<span class="co">#&gt;  [69] "Basic leucine zipper proteins"                                               </span>
<span class="co">#&gt;  [70] "BTG/Tob family"                                                              </span>
<span class="co">#&gt;  [71] "Signal transduction and activation of RNA metabolism family"                 </span>
<span class="co">#&gt;  [72] "Granins"                                                                     </span>
<span class="co">#&gt;  [73] "Dynein 1 complex subunits"                                                   </span>
<span class="co">#&gt;  [74] "Polycomb repressive complex 2"                                               </span>
<span class="co">#&gt;  [75] "Histone deacetylases, class I"                                               </span>
<span class="co">#&gt;  [76] "G protein subunits alpha, group s"                                           </span>
<span class="co">#&gt;  [77] "NURF complex"                                                                </span>
<span class="co">#&gt;  [78] "U3 small nucleolar ribonucleoprotein"                                        </span>
<span class="co">#&gt;  [79] "Transmembrane p24 trafficking proteins"                                      </span>
<span class="co">#&gt;  [80] "Membrane associated progesterone receptor family"                            </span>
<span class="co">#&gt;  [81] "Heparin binding growth factor family"                                        </span>
<span class="co">#&gt;  [82] "Tropomyosins"                                                                </span>
<span class="co">#&gt;  [83] "CDGSH iron sulfur domain containing"                                         </span>
<span class="co">#&gt;  [84] "WRAD complex"                                                                </span>
<span class="co">#&gt;  [85] "U2 small nuclear ribonucleoprotein "                                         </span>
<span class="co">#&gt;  [86] "Mitochondrial RNase P complex"                                               </span>
<span class="co">#&gt;  [87] "Heat shock 70kDa proteins"                                                   </span>
<span class="co">#&gt;  [88] "Small nucleolar RNA non-coding host genes"                                   </span>
<span class="co">#&gt;  [89] "Isocitrate dehydrogenases"                                                   </span>
<span class="co">#&gt;  [90] "Insulin like growth factor binding proteins"                                 </span>
<span class="co">#&gt;  [91] "Large subunit mitochondrial ribosomal proteins"                              </span>
<span class="co">#&gt;  [92] "LYR motif containing"                                                        </span>
<span class="co">#&gt;  [93] "KDEL endoplasmic reticulum protein retention receptors"                      </span>
<span class="co">#&gt;  [94] "NuRD complex"                                                                </span>
<span class="co">#&gt;  [95] "Protein disulfide isomerases"                                                </span>
<span class="co">#&gt;  [96] "Nuclear factor I family"                                                     </span>
<span class="co">#&gt;  [97] "Spliceosomal P complex"                                                      </span>
<span class="co">#&gt;  [98] "Mitochondrial ribosomal proteins"                                            </span>
<span class="co">#&gt;  [99] "Transmembrane BAX inhibitor motif containing"                                </span>
<span class="co">#&gt; [100] "Chromosomal passenger complex"                                               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Top 100 transcripts in ranked families for pCSC_group_2_SC4.LB17009:</span>
<span class="co">#&gt;   [1] "YBX1"     "MT-ATP6"  "ALDOA"    "SMS"      "JUN"      "ACTB"    </span>
<span class="co">#&gt;   [7] "FOS"      "PFN1"     "MYL6"     "SET"      "HINT1"    "STMN1"   </span>
<span class="co">#&gt;  [13] "ENO1"     "UCHL1"    "FTL"      "MT-CYB"   "FTH1"     "SOD1"    </span>
<span class="co">#&gt;  [19] "ACTG1"    "HSP90AA1" "CHCHD2"   "U2AF1"    "BEX1"     "SKP1"    </span>
<span class="co">#&gt;  [25] "MT-ND4"   "ANP32B"   "VDAC2"    "RBX1"     "KHDRBS1"  "HSP90AB1"</span>
<span class="co">#&gt;  [31] "TUBA1B"   "PIN1"     "MT-CO1"   "PRDX2"    "MGST3"    "HNRNPA1" </span>
<span class="co">#&gt;  [37] "MT-CO3"   "MT2A"     "MGST1"    "PRDX1"    "PFDN5"    "GNAS"    </span>
<span class="co">#&gt;  [43] "RNPS1"    "TUBA1A"   "BEX2"     "MT-CO2"   "VDAC3"    "JUNB"    </span>
<span class="co">#&gt;  [49] "DYNLL1"   "HMGB2"    "PPIA"     "SAP18"    "NHP2"     "SNRNP70" </span>
<span class="co">#&gt;  [55] "RBM8A"    "MT-ND2"   "TOMM7"    "YWHAE"    "SRP9"     "ANP32A"  </span>
<span class="co">#&gt;  [61] "HMGN2"    "HMGB1"    "DYNLT1"   "YWHAQ"    "SNRPG"    "SF3B5"   </span>
<span class="co">#&gt;  [67] "VDAC1"    "MAGOH"    "PFN2"     "PRDX5"    "AP2M1"    "MT-ND1"  </span>
<span class="co">#&gt;  [73] "SRP14"    "HSP90B1"  "TMEM258"  "SRSF3"    "MT1E"     "SNRPE"   </span>
<span class="co">#&gt;  [79] "SNRPC"    "PSIP1"    "SNRPD2"   "UQCRH"    "PIN4"     "RBBP4"   </span>
<span class="co">#&gt;  [85] "NOP10"    "DPY30"    "TUBB"     "YWHAH"    "HDAC2"    "FBL"     </span>
<span class="co">#&gt;  [91] "AP2S1"    "NENF"     "EIF4A3"   "ENO2"     "SNRPA"    "APEX1"   </span>
<span class="co">#&gt;  [97] "SNRPD1"   "LAMTOR5"  "BTG1"     "SRSF7"</span></code></pre></div>
</div>
<div id="step-8-t-snes-for-highest-ranked-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#step-8-t-snes-for-highest-ranked-genes" class="anchor"></a>Step 8: t-SNEs for Highest Ranked Genes</h3>
<p>To understand further the importance of the the highest ranked preferentially expressed genes, we’ll view their expression patterns within the t-SNEs.</p>
<p>For example, let’s look at AP-1 transcription factor components FOS, JUN &amp; JUNB, as they ranked highly.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_names_to_print</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'FOS'</span>, <span class="st">'JUN'</span>, <span class="st">'JUNB'</span><span class="op">)</span>

<span class="va">perpl_index</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">run_name</span> <span class="op">&lt;-</span> <span class="va">tSNE_run_names</span><span class="op">[</span><span class="va">perpl_index</span><span class="op">]</span>

<span class="kw">for</span><span class="op">(</span><span class="va">cohort_index</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="../reference/print_tSNE_coloured_by_gene.html">print_tSNE_coloured_by_gene</a></span><span class="op">(</span><span class="va">gene_names_to_print</span>,
                                <span class="va">CDX_tSNE_list_with_annotations</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">perpl_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">CDX_tRawCounts_list</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>,
                                <span class="va">run_name</span>, <span class="va">outDir</span> <span class="op">&lt;-</span> <span class="va">tSNEsDirs</span><span class="op">[[</span><span class="va">cohort_index</span><span class="op">]</span><span class="op">]</span>, <span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">2</span>,
                                <span class="va">all_HUGO_abbreviations_and_fullnames</span>, <span class="va">dataDir</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt;   FOS [Fos proto-oncogene, AP-1 transcription factor subunit]</span></code></pre></div>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-1.png" width="1152"></p>
<pre><code>#&gt;   JUN [Jun proto-oncogene, AP-1 transcription factor subunit]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-2.png" width="1152"></p>
<pre><code>#&gt;   JUNB [JunB proto-oncogene, AP-1 transcription factor subunit]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-3.png" width="1152"></p>
<pre><code>#&gt;   FOS [Fos proto-oncogene, AP-1 transcription factor subunit]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-4.png" width="1152"></p>
<pre><code>#&gt;   JUN [Jun proto-oncogene, AP-1 transcription factor subunit]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-5.png" width="1152"></p>
<pre><code>#&gt;   JUNB [JunB proto-oncogene, AP-1 transcription factor subunit]</code></pre>
<p><img src="tutorial_1_files/figure-html/unnamed-chunk-37-6.png" width="1152"></p>
<p>This is the end of CSCPlots tutorial 1.</p>
<p><strong>Now try <a href="tutorial_2.html">tutorial 2</a></strong>, which looks at CSCs from melanoma and evaluates their preferential expression. This is a more complex example from the perspective of having one cell population composed of multiple cell types from multiple samples.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Phil Carter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
